<!DOCTYPE html>
<html lang="en" dir="auto">
    <head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=5, user-scalable=no">
<meta name="description" content="hail Web: https://hail.is/ as with hail on cloud. Preparations We will need to set up an virtual environment as follows, module load python/3.7 had...">
<meta name="revised" content="2109d8f596b4cd432c7a5937f50b592827e6194f">
<meta name="author" content="cambridge-ceu">
<meta name="generator" content="jekyll-rtd-theme v2.0.10"><meta name="theme-color" content="#2980b9">
<title>hail 路 csd3</title>
<meta name="twitter:title" content="hail 路 csd3">
<meta name="twitter:description" content="hail Web: https://hail.is/ as with hail on cloud. Preparations We will need to set up an virtual environment as follows, module load python/3.7 had...">
<meta name="twitter:card" content="summary">
<meta name="twitter:site" content="@cambridge-ceu">
<meta name="twitter:url" content="/applications/hail.html">
<meta name="twitter:creator" content="@jekyll-rtd-theme v2.0.10">
<meta property="og:title" content="hail 路 csd3">
<meta property="og:description" content="hail Web: https://hail.is/ as with hail on cloud. Preparations We will need to set up an virtual environment as follows, module load python/3.7 had...">
<meta property="og:locale" content="en">
<meta property="og:url" content="/applications/hail.html">
<meta property="og:type" content="article">
<meta property="article:author" content="cambridge-ceu">
<meta property="article:published_time" content="2021-12-23T16:14:30+00:00">
<meta property="article:modified_time" content="2021-12-23T16:14:30+00:00">
<script type="application/ld+json">
    {
        "@context": "https://schema.org",
        "@type": "Article",
        "mainEntityOfPage": {
            "@type": "WebPage",
            "@id": "/applications/hail.html"
        },
        "headline": "hail 路 csd3",
        "image": [],
        "author": {
            "@type": "Person",
            "name": "cambridge-ceu"
        },
        "datePublished": "2021-12-23T16:14:30+00:00",
        "dateModified": "2021-12-23T16:14:30+00:00",
        "publisher": {
            "@type": "",
            "name": "cambridge-ceu",
            "logo": {
                "@type": "ImageObject",
                "url": ""
            }
        },
        "description": "hail Web: https://hail.is/ as with hail on cloud. Preparations We will need to set up an virtual environment as follows, module load python/3.7 had..."
    }
</script>

<link rel="dns-prefetch" href="https://cdn.jsdelivr.net"><link rel="prev" href="/applications/gwas2vcf.html"><link rel="next" href="/applications/HESS.html"><link rel="canonical" href="/applications/hail.html"><link rel="icon" type="image/svg+xml" href="/assets/images/favicon.svg"><link rel="icon" type="image/png" href="/assets/images/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="/assets/images/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="/assets/images/favicon-96x96.png" sizes="96x96"><link rel="mask-icon" href="/assets/images/favicon.svg" color="#2980b9"><link rel="apple-touch-icon" href="/assets/images/apple-touch-icon-300x300.jpg">
<link rel="stylesheet" href="/assets/css/theme.css"><script>
    window.ui = {
        title: "csd3",
        baseurl: "",
        i18n: {
            search_results: "Search Results",
            search_results_found: "Search finished, found # page(s) matching the search query.",
            search_results_not_found: "Your search did not match any documents, please make sure that all characters are spelled correctly!"
        }
    };
</script>
</head>
    <body class="container"><div class="sidebar-wrap overflow-hidden">
    <div class="sidebar height-full overflow-y-scroll overflow-x-hidden">
        <div class="header d-flex flex-column p-3 text-center">
            <div class="title pb-1">
                <a class="h4 no-underline py-1 px-2 rounded-1" href="/" title="A collection of CSD3 notes">
                    <i class="fa fa-home"></i> csd3
                </a>
            </div>
            <span class="version"></span>
            <form class="search pt-2" action="/search.html" method="get" autocomplete="off">
                <input class="form-control input-block input-sm" type="text" name="q" placeholder="Search docs...">
            </form>
        </div>
        <div class="toctree py-2" data-spy="affix" role="navigation" aria-label="main navigation">
            


<ul>
</ul>




    
    <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/systems/">
        The system
    </a><ul>
    

        <li class="toc level-1 " data-sort="1" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/system.html">1. System</a>
        </li>

        <li class="toc level-1 " data-sort="2" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/login.html">2. Login</a>
        </li>

        <li class="toc level-1 " data-sort="3" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/directories.html">3. Directories</a>
        </li>

        <li class="toc level-1 " data-sort="4" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/policies.html">4. Policies</a>
        </li>

        <li class="toc level-1 " data-sort="5" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/email.html">5. email</a>
        </li>

        <li class="toc level-1 " data-sort="6" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/software.html">6. Software</a>
        </li>

        <li class="toc level-1 " data-sort="7" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/training.html">7. Training</a>
        </li>

        <li class="toc level-1 " data-sort="8" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/acknowledgement.html">8. Acknowledgement</a>
        </li>

        <li class="toc level-1 " data-sort="8" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/contacts.html">8. Contacts</a>
        </li></ul>



    
    <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/cardio/">
        cardio
    </a><ul>
    </ul>



    
    <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/applications/">
        Applications
    </a><ul>
    

        <li class="toc level-1 " data-sort="1" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/ABCtoolbox.html">1. ABCtoolbox</a>
        </li>

        <li class="toc level-1 " data-sort="2" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/AEGIS.html">2. AEGIS</a>
        </li>

        <li class="toc level-1 " data-sort="3" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/akt.html">3. akt</a>
        </li>

        <li class="toc level-1 " data-sort="4" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/ANNOVAR.html">4. ANNOVAR</a>
        </li>

        <li class="toc level-1 " data-sort="5" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/aria2.html">5. aria2</a>
        </li>

        <li class="toc level-1 " data-sort="6" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/bcftools.html">6. bcftools</a>
        </li>

        <li class="toc level-1 " data-sort="7" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/bedops.html">7. bedops</a>
        </li>

        <li class="toc level-1 " data-sort="8" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/bgenix.html">8. bgenix</a>
        </li>

        <li class="toc level-1 " data-sort="9" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/CaVEMaN.html">9. CaVEMaN</a>
        </li>

        <li class="toc level-1 " data-sort="10" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/citeproc.html">10. citeproc</a>
        </li>

        <li class="toc level-1 " data-sort="11" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/Cytoscape.html">11. Cytoscape</a>
        </li>

        <li class="toc level-1 " data-sort="12" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/DosageConverter.html">12. DosageConverter</a>
        </li>

        <li class="toc level-1 " data-sort="13" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/gcta.html">13. GCTA</a>
        </li>

        <li class="toc level-1 " data-sort="14" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/gsutil.html">14. gsutil</a>
        </li>

        <li class="toc level-1 " data-sort="15" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/gwas2vcf.html">15. gwas2vcf</a>
        </li>

        <li class="toc level-1 current" data-sort="16" data-level="1">
            <a class="d-flex flex-items-baseline current" href="/applications/hail.html">16. hail</a>
        </li>

        <li class="toc level-1 " data-sort="17" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/HESS.html">17. HESS</a>
        </li>

        <li class="toc level-1 " data-sort="18" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/KentUtils.html">18. KentUtils</a>
        </li>

        <li class="toc level-1 " data-sort="19" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/lapack.html">19. lapack</a>
        </li>

        <li class="toc level-1 " data-sort="20" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/ldsc.html">20. ldsc</a>
        </li>

        <li class="toc level-1 " data-sort="21" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/LEMMA.html">21. LEMMA</a>
        </li>

        <li class="toc level-1 " data-sort="22" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/MetaXcan.html">22. MetaXcan</a>
        </li>

        <li class="toc level-1 " data-sort="23" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/pandoc.html">23. pandoc</a>
        </li>

        <li class="toc level-1 " data-sort="24" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/pca_projection.html">24. PCA projection</a>
        </li>

        <li class="toc level-1 " data-sort="25" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/plink.html">25. PLINK</a>
        </li>

        <li class="toc level-1 " data-sort="26" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/polyphen-2.html">26. polyphen-2</a>
        </li>

        <li class="toc level-1 " data-sort="27" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/poppler.html">27. poppler</a>
        </li>

        <li class="toc level-1 " data-sort="28" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/PRSice.html">28. PRSice</a>
        </li>

        <li class="toc level-1 " data-sort="29" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/PRSoS.html">29. PRSoS</a>
        </li>

        <li class="toc level-1 " data-sort="30" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/pspp.html">30. pspp</a>
        </li>

        <li class="toc level-1 " data-sort="31" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/qpdf.html">31. qpdf</a>
        </li>

        <li class="toc level-1 " data-sort="32" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/regenie.html">32. regenie</a>
        </li>

        <li class="toc level-1 " data-sort="33" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/R.html">33. R</a>
        </li>

        <li class="toc level-1 " data-sort="34" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/RStudio.html">34. RStudio</a>
        </li>

        <li class="toc level-1 " data-sort="35" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/shapeit3.html">35. shapeit3</a>
        </li>

        <li class="toc level-1 " data-sort="36" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/SMR.html">36. SMR</a>
        </li>

        <li class="toc level-1 " data-sort="37" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/snakemake.html">37. snakemake</a>
        </li>

        <li class="toc level-1 " data-sort="38" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/SNPTEST.html">38. SNPTEST</a>
        </li>

        <li class="toc level-1 " data-sort="39" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/tabix.html">39. tabix</a>
        </li>

        <li class="toc level-1 " data-sort="40" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/VEP.html">40. VEP</a>
        </li>

        <li class="toc level-1 " data-sort="41" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/VScode.html">41. Visual Studio code</a>
        </li>
        <li class="toc level-1">


    <a class="d-flex flex-items-baseline" href="/applications/R/">
        R packages
    </a><ul>
    

        <li class="toc level-2 " data-sort="1" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/brms.html">1. R/brms</a>
        </li>

        <li class="toc level-2 " data-sort="2" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/EasyQC.html">2. R/EasyQC</a>
        </li>

        <li class="toc level-2 " data-sort="3" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/gnn.html">3. R/gnn</a>
        </li>

        <li class="toc level-2 " data-sort="4" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/gsl.html">4. R/gsl</a>
        </li>

        <li class="toc level-2 " data-sort="5" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/LDlinkR.html">5. R/LDlinkR</a>
        </li>

        <li class="toc level-2 " data-sort="6" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/magick.html">6. R/magick</a>
        </li>

        <li class="toc level-2 " data-sort="7" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/PhenoScanner.html">7. R/PhenoScanner</a>
        </li>

        <li class="toc level-2 " data-sort="8" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/PKI.html">8. R/PKI</a>
        </li>

        <li class="toc level-2 " data-sort="9" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/plumber.html">9. R/plumber</a>
        </li>

        <li class="toc level-2 " data-sort="10" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/protolite.html">10. R/protolite</a>
        </li>

        <li class="toc level-2 " data-sort="11" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/Rfast.html">11. R/Rfast</a>
        </li>

        <li class="toc level-2 " data-sort="12" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/rgdal.html">12. R/rgdal</a>
        </li>

        <li class="toc level-2 " data-sort="13" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/rgeos.html">13. R/rgeos</a>
        </li>

        <li class="toc level-2 " data-sort="14" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/Rhdf5lib.html">14. R/Rhdf5lib</a>
        </li>

        <li class="toc level-2 " data-sort="15" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/rjags.html">15. R/rjags</a>
        </li>

        <li class="toc level-2 " data-sort="16" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/rJava.html">16. R/rJava</a>
        </li>

        <li class="toc level-2 " data-sort="17" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/rstan.html">17. R/rstan</a>
        </li>

        <li class="toc level-2 " data-sort="18" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/SAIGE.html">18. R/SAIGE 0.36.6 and 0.39.2</a>
        </li>

        <li class="toc level-2 " data-sort="19" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/sf.html">19. R/sf</a>
        </li>

        <li class="toc level-2 " data-sort="20" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/snpnet.html">20. R/snpnet</a>
        </li>

        <li class="toc level-2 " data-sort="21" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/sojo.html">21. R/sojo</a>
        </li>

        <li class="toc level-2 " data-sort="22" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/xlsx.html">22. R/xlsx</a>
        </li></ul>
</li></ul>



    
    <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/writing/">
        Writing Related
    </a><ul>
    

        <li class="toc level-1 " data-sort="1" data-level="1">
            <a class="d-flex flex-items-baseline " href="/writing/front-matter.html">1. Front matter</a>
        </li>

        <li class="toc level-1 " data-sort="2" data-level="1">
            <a class="d-flex flex-items-baseline " href="/writing/list.html">2. List directory</a>
        </li>

        <li class="toc level-1 " data-sort="3" data-level="1">
            <a class="d-flex flex-items-baseline " href="/writing/mathjax.html">3. Mathjax</a>
        </li>

        <li class="toc level-1 " data-sort="4" data-level="1">
            <a class="d-flex flex-items-baseline " href="/writing/admonitions.html">4. Admonition card</a>
        </li>
        <li class="toc level-1">


    <a class="d-flex flex-items-baseline" href="/writing/test/">
        Test Documentation
    </a><ul>
    

        <li class="toc level-2 " data-sort="1" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/markdown.html">1. Markdown Elements</a>
        </li>

        <li class="toc level-2 " data-sort="2" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/toasts.html">2. Toasts Card</a>
        </li>

        <li class="toc level-2 " data-sort="3" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/codes.html">3. Code Blocks</a>
        </li>

        <li class="toc level-2 " data-sort="4" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/mermaid.html">4. Mermaid Test</a>
        </li>

        <li class="toc level-2 " data-sort="5" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/emoji.html">5. Emoji Test</a>
        </li>

        <li class="toc level-2 " data-sort="6" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/gist.html">6. Gist Test</a>
        </li>

        <li class="toc level-2 " data-sort="7" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/avatar.html">7. Avatar Test</a>
        </li>

        <li class="toc level-2 " data-sort="8" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/mentions.html">8. Mentions Test</a>
        </li>

        <li class="toc level-2 " data-sort="9" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/fonts.html">9. Fonts Test</a>
        </li>

        <li class="toc level-2 " data-sort="10" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/utilities.html">10. Primer Utilities Test</a>
        </li></ul>
</li></ul>

        </div>
    </div>
</div>
<div class="content-wrap">
    <div class="header d-flex flex-justify-between p-2 hide-lg hide-xl" aria-label="top navigation">
        <button id="toggle" aria-label="Toggle menu" class="btn-octicon p-2 m-0 text-white" type="button">
            <i class="fa fa-bars"></i>
        </button>
        <div class="title flex-1 d-flex flex-justify-center">
            <a class="h4 no-underline py-1 px-2 rounded-1" href="/">csd3</a>
        </div>
    </div>
    <div class="content p-3 p-sm-5">
        <div class="navigation-top d-flex flex-justify-between">
    <ul class="breadcrumb" role="navigation" aria-label="breadcrumbs navigation">
        <li class="breadcrumb-item">
            <a class="no-underline" href="/" title="/">
                <i class="fa fa-home"></i>
            </a>
        </li><li class="breadcrumb-item" ><a href="/applications/">applications</a></li><li class="breadcrumb-item" aria-current="page">hail.md</li></ul>
    <a class="edit" href="https://github.com/cambridge-ceu/csd3/edit/gh-pages/applications/hail.md" title="Edit on GitHub" rel="noreferrer" target="_blank">
        <i class="fa fa-edit"></i>
    </a>
</div>

        <hr>
        <div role="main" itemscope="itemscope" itemtype="https://schema.org/Article">
            <div class="markdown-body" itemprop="articleBody">
                <h1 id="hail">hail</h1>

<p>Web: <a href="https://hail.is/">https://hail.is/</a> as with <a href="https://github.com/danking/hail-cloud-docs/blob/master/how-to-cloud.md">hail on cloud</a>.</p>

<h2 id="preparations">Preparations</h2>

<p>We will need to set up an virtual environment as follows,</p>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>module load python/3.7 hadoop/2.7.7
virtualenv py37
<span class="nb">source </span>py37/bin/activate
</code></pre> </div></div>

<p>Later, only two of the commands are necessary, i.e.,</p>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>module load python/3.7 hadoop/2.7.7
<span class="nb">source </span>py37/bin/activate
</code></pre> </div></div>

<h2 id="installation">Installation</h2>

<p>We proceed with,</p>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>pip <span class="nb">install </span>hail
pip <span class="nb">install </span>gnomAD
pip <span class="nb">install </span>pyensembl
pip <span class="nb">install </span>varcode
pyensembl <span class="nb">install</span> <span class="nt">--release</span> 99 <span class="nt">--species</span> human
</code></pre> </div></div>

<p>where <code class="language-plaintext highlighter-rouge notranslate">gnomAD</code>, <code class="language-plaintext highlighter-rouge notranslate">pyensembl</code> (<a href="https://github.com/openvax/pyensembl">https://github.com/openvax/pyensembl</a>) and <code class="language-plaintext highlighter-rouge notranslate">varcode</code> (<a href="https://github.com/openvax/varcode">https://github.com/openvax/varcode</a>) are optional, and we could invoke Python as follows,</p>

<div class="language-python highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="n">python</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="n">hl</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">mt</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">balding_nichols_model</span><span class="p">(</span><span class="n">n_populations</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">n_samples</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">n_variants</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">mt</span><span class="p">.</span><span class="n">describe</span><span class="p">()</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">mt</span><span class="p">.</span><span class="n">summarize</span><span class="p">()</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">mt</span><span class="p">.</span><span class="n">count</span><span class="p">()</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">mt</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">mt</span><span class="p">.</span><span class="n">write</span><span class="p">(</span><span class="s">"hail"</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">at</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">import_vcf</span><span class="p">(</span><span class="s">"INTERVAL.vcf.bgz"</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">hl</span><span class="p">.</span><span class="n">export_vcf</span><span class="p">(</span><span class="n">at</span><span class="p">,</span><span class="s">"at.vcf.bgz"</span><span class="p">)</span>
</code></pre> </div></div>

<p>Alternatively, we could also start a session using <code class="language-plaintext highlighter-rouge notranslate">ipython</code>.</p>

<h2 id="tutorials">Tutorials</h2>

<p>We can get the tutorials at the current directory,</p>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>wget <span class="nt">-qO-</span> https://hail.is/docs/0.2/tutorials.tar.gz | <span class="nb">tar </span>xf -
</code></pre> </div></div>

<p>and cut/paste code from the hail website to our command-line session. A better quality can be achieved with a browser, which unfortunately does not function so well on csd3 and requires some tweak from elsewhere,</p>

<p>First, from a csd3 session we issue commands,</p>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="nb">hostname
</span>jupyter notebook tutorials/ <span class="nt">--ip</span><span class="o">=</span>127.0.0.1 <span class="nt">--no-browser</span> <span class="nt">--port</span> 8081
</code></pre> </div></div>

<p>and for this instance we have <code class="language-plaintext highlighter-rouge notranslate">hostname</code> as <code class="language-plaintext highlighter-rouge notranslate">login-e-12</code> and additional information as follows,</p>

<div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>[I 20:21:59.258 NotebookApp] Serving notebooks from local directory: /rds/project/jmmh2/rds-jmmh2-results/public/gwas/ukb_exomes/tutorials
[I 20:21:59.258 NotebookApp] The Jupyter Notebook is running at:
[I 20:21:59.258 NotebookApp] http://127.0.0.1:8081/?token=2d4ba02d3d0d782ad1dbafc73f96a3fc3d0bdff4573899c9
[I 20:21:59.258 NotebookApp]  or http://127.0.0.1:8081/?token=2d4ba02d3d0d782ad1dbafc73f96a3fc3d0bdff4573899c9
[I 20:21:59.258 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 20:21:59.300 NotebookApp]

    To access the notebook, open this file in a browser:
        file:///home/jhz22/.local/share/jupyter/runtime/nbserver-235186-open.html
    Or copy and paste one of these URLs:
        http://127.0.0.1:8081/?token=2d4ba02d3d0d782ad1dbafc73f96a3fc3d0bdff4573899c9
     or http://127.0.0.1:8081/?token=2d4ba02d3d0d782ad1dbafc73f96a3fc3d0bdff4573899c9
</code></pre> </div></div>

<p>from a console other than a csd3 session (e.g., from a <code class="language-plaintext highlighter-rouge notranslate">srcf</code> or <code class="language-plaintext highlighter-rouge notranslate">ds</code> or Windows <code class="language-plaintext highlighter-rouge notranslate">virtualbox</code> virtual machine) we access the csd3 session above.</p>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>ssh <span class="nt">-4</span> <span class="nt">-L</span> 8081:127.0.0.1:8081 <span class="nt">-fN</span> login-e-12.hpc.cam.ac.uk
firefox http://127.0.0.1:8081/?token<span class="o">=</span>2d4ba02d3d0d782ad1dbafc73f96a3fc3d0bdff4573899c9
</code></pre> </div></div>

<p>and run the tutorial from firefox. Similar process is also described at the system/software section on Python.</p>

<p>For convenience, we list the code from the tutorials on 1000Genomes data, see https://notebook.community/danking/hail/notebook/images/hail/resources/Hail-Workshop-Notebook.</p>

<div class="language-python highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="n">hl</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="nb">dir</span><span class="o">=</span><span class="n">os</span><span class="p">.</span><span class="n">environ</span><span class="p">[</span><span class="s">'HPC_WORK'</span><span class="p">]</span><span class="o">+</span><span class="s">"/lib64/"</span>
<span class="n">files</span><span class="o">=</span><span class="nb">dir</span><span class="o">+</span><span class="s">"libblas.so:"</span><span class="o">+</span><span class="nb">dir</span><span class="o">+</span><span class="s">"libcblas.so:"</span><span class="o">+</span><span class="nb">dir</span><span class="o">+</span><span class="s">"liblapack.so"</span>
<span class="k">print</span><span class="p">(</span><span class="n">files</span><span class="p">)</span>
<span class="c1"># os.environ['PYENSEMBL_CACHE_DIR'] = '/custom/cache/dir'
</span><span class="n">hl</span><span class="p">.</span><span class="n">init</span><span class="p">(</span><span class="n">spark_conf</span><span class="o">=</span><span class="p">{</span><span class="s">"spark.executor.extraClassPath"</span><span class="p">:</span> <span class="n">files</span><span class="p">})</span>
<span class="c1"># Gene information
</span><span class="n">gene_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">import_table</span><span class="p">(</span><span class="s">'tutorials/data/ensembl_gene_annotations.txt'</span><span class="p">,</span> <span class="n">impute</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">gene_ht</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">gene_ht</span><span class="p">.</span><span class="n">count</span><span class="p">()</span>
<span class="n">gene_ht</span> <span class="o">=</span> <span class="n">gene_ht</span><span class="p">.</span><span class="n">transmute</span><span class="p">(</span><span class="n">interval</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">locus_interval</span><span class="p">(</span><span class="n">gene_ht</span><span class="p">[</span><span class="s">'Chromosome'</span><span class="p">],</span>
                                                         <span class="n">gene_ht</span><span class="p">[</span><span class="s">'Gene start'</span><span class="p">],</span>
                                                         <span class="n">gene_ht</span><span class="p">[</span><span class="s">'Gene end'</span><span class="p">],</span>
                                                         <span class="n">reference_genome</span><span class="o">=</span><span class="s">'GRCh37'</span><span class="p">))</span>
<span class="n">gene_ht</span> <span class="o">=</span> <span class="n">gene_ht</span><span class="p">.</span><span class="n">key_by</span><span class="p">(</span><span class="s">'interval'</span><span class="p">)</span>
<span class="c1"># 1000Genomes data from the tutorials
</span><span class="n">sa</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">import_table</span><span class="p">(</span><span class="s">'tutorials/data/1kg_annotations.txt'</span><span class="p">,</span> <span class="n">impute</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s">'Sample'</span><span class="p">)</span>
<span class="n">sa</span><span class="p">.</span><span class="n">describe</span><span class="p">()</span>
<span class="n">sa</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">read_matrix_table</span><span class="p">(</span><span class="s">'tutorials/data/1kg.mt'</span><span class="p">)</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="p">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="n">gene_info</span> <span class="o">=</span> <span class="n">gene_ht</span><span class="p">[</span><span class="n">mt</span><span class="p">.</span><span class="n">locus</span><span class="p">])</span>
<span class="n">mt</span><span class="p">.</span><span class="n">s</span><span class="p">.</span><span class="n">show</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="n">mt</span><span class="p">.</span><span class="n">locus</span><span class="p">.</span><span class="n">show</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="n">hl</span><span class="p">.</span><span class="n">summarize_variants</span><span class="p">(</span><span class="n">mt</span><span class="p">)</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="p">.</span><span class="n">annotate_cols</span><span class="p">(</span><span class="n">pheno</span> <span class="o">=</span> <span class="n">sa</span><span class="p">[</span><span class="n">mt</span><span class="p">.</span><span class="n">s</span><span class="p">])</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">sample_qc</span><span class="p">(</span><span class="n">mt</span><span class="p">)</span>
<span class="n">mt</span><span class="p">.</span><span class="n">sample_qc</span><span class="p">.</span><span class="n">describe</span><span class="p">()</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">plot</span><span class="p">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">mt</span><span class="p">.</span><span class="n">sample_qc</span><span class="p">.</span><span class="n">r_het_hom_var</span><span class="p">,</span>
                    <span class="n">y</span><span class="o">=</span><span class="n">mt</span><span class="p">.</span><span class="n">sample_qc</span><span class="p">.</span><span class="n">call_rate</span><span class="p">)</span>
<span class="n">show</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="p">.</span><span class="n">filter_cols</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">sample_qc</span><span class="p">.</span><span class="n">dp_stats</span><span class="p">.</span><span class="n">mean</span> <span class="o">&gt;=</span> <span class="mi">4</span><span class="p">)</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="p">.</span><span class="n">filter_cols</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">sample_qc</span><span class="p">.</span><span class="n">call_rate</span> <span class="o">&gt;=</span> <span class="mf">0.97</span><span class="p">)</span>
<span class="n">mt</span><span class="p">.</span><span class="n">aggregate_entries</span><span class="p">(</span><span class="n">hl</span><span class="p">.</span><span class="n">agg</span><span class="p">.</span><span class="n">fraction</span><span class="p">(</span><span class="n">hl</span><span class="p">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">GT</span><span class="p">)))</span>
<span class="n">ab</span> <span class="o">=</span> <span class="n">mt</span><span class="p">.</span><span class="n">AD</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">hl</span><span class="p">.</span><span class="nb">sum</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">AD</span><span class="p">)</span>
<span class="n">filter_condition_ab</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">hl</span><span class="p">.</span><span class="n">case</span><span class="p">()</span>
    <span class="p">.</span><span class="n">when</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">GT</span><span class="p">.</span><span class="n">is_hom_ref</span><span class="p">(),</span> <span class="n">ab</span> <span class="o">&lt;=</span> <span class="mf">0.1</span><span class="p">)</span>
    <span class="p">.</span><span class="n">when</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">GT</span><span class="p">.</span><span class="n">is_het</span><span class="p">(),</span> <span class="p">(</span><span class="n">ab</span> <span class="o">&gt;=</span> <span class="mf">0.25</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ab</span> <span class="o">&lt;=</span> <span class="mf">0.75</span><span class="p">))</span>
    <span class="p">.</span><span class="n">default</span><span class="p">(</span><span class="n">ab</span> <span class="o">&gt;=</span> <span class="mf">0.9</span><span class="p">)</span> <span class="c1"># hom-var
</span><span class="p">)</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="p">.</span><span class="n">filter_entries</span><span class="p">(</span><span class="n">filter_condition_ab</span><span class="p">)</span>
<span class="n">mt</span><span class="p">.</span><span class="n">aggregate_entries</span><span class="p">(</span><span class="n">hl</span><span class="p">.</span><span class="n">agg</span><span class="p">.</span><span class="n">fraction</span><span class="p">(</span><span class="n">hl</span><span class="p">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">GT</span><span class="p">)))</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">variant_qc</span><span class="p">(</span><span class="n">mt</span><span class="p">)</span>
<span class="n">mt</span><span class="p">.</span><span class="n">variant_qc</span><span class="p">.</span><span class="n">describe</span><span class="p">()</span>
<span class="n">mt</span><span class="p">.</span><span class="n">variant_qc</span><span class="p">.</span><span class="n">AF</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="p">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">hl</span><span class="p">.</span><span class="nb">min</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">variant_qc</span><span class="p">.</span><span class="n">AF</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mf">1e-6</span><span class="p">)</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="p">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">variant_qc</span><span class="p">.</span><span class="n">p_value_hwe</span> <span class="o">&gt;</span> <span class="mf">0.005</span><span class="p">)</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="p">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="n">gene_info</span> <span class="o">=</span> <span class="n">gene_ht</span><span class="p">[</span><span class="n">mt</span><span class="p">.</span><span class="n">locus</span><span class="p">])</span>
<span class="n">mt</span><span class="p">.</span><span class="n">gene_info</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">burden_mt</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">mt</span>
    <span class="p">.</span><span class="n">group_rows_by</span><span class="p">(</span><span class="n">gene</span> <span class="o">=</span> <span class="n">mt</span><span class="p">.</span><span class="n">gene_info</span><span class="p">[</span><span class="s">'Gene name'</span><span class="p">])</span>
    <span class="p">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">n_variants</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">agg</span><span class="p">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">GT</span><span class="p">.</span><span class="n">n_alt_alleles</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">))</span>
<span class="p">)</span>
<span class="n">burden_mt</span><span class="p">.</span><span class="n">describe</span><span class="p">()</span>
<span class="n">pca_eigenvalues</span><span class="p">,</span> <span class="n">pca_scores</span><span class="p">,</span> <span class="n">pca_loadings</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">hwe_normalized_pca</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">GT</span><span class="p">,</span> <span class="n">compute_loadings</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">pca_eigenvalues</span>
<span class="n">pca_scores</span><span class="p">.</span><span class="n">describe</span><span class="p">()</span>
<span class="n">pca_scores</span><span class="p">.</span><span class="n">scores</span><span class="p">[</span><span class="mi">0</span><span class="p">].</span><span class="n">show</span><span class="p">()</span>
<span class="n">pca_loadings</span><span class="p">.</span><span class="n">describe</span><span class="p">()</span>
<span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="p">.</span><span class="n">annotate_cols</span><span class="p">(</span><span class="n">pca</span> <span class="o">=</span> <span class="n">pca_scores</span><span class="p">[</span><span class="n">mt</span><span class="p">.</span><span class="n">s</span><span class="p">])</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">plot</span><span class="p">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">mt</span><span class="p">.</span><span class="n">pca</span><span class="p">.</span><span class="n">scores</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                    <span class="n">mt</span><span class="p">.</span><span class="n">pca</span><span class="p">.</span><span class="n">scores</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                    <span class="n">label</span><span class="o">=</span><span class="n">mt</span><span class="p">.</span><span class="n">pheno</span><span class="p">.</span><span class="n">SuperPopulation</span><span class="p">)</span>
<span class="n">show</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="c1"># errors from here though hl.init() above
# https://discuss.hail.is/t/undefined-symbol-cblas-dgemm/1488
</span><span class="n">gwas</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">linear_regression_rows</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">mt</span><span class="p">.</span><span class="n">pheno</span><span class="p">.</span><span class="n">CaffeineConsumption</span><span class="p">,</span>
                                 <span class="n">x</span><span class="o">=</span><span class="n">mt</span><span class="p">.</span><span class="n">GT</span><span class="p">.</span><span class="n">n_alt_alleles</span><span class="p">(),</span>
                                 <span class="n">covariates</span><span class="o">=</span><span class="p">[</span><span class="mf">1.0</span><span class="p">])</span>
<span class="n">gwas</span><span class="p">.</span><span class="n">describe</span><span class="p">()</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">plot</span><span class="p">.</span><span class="n">manhattan</span><span class="p">(</span><span class="n">gwas</span><span class="p">.</span><span class="n">p_value</span><span class="p">)</span>
<span class="n">show</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">plot</span><span class="p">.</span><span class="n">qq</span><span class="p">(</span><span class="n">gwas</span><span class="p">.</span><span class="n">p_value</span><span class="p">)</span>
<span class="n">show</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="n">gwas</span> <span class="o">=</span> <span class="n">hl</span><span class="p">.</span><span class="n">linear_regression_rows</span><span class="p">(</span>
    <span class="n">y</span><span class="o">=</span><span class="n">mt</span><span class="p">.</span><span class="n">pheno</span><span class="p">.</span><span class="n">CaffeineConsumption</span><span class="p">,</span>
    <span class="n">x</span><span class="o">=</span><span class="n">mt</span><span class="p">.</span><span class="n">GT</span><span class="p">.</span><span class="n">n_alt_alleles</span><span class="p">(),</span>
    <span class="n">covariates</span><span class="o">=</span><span class="p">[</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">mt</span><span class="p">.</span><span class="n">pheno</span><span class="p">.</span><span class="n">isFemale</span><span class="p">,</span> <span class="n">mt</span><span class="p">.</span><span class="n">pca</span><span class="p">.</span><span class="n">scores</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">mt</span><span class="p">.</span><span class="n">pca</span><span class="p">.</span><span class="n">scores</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">mt</span><span class="p">.</span><span class="n">pca</span><span class="p">.</span><span class="n">scores</span><span class="p">[</span><span class="mi">2</span><span class="p">]])</span><span class="sb">``</span><span class="err">`</span>
</code></pre> </div></div>

<h2 id="resources">Resources</h2>

<p>Some files can be made available with <code class="language-plaintext highlighter-rouge notranslate">gsutil</code> installed, e.g.,</p>

<ul>
  <li>The HGDP release 3.1, <code class="language-plaintext highlighter-rouge notranslate">gs://gcp-public-data--gnomad/release/3.1</code>.
    <ul>
      <li>See also, <code class="language-plaintext highlighter-rouge notranslate">gs://gnomad-public/release</code>.</li>
    </ul>
  </li>
  <li>UK Biobank results, <code class="language-plaintext highlighter-rouge notranslate">gs://hail-datasets/ukbb_imputed_v3_gwas_results_both_sexes.GRCh37.mt</code>.</li>
  <li>Pan-ancestry genetic analysis of the UK Biobank, <a href="https://pan.ukbb.broadinstitute.org/docs/hail-format">https://pan.ukbb.broadinstitute.org/docs/hail-format</a>.</li>
  <li>Exome-based association statistics, see below.</li>
</ul>

<h2 id="genebass">genebass</h2>

<p>Web: <a href="https://genebass.org/">https://genebass.org/</a></p>

<h3 id="download">Download</h3>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="c"># genebass and csd3 locations</span>
<span class="nb">export </span><span class="nv">src</span><span class="o">=</span>gs://ukbb-exome-public/300k/results
<span class="nb">export </span><span class="nv">dest</span><span class="o">=</span>~/rds/results/public/gwas/ukb_exomes
<span class="nb">mkdir</span> <span class="nt">-p</span> <span class="k">${</span><span class="nv">dest</span><span class="k">}</span>
<span class="nb">cd</span> <span class="k">${</span><span class="nv">dest</span><span class="k">}</span>
<span class="c"># Gene burden results</span>
gsutil <span class="nt">-m</span> <span class="nb">cp</span> <span class="nt">-r</span> <span class="k">${</span><span class="nv">src</span><span class="k">}</span>/results.mt <span class="nb">.</span>
<span class="c"># Single variant association results</span>
gsutil <span class="nt">-m</span> <span class="nb">cp</span> <span class="nt">-r</span> <span class="k">${</span><span class="nv">src</span><span class="k">}</span>/variant_results.mt <span class="nb">.</span>
</code></pre> </div></div>

<h3 id="access">Access</h3>

<p>We could run the script directly as follows,</p>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="nb">cd</span> ~/rds/results/public/gwas/ukb_exomes
python <span class="o">&lt;&lt;</span><span class="no">END</span><span class="sh">

import hail as hl

# Gene (burden) results
gr = hl.read_matrix_table('results.mt')

## Summaries
gr.count()
gr.describe()
gr.show()
gr.cols().show()
gr.rows().show()
gr.entry.take(5)
gr.summarize()

## filtering
FGR = gr.filter_rows(gr.gene_symbol=="FGR")
FGR1065 = FGR.filter_cols(FGR.coding == '1065')
FGR1065burden = FGR1065.select_entries(FGR1065.BETA_Burden)

## col operations
trait_gr = gr.cols().collect_by_key()
trait_gr.group_by('trait_type').aggregate(n_phenos=hl.agg.count()).show()
gr_cols = gr.cols()
gr_cols.count()
gr_cols.show(truncate=40, width=85)
gr_cols.phenocode.show()
gr_cols.filter(gb_cols.phenocode == '1777').show()
fields_to_drop = ['category', 'coding_description', 'description_more', 'inv_normalized',
                  'n_cases_both_sexes', 'n_cases_females', 'n_cases_males', 'saige_version']
gr_cols_reduced=gb_cols.drop(*fields_to_drop)

# Variant results
vr = hl.read_matrix_table('variant_results.mt')
hl.summarize_variants(vr)
vr.count()
vr.describe()
vr.select_cols().show(1)
vr.select_rows().show(1)
vr.locus.show(1)
co1777=vr.filter_cols(vr.coding=="1777").show()
co1777.entry
co1777.entry.take(5)

# IL12B
</span><span class="no">END
</span></code></pre> </div></div>

<h2 id="reference">Reference</h2>

<p>Karczewski, K.J. et al. Systematic single-variant and gene-based association testing of 3,700 phenotypes in 281,850 UK Biobank exomes. medRxiv, 2021.06.19.21259117 (2021).</p>

            </div>
        </div>
        <div class="navigation-bottom d-flex flex-justify-between py-3" role="navigation" aria-label="footer navigation">
            <div class="prev"><a href="/applications/gwas2vcf.html" class="btn" title="gwas2vcf" accesskey="p" rel="prev">
                        <i class="fa fa-arrow-circle-left"></i> Previous
                    </a></div>
            <div class="next"><a href="/applications/HESS.html" class="btn" title="HESS" accesskey="n" rel="next">
                        Next <i class="fa fa-arrow-circle-right"></i>
                    </a></div>
        </div><hr>

<div class="copyright text-center text-gray" role="contentinfo">
    <i class="fa fa-copyright"></i>
    <span class="time">2021,</span>
    <a class="text-gray" href="https://github.com/cambridge-ceu" rel="noreferrer" target="_blank">cambridge-ceu</a>
    Revision
    <a class="text-gray" href="https://github.com/cambridge-ceu/csd3/commit/2109d8f596b4cd432c7a5937f50b592827e6194f" title="2109d8f596b4cd432c7a5937f50b592827e6194f" rel="noreferrer" target="_blank">2109d8f</a>
    
        <br>
        <div class="generator">
            Built with
            <a href="https://pages.github.com" rel="noreferrer" target="_blank" title="github-pages v209">GitHub Pages</a>
            using a
            <a href="https://github.com/rundocs/jekyll-rtd-theme" rel="noreferrer" target="_blank" title="jekyll-rtd-theme v2.0.10">theme</a>
            provided by
            <a href="https://rundocs.io" rel="noreferrer" target="_blank">RunDocs</a>.
        </div>
    
</div>

    </div>
</div>
<div class="addons-wrap d-flex flex-column overflow-y-auto">
    <div class="status d-flex flex-justify-between p-2">
        <div class="title p-1">
            <i class="fa fa-book"></i> csd3
        </div>
        <div class="branch p-1">
            <span class="name">
                gh-pages
            </span>
            <i class="fa fa-caret-down"></i>
        </div>
    </div>
    <div class="addons d-flex flex-column height-full p-2 d-none">
        <dl>
            <dt>GitHub</dt>
            <dd>
                <a href="https://github.com/cambridge-ceu/csd3" title="Stars: ">
                    <i class="fa fa-github"></i>
                    Homepage
                </a>
            </dd>
            <dd>
                <a href="" title="Open issues: ">
                    <i class="fa fa-question-circle-o"></i>
                    Issues
                </a>
            </dd>
            <dd>
                <a href="https://github.com/cambridge-ceu/csd3/zipball/gh-pages" title="Size:  Kb">
                    <i class="fa fa-download"></i>
                    Download
                </a>
            </dd>
        </dl>
        <hr>
        <div class="license f6 pb-2">
            This
            <a href="/" title="csd3">Software</a>
            is under the terms of
            <a href="https://github.com/cambridge-ceu/csd3">The Unlicense</a>.
        </div>
    </div>
</div>
<script src="/assets/js/jquery.min.js"></script><script src="/assets/js/theme.js"></script>


</body>
</html>
