<!DOCTYPE html>
<html lang="en" dir="auto">
    <head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=5, user-scalable=no">
<meta name="description" content="PCA projection Official page: https://github.com/covid19-hg/pca_projection. git clone https://github.com/covid19-hg/pca_projection cd pca_projectio...">
<meta name="revised" content="2109d8f596b4cd432c7a5937f50b592827e6194f">
<meta name="author" content="cambridge-ceu">
<meta name="generator" content="jekyll-rtd-theme v2.0.10"><meta name="theme-color" content="#2980b9">
<title>PCA projection · csd3</title>
<meta name="twitter:title" content="PCA projection · csd3">
<meta name="twitter:description" content="PCA projection Official page: https://github.com/covid19-hg/pca_projection. git clone https://github.com/covid19-hg/pca_projection cd pca_projectio...">
<meta name="twitter:card" content="summary">
<meta name="twitter:site" content="@cambridge-ceu">
<meta name="twitter:url" content="/applications/pca_projection.html">
<meta name="twitter:creator" content="@jekyll-rtd-theme v2.0.10">
<meta property="og:title" content="PCA projection · csd3">
<meta property="og:description" content="PCA projection Official page: https://github.com/covid19-hg/pca_projection. git clone https://github.com/covid19-hg/pca_projection cd pca_projectio...">
<meta property="og:locale" content="en">
<meta property="og:url" content="/applications/pca_projection.html">
<meta property="og:type" content="article">
<meta property="article:author" content="cambridge-ceu">
<meta property="article:published_time" content="2021-12-23T16:14:30+00:00">
<meta property="article:modified_time" content="2021-12-23T16:14:30+00:00">
<script type="application/ld+json">
    {
        "@context": "https://schema.org",
        "@type": "Article",
        "mainEntityOfPage": {
            "@type": "WebPage",
            "@id": "/applications/pca_projection.html"
        },
        "headline": "PCA projection · csd3",
        "image": [],
        "author": {
            "@type": "Person",
            "name": "cambridge-ceu"
        },
        "datePublished": "2021-12-23T16:14:30+00:00",
        "dateModified": "2021-12-23T16:14:30+00:00",
        "publisher": {
            "@type": "",
            "name": "cambridge-ceu",
            "logo": {
                "@type": "ImageObject",
                "url": ""
            }
        },
        "description": "PCA projection Official page: https://github.com/covid19-hg/pca_projection. git clone https://github.com/covid19-hg/pca_projection cd pca_projectio..."
    }
</script>

<link rel="dns-prefetch" href="https://cdn.jsdelivr.net"><link rel="prev" href="/applications/pandoc.html"><link rel="next" href="/applications/plink.html"><link rel="canonical" href="/applications/pca_projection.html"><link rel="icon" type="image/svg+xml" href="/assets/images/favicon.svg"><link rel="icon" type="image/png" href="/assets/images/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="/assets/images/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="/assets/images/favicon-96x96.png" sizes="96x96"><link rel="mask-icon" href="/assets/images/favicon.svg" color="#2980b9"><link rel="apple-touch-icon" href="/assets/images/apple-touch-icon-300x300.jpg">
<link rel="stylesheet" href="/assets/css/theme.css"><script>
    window.ui = {
        title: "csd3",
        baseurl: "",
        i18n: {
            search_results: "Search Results",
            search_results_found: "Search finished, found # page(s) matching the search query.",
            search_results_not_found: "Your search did not match any documents, please make sure that all characters are spelled correctly!"
        }
    };
</script>
</head>
    <body class="container"><div class="sidebar-wrap overflow-hidden">
    <div class="sidebar height-full overflow-y-scroll overflow-x-hidden">
        <div class="header d-flex flex-column p-3 text-center">
            <div class="title pb-1">
                <a class="h4 no-underline py-1 px-2 rounded-1" href="/" title="A collection of CSD3 notes">
                    <i class="fa fa-home"></i> csd3
                </a>
            </div>
            <span class="version"></span>
            <form class="search pt-2" action="/search.html" method="get" autocomplete="off">
                <input class="form-control input-block input-sm" type="text" name="q" placeholder="Search docs...">
            </form>
        </div>
        <div class="toctree py-2" data-spy="affix" role="navigation" aria-label="main navigation">
            


<ul>
</ul>




    
    <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/systems/">
        The system
    </a><ul>
    

        <li class="toc level-1 " data-sort="1" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/system.html">1. System</a>
        </li>

        <li class="toc level-1 " data-sort="2" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/login.html">2. Login</a>
        </li>

        <li class="toc level-1 " data-sort="3" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/directories.html">3. Directories</a>
        </li>

        <li class="toc level-1 " data-sort="4" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/policies.html">4. Policies</a>
        </li>

        <li class="toc level-1 " data-sort="5" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/email.html">5. email</a>
        </li>

        <li class="toc level-1 " data-sort="6" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/software.html">6. Software</a>
        </li>

        <li class="toc level-1 " data-sort="7" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/training.html">7. Training</a>
        </li>

        <li class="toc level-1 " data-sort="8" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/acknowledgement.html">8. Acknowledgement</a>
        </li>

        <li class="toc level-1 " data-sort="8" data-level="1">
            <a class="d-flex flex-items-baseline " href="/systems/contacts.html">8. Contacts</a>
        </li></ul>



    
    <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/cardio/">
        cardio
    </a><ul>
    </ul>



    
    <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/applications/">
        Applications
    </a><ul>
    

        <li class="toc level-1 " data-sort="1" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/ABCtoolbox.html">1. ABCtoolbox</a>
        </li>

        <li class="toc level-1 " data-sort="2" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/AEGIS.html">2. AEGIS</a>
        </li>

        <li class="toc level-1 " data-sort="3" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/akt.html">3. akt</a>
        </li>

        <li class="toc level-1 " data-sort="4" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/ANNOVAR.html">4. ANNOVAR</a>
        </li>

        <li class="toc level-1 " data-sort="5" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/aria2.html">5. aria2</a>
        </li>

        <li class="toc level-1 " data-sort="6" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/bcftools.html">6. bcftools</a>
        </li>

        <li class="toc level-1 " data-sort="7" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/bedops.html">7. bedops</a>
        </li>

        <li class="toc level-1 " data-sort="8" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/bgenix.html">8. bgenix</a>
        </li>

        <li class="toc level-1 " data-sort="9" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/CaVEMaN.html">9. CaVEMaN</a>
        </li>

        <li class="toc level-1 " data-sort="10" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/citeproc.html">10. citeproc</a>
        </li>

        <li class="toc level-1 " data-sort="11" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/Cytoscape.html">11. Cytoscape</a>
        </li>

        <li class="toc level-1 " data-sort="12" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/DosageConverter.html">12. DosageConverter</a>
        </li>

        <li class="toc level-1 " data-sort="13" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/gcta.html">13. GCTA</a>
        </li>

        <li class="toc level-1 " data-sort="14" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/gsutil.html">14. gsutil</a>
        </li>

        <li class="toc level-1 " data-sort="15" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/gwas2vcf.html">15. gwas2vcf</a>
        </li>

        <li class="toc level-1 " data-sort="16" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/hail.html">16. hail</a>
        </li>

        <li class="toc level-1 " data-sort="17" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/HESS.html">17. HESS</a>
        </li>

        <li class="toc level-1 " data-sort="18" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/KentUtils.html">18. KentUtils</a>
        </li>

        <li class="toc level-1 " data-sort="19" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/lapack.html">19. lapack</a>
        </li>

        <li class="toc level-1 " data-sort="20" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/ldsc.html">20. ldsc</a>
        </li>

        <li class="toc level-1 " data-sort="21" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/LEMMA.html">21. LEMMA</a>
        </li>

        <li class="toc level-1 " data-sort="22" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/MetaXcan.html">22. MetaXcan</a>
        </li>

        <li class="toc level-1 " data-sort="23" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/pandoc.html">23. pandoc</a>
        </li>

        <li class="toc level-1 current" data-sort="24" data-level="1">
            <a class="d-flex flex-items-baseline current" href="/applications/pca_projection.html">24. PCA projection</a>
        </li>

        <li class="toc level-1 " data-sort="25" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/plink.html">25. PLINK</a>
        </li>

        <li class="toc level-1 " data-sort="26" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/polyphen-2.html">26. polyphen-2</a>
        </li>

        <li class="toc level-1 " data-sort="27" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/poppler.html">27. poppler</a>
        </li>

        <li class="toc level-1 " data-sort="28" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/PRSice.html">28. PRSice</a>
        </li>

        <li class="toc level-1 " data-sort="29" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/PRSoS.html">29. PRSoS</a>
        </li>

        <li class="toc level-1 " data-sort="30" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/pspp.html">30. pspp</a>
        </li>

        <li class="toc level-1 " data-sort="31" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/qpdf.html">31. qpdf</a>
        </li>

        <li class="toc level-1 " data-sort="32" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/regenie.html">32. regenie</a>
        </li>

        <li class="toc level-1 " data-sort="33" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/R.html">33. R</a>
        </li>

        <li class="toc level-1 " data-sort="34" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/RStudio.html">34. RStudio</a>
        </li>

        <li class="toc level-1 " data-sort="35" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/shapeit3.html">35. shapeit3</a>
        </li>

        <li class="toc level-1 " data-sort="36" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/SMR.html">36. SMR</a>
        </li>

        <li class="toc level-1 " data-sort="37" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/snakemake.html">37. snakemake</a>
        </li>

        <li class="toc level-1 " data-sort="38" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/SNPTEST.html">38. SNPTEST</a>
        </li>

        <li class="toc level-1 " data-sort="39" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/tabix.html">39. tabix</a>
        </li>

        <li class="toc level-1 " data-sort="40" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/VEP.html">40. VEP</a>
        </li>

        <li class="toc level-1 " data-sort="41" data-level="1">
            <a class="d-flex flex-items-baseline " href="/applications/VScode.html">41. Visual Studio code</a>
        </li>
        <li class="toc level-1">


    <a class="d-flex flex-items-baseline" href="/applications/R/">
        R packages
    </a><ul>
    

        <li class="toc level-2 " data-sort="1" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/brms.html">1. R/brms</a>
        </li>

        <li class="toc level-2 " data-sort="2" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/EasyQC.html">2. R/EasyQC</a>
        </li>

        <li class="toc level-2 " data-sort="3" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/gnn.html">3. R/gnn</a>
        </li>

        <li class="toc level-2 " data-sort="4" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/gsl.html">4. R/gsl</a>
        </li>

        <li class="toc level-2 " data-sort="5" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/LDlinkR.html">5. R/LDlinkR</a>
        </li>

        <li class="toc level-2 " data-sort="6" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/magick.html">6. R/magick</a>
        </li>

        <li class="toc level-2 " data-sort="7" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/PhenoScanner.html">7. R/PhenoScanner</a>
        </li>

        <li class="toc level-2 " data-sort="8" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/PKI.html">8. R/PKI</a>
        </li>

        <li class="toc level-2 " data-sort="9" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/plumber.html">9. R/plumber</a>
        </li>

        <li class="toc level-2 " data-sort="10" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/protolite.html">10. R/protolite</a>
        </li>

        <li class="toc level-2 " data-sort="11" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/Rfast.html">11. R/Rfast</a>
        </li>

        <li class="toc level-2 " data-sort="12" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/rgdal.html">12. R/rgdal</a>
        </li>

        <li class="toc level-2 " data-sort="13" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/rgeos.html">13. R/rgeos</a>
        </li>

        <li class="toc level-2 " data-sort="14" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/Rhdf5lib.html">14. R/Rhdf5lib</a>
        </li>

        <li class="toc level-2 " data-sort="15" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/rjags.html">15. R/rjags</a>
        </li>

        <li class="toc level-2 " data-sort="16" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/rJava.html">16. R/rJava</a>
        </li>

        <li class="toc level-2 " data-sort="17" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/rstan.html">17. R/rstan</a>
        </li>

        <li class="toc level-2 " data-sort="18" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/SAIGE.html">18. R/SAIGE 0.36.6 and 0.39.2</a>
        </li>

        <li class="toc level-2 " data-sort="19" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/sf.html">19. R/sf</a>
        </li>

        <li class="toc level-2 " data-sort="20" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/snpnet.html">20. R/snpnet</a>
        </li>

        <li class="toc level-2 " data-sort="21" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/sojo.html">21. R/sojo</a>
        </li>

        <li class="toc level-2 " data-sort="22" data-level="2">
            <a class="d-flex flex-items-baseline " href="/applications/R/xlsx.html">22. R/xlsx</a>
        </li></ul>
</li></ul>



    
    <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/writing/">
        Writing Related
    </a><ul>
    

        <li class="toc level-1 " data-sort="1" data-level="1">
            <a class="d-flex flex-items-baseline " href="/writing/front-matter.html">1. Front matter</a>
        </li>

        <li class="toc level-1 " data-sort="2" data-level="1">
            <a class="d-flex flex-items-baseline " href="/writing/list.html">2. List directory</a>
        </li>

        <li class="toc level-1 " data-sort="3" data-level="1">
            <a class="d-flex flex-items-baseline " href="/writing/mathjax.html">3. Mathjax</a>
        </li>

        <li class="toc level-1 " data-sort="4" data-level="1">
            <a class="d-flex flex-items-baseline " href="/writing/admonitions.html">4. Admonition card</a>
        </li>
        <li class="toc level-1">


    <a class="d-flex flex-items-baseline" href="/writing/test/">
        Test Documentation
    </a><ul>
    

        <li class="toc level-2 " data-sort="1" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/markdown.html">1. Markdown Elements</a>
        </li>

        <li class="toc level-2 " data-sort="2" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/toasts.html">2. Toasts Card</a>
        </li>

        <li class="toc level-2 " data-sort="3" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/codes.html">3. Code Blocks</a>
        </li>

        <li class="toc level-2 " data-sort="4" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/mermaid.html">4. Mermaid Test</a>
        </li>

        <li class="toc level-2 " data-sort="5" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/emoji.html">5. Emoji Test</a>
        </li>

        <li class="toc level-2 " data-sort="6" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/gist.html">6. Gist Test</a>
        </li>

        <li class="toc level-2 " data-sort="7" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/avatar.html">7. Avatar Test</a>
        </li>

        <li class="toc level-2 " data-sort="8" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/mentions.html">8. Mentions Test</a>
        </li>

        <li class="toc level-2 " data-sort="9" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/fonts.html">9. Fonts Test</a>
        </li>

        <li class="toc level-2 " data-sort="10" data-level="2">
            <a class="d-flex flex-items-baseline " href="/writing/test/utilities.html">10. Primer Utilities Test</a>
        </li></ul>
</li></ul>

        </div>
    </div>
</div>
<div class="content-wrap">
    <div class="header d-flex flex-justify-between p-2 hide-lg hide-xl" aria-label="top navigation">
        <button id="toggle" aria-label="Toggle menu" class="btn-octicon p-2 m-0 text-white" type="button">
            <i class="fa fa-bars"></i>
        </button>
        <div class="title flex-1 d-flex flex-justify-center">
            <a class="h4 no-underline py-1 px-2 rounded-1" href="/">csd3</a>
        </div>
    </div>
    <div class="content p-3 p-sm-5">
        <div class="navigation-top d-flex flex-justify-between">
    <ul class="breadcrumb" role="navigation" aria-label="breadcrumbs navigation">
        <li class="breadcrumb-item">
            <a class="no-underline" href="/" title="/">
                <i class="fa fa-home"></i>
            </a>
        </li><li class="breadcrumb-item" ><a href="/applications/">applications</a></li><li class="breadcrumb-item" aria-current="page">pca_projection.md</li></ul>
    <a class="edit" href="https://github.com/cambridge-ceu/csd3/edit/gh-pages/applications/pca_projection.md" title="Edit on GitHub" rel="noreferrer" target="_blank">
        <i class="fa fa-edit"></i>
    </a>
</div>

        <hr>
        <div role="main" itemscope="itemscope" itemtype="https://schema.org/Article">
            <div class="markdown-body" itemprop="articleBody">
                <h1 id="pca-projection">PCA projection</h1>

<p>Official page: <a href="https://github.com/covid19-hg/pca_projection">https://github.com/covid19-hg/pca_projection</a>.</p>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>git clone https://github.com/covid19-hg/pca_projection
<span class="nb">cd </span>pca_projection
</code></pre> </div></div>

<h2 id="setup">setup</h2>

<p>The required steps according to the documentation can be summarised as follows,</p>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="c"># Pre-computed PCA loadings, reference allele frequencies and scores</span>
wget https://storage.googleapis.com/covid19-hg-public/pca_projection/hgdp_tgp_pca_covid19hgi_snps_loadings.GRCh37.plink.tsv
wget https://storage.googleapis.com/covid19-hg-public/pca_projection/hgdp_tgp_pca_covid19hgi_snps_loadings.GRCh37.plink.afreq
wget https://storage.googleapis.com/covid19-hg-public/pca_projection/hgdp_tgp_pca_covid19hgi_snps_loadings.GRCh38.plink.tsv
wget https://storage.googleapis.com/covid19-hg-public/pca_projection/hgdp_tgp_pca_covid19hgi_snps_loadings.GRCh38.plink.afreq
wget https://storage.googleapis.com/covid19-hg-public/pca_projection/hgdp_tgp_pca_covid19hgi_snps_loadings.rsid.plink.tsv
wget https://storage.googleapis.com/covid19-hg-public/pca_projection/hgdp_tgp_pca_covid19hgi_snps_loadings.rsid.plink.afreq
wget https://storage.googleapis.com/covid19-hg-public/pca_projection/hgdp_tgp_pca_covid19hgi_snps_scores.txt.gz

<span class="c"># imputed PLINK2 dosage files</span>

<span class="nb">cut</span> <span class="nt">-f1</span> <span class="o">[</span>path to the pre-computed loadings file] | <span class="nb">tail</span> <span class="nt">-n</span> +2 <span class="o">&gt;</span> variants.extract

<span class="c">## .pgen/.pvar/.psam</span>
plink2 <span class="se">\</span>
  <span class="nt">--pfile</span> <span class="o">[</span>path to your per-chromosome pfile] <span class="se">\</span>
  <span class="nt">--extract</span> variants.extract <span class="se">\</span>
  <span class="nt">--make-pfile</span> <span class="se">\</span>
  <span class="nt">--out</span> <span class="o">[</span>per-chromosome output name]
<span class="nb">ls</span> <span class="o">[</span>the previous per-chromosome output prefix].<span class="k">*</span>.pgen | <span class="nb">sed</span> <span class="nt">-e</span> ‘s/.pgen//' <span class="o">&gt;</span> merge-list.txt
plink2 <span class="nt">--pmerge-list</span> merge-list.txt <span class="nt">--out</span> <span class="o">[</span>all-chromosome output name]

<span class="c">## .bed/.bim/.fam</span>

plink <span class="se">\</span>
  <span class="nt">--bfile</span> <span class="o">[</span>path to your per-chromosome pfile] <span class="se">\</span>
  <span class="nt">--extract</span> variants.extract <span class="se">\</span>
  <span class="nt">--make-bed</span> <span class="se">\</span>
  <span class="nt">--out</span> <span class="o">[</span>per-chromosome output name]
<span class="nb">ls</span> <span class="o">[</span>outname].<span class="k">*</span>.bed | <span class="nb">sed</span> <span class="nt">-e</span> ‘s/.bed//' <span class="o">&gt;</span> merge-list.txt
plink <span class="nt">--merge-list</span> merge-list.txt <span class="nt">--out</span> <span class="o">[</span>all-chromosome output name]

<span class="c">## .bgen/.sample (in doubt)</span>

bgenix <span class="se">\</span>
  <span class="nt">-g</span> <span class="o">[</span>path to your per-chromosome bgen] <span class="se">\</span>
  <span class="nt">-incl-rsids</span> variant.extract <span class="se">\</span>
  <span class="o">&gt;</span> <span class="o">[</span>per-chromosome output name].bgen
cat-bgen <span class="se">\</span>
<span class="nt">-g</span> <span class="o">[</span>path to your per-chromosome bgen 1] <span class="se">\</span>
   <span class="o">[</span>path to your per-chromosome bgen 2] <span class="se">\</span>
...
   <span class="o">[</span>path to your per-chromosome bgen 22] <span class="se">\</span>
<span class="nt">-og</span> <span class="o">[</span>all-chromosome outname]
plink2 <span class="se">\</span>
  <span class="nt">--bgen</span> <span class="o">[</span>path to all-chromosome bgen] <span class="o">[</span>REF/ALT mode] <span class="se">\</span>
  <span class="nt">--make-pfile</span> <span class="se">\</span>
  <span class="nt">--out</span> <span class="o">[</span>output pfile name]

<span class="c">## .vcf</span>

bcftools view <span class="nt">-Oz</span> <span class="se">\</span>
  <span class="nt">-i</span> "ID <span class="o">=</span> @variants.extract" <span class="se">\</span>
  <span class="o">[</span>path to your per-chromosome vcf file] <span class="se">\</span>
  <span class="o">&gt;</span> <span class="o">[</span>per-chromosome outname&gt;.vcf.gz]
bcftools concat <span class="nt">-Oz</span> <span class="o">[</span>per-chromosome vcf files] <span class="o">&gt;</span> <span class="o">[</span>all-chromosome outname].vcf.gz
plink2 <span class="se">\</span>
  <span class="nt">--vcf</span> <span class="o">[</span>all-chromosome outname].vcf.gz <span class="se">\</span>
  <span class="nv">dosage</span><span class="o">=[</span>dosage field name] <span class="se">\</span>
  <span class="nt">--make-pfile</span> <span class="se">\</span>
  <span class="nt">--out</span> <span class="o">[</span>outname]

<span class="c"># project and plot PC</span>

plink2 <span class="se">\</span>
  <span class="k">${</span><span class="nv">input_command</span><span class="k">}</span> <span class="se">\</span>
  <span class="nt">--score</span> <span class="k">${</span><span class="nv">PCA_LOADINGS</span><span class="k">}</span> <span class="se">\</span>
  center <span class="se">\</span>
  <span class="nv">cols</span><span class="o">=</span><span class="nt">-scoreavgs</span>,+scoresums <span class="se">\</span>
  list-variants <span class="se">\</span>
  header-read <span class="se">\</span>
  <span class="nt">--score-col-nums</span> 3-22 <span class="se">\</span>
  <span class="nt">--read-freq</span> <span class="k">${</span><span class="nv">PCA_AF</span><span class="k">}</span> <span class="se">\</span>
  <span class="nt">--out</span> <span class="k">${</span><span class="nv">OUTNAME</span><span class="k">}</span>

Rscript <span class="nt">-e</span> <span class="s2">"install.packages(c("</span>data.table<span class="s2">", "</span>hexbin<span class="s2">", "</span>optparse<span class="s2">", "</span>patchwork<span class="s2">", "</span>R.utils<span class="s2">", "</span>tidyverse<span class="s2">"))"</span>
Rscript plot_projected_pc.R <span class="se">\</span>
  <span class="nt">--sscore</span> <span class="o">[</span>path to .sscore output] <span class="se">\</span>
  <span class="nt">--phenotype-file</span> <span class="o">[</span>path to phenotype file] <span class="se">\</span>
  <span class="nt">--phenotype-col</span> <span class="o">[</span>phenotype column name]
  <span class="nt">--covariate-file</span> <span class="o">[</span>path to covariate file] <span class="se">\</span>
  <span class="nt">--pc-prefix</span> <span class="o">[</span>prefix of PC columns] <span class="se">\</span>
  <span class="nt">--pc-num</span> <span class="o">[</span>number of PCs used <span class="k">in </span>GWAS] <span class="se">\</span>
  <span class="nt">--ancestry</span> <span class="o">[</span>ancestry code: AFR, AMR, EAS, EUR, MID, or SAS] <span class="se">\</span>
  <span class="nt">--study</span> <span class="o">[</span>your study name] <span class="se">\</span>
  <span class="nt">--out</span> <span class="o">[</span>output name prefix]

<span class="c"># --ancestry-file [path to ancestry file] \</span>
<span class="c"># --ancestry-col [ancestry column name]</span>
<span class="c"># --reference-score-file</span>
</code></pre> </div></div>

<h2 id="implementation">Implementation</h2>

<p>This is detailed below by section.</p>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="c">#!/usr/bin/bash</span>

<span class="nb">export </span><span class="nv">TMPDIR</span><span class="o">=</span><span class="k">${</span><span class="nv">HPC_WORK</span><span class="k">}</span>/work
<span class="c"># genotype</span>
<span class="nb">export </span><span class="nv">autosomes</span><span class="o">=</span>~/rds/post_qc_data/interval/imputed/uk10k_1000g_b37/
<span class="c"># location of PCs</span>
<span class="nb">export </span><span class="nv">ref</span><span class="o">=</span>~/rds/post_qc_data/interval/reference_files/genetic/reference_files_genotyped_imputed/
<span class="c"># HGI working directory</span>
<span class="nb">export </span><span class="nv">prefix</span><span class="o">=</span>~/rds/rds-asb38-ceu-restricted/projects/covid/HGI
<span class="nb">export dir</span><span class="o">=</span>20201201-ANA_C2_V2

<span class="nb">export </span><span class="nv">PCA_projection</span><span class="o">=</span>pca_projection
<span class="nb">export </span><span class="nv">PCA_loadings</span><span class="o">=</span>hgdp_tgp_pca_covid19hgi_snps_loadings.GRCh37.plink.tsv
<span class="nb">export </span><span class="nv">PCA_af</span><span class="o">=</span>hgdp_tgp_pca_covid19hgi_snps_loadings.GRCh37.plink.afreq
<span class="nb">export </span><span class="nv">sscore</span><span class="o">=</span>hgdp_tgp_pca_covid19hgi_snps_scores.txt.gz

module load plink/2.00-alpha ceuadmin/stata

<span class="k">function </span>extract_data<span class="o">()</span>
<span class="o">{</span>
  <span class="nb">cut</span> <span class="nt">-f1</span> <span class="k">${</span><span class="nv">PCA_projection</span><span class="k">}</span>/<span class="k">${</span><span class="nv">PCA_loadings</span><span class="k">}</span> | <span class="nb">tail</span> <span class="nt">-n</span> +2 <span class="o">&gt;</span> variants.extract
  <span class="o">(</span>
    <span class="nb">cat </span>variants.extract
    <span class="nb">awk</span> <span class="s1">'{split($1,a,":");print "chr"a[1]":"a[2]"_"a[4]"_"a[3]}'</span> variants.extract
  <span class="o">)</span> <span class="o">&gt;</span> variants.extract2

  <span class="nb">cp</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/<span class="k">${</span><span class="nv">dir</span><span class="k">}</span>/output/INTERVAL-<span class="k">*</span>.bgen.bgi <span class="k">${</span><span class="nv">TMPDIR</span><span class="k">}</span>
  <span class="nb">seq </span>22 | <span class="se">\</span>
  parallel <span class="nt">-C</span><span class="s1">' '</span> <span class="nt">--env</span> prefix <span class="nt">--env</span> <span class="nb">dir</span> <span class="s1">'
    ln -sf ${prefix}/${dir}/output/INTERVAL-{}.bgen ${TMPDIR}/INTERVAL-{}.bgen
    python update_bgi.py --bgi ${TMPDIR}/INTERVAL-{}.bgen.bgi
    bgenix -g ${TMPDIR}/INTERVAL-{}.bgen -incl-rsids variants.extract2 &gt; ${prefix}/work/INTERVAL-{}.bgen
  '</span>
<span class="o">}</span>

<span class="k">function </span>twist<span class="o">()</span>
<span class="o">{</span>
  cat-bgen <span class="nt">-g</span> <span class="si">$(</span><span class="nb">echo</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/INTERVAL-<span class="o">{</span>1..22<span class="o">}</span>.bgen<span class="si">)</span> <span class="nt">-og</span> INTERVAL.bgen <span class="nt">-clobber</span>
  bgenix <span class="nt">-g</span> INTERVAL.bgen <span class="nt">-index</span> <span class="nt">-clobber</span>
  <span class="nb">export </span><span class="nv">csvfile</span><span class="o">=</span>INTERVAL.csv
  python update_bgi.py <span class="nt">--bgi</span> INTERVAL.bgen.bgi
  plink2 <span class="nt">--bgen</span> INTERVAL.bgen ref-first <span class="nt">--make-pfile</span> <span class="nt">--out</span> INTERVAL
  <span class="nb">cp </span>INTERVAL.p??? <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work
  <span class="o">(</span>
    <span class="nb">head</span> <span class="nt">-1</span> INTERVAL.pvar
    <span class="nb">paste</span> &lt;<span class="o">(</span><span class="nb">sed</span> <span class="s1">'1d'</span> INTERVAL.pvar | <span class="nb">cut</span> <span class="nt">-f1</span>,2<span class="o">)</span> <span class="se">\</span>
          &lt;<span class="o">(</span><span class="nb">sed</span> <span class="s1">'1d'</span> INTERVAL.csv | <span class="nb">cut</span> <span class="nt">-d</span>, <span class="nt">-f3</span><span class="o">)</span> | <span class="se">\</span>
    <span class="nb">paste</span> - &lt;<span class="o">(</span><span class="nb">sed</span> <span class="s1">'1d'</span> INTERVAL.pvar | <span class="nb">cut</span> <span class="nt">-f4</span>,5<span class="o">)</span>
  <span class="o">)</span> <span class="o">&gt;</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/INTERVAL.pvar
<span class="o">}</span>

<span class="k">function </span>project_pc<span class="o">()</span>
<span class="o">{</span>
<span class="c">#!/bin/bash</span>

  <span class="nb">set</span> <span class="nt">-eu</span>
<span class="c">################################################################################</span>
<span class="c"># Please fill in the below variables</span>
<span class="c">################################################################################</span>
<span class="c"># Metadata</span>
  <span class="nv">STUDY_NAME</span><span class="o">=</span><span class="s2">"INTERVAL"</span>
  <span class="nv">ANALYST_LAST_NAME</span><span class="o">=</span><span class="s2">"ZHAO"</span>
  <span class="nv">DATE</span><span class="o">=</span><span class="s2">"</span><span class="si">$(</span><span class="nb">date</span> +<span class="s1">'%Y%m%d'</span><span class="si">)</span><span class="s2">"</span>
  <span class="nv">OUTNAME</span><span class="o">=</span><span class="s2">"</span><span class="k">${</span><span class="nv">prefix</span><span class="k">}</span><span class="s2">/work/</span><span class="k">${</span><span class="nv">STUDY_NAME</span><span class="k">}</span><span class="s2">.</span><span class="k">${</span><span class="nv">ANALYST_LAST_NAME</span><span class="k">}</span><span class="s2">.</span><span class="k">${</span><span class="nv">DATE</span><span class="k">}</span><span class="s2">"</span>
<span class="c">################################################################################</span>
<span class="c"># Location of downloaded input files</span>
  <span class="nv">PCA_LOADINGS</span><span class="o">=</span><span class="s2">"</span><span class="k">${</span><span class="nv">PCA_projection</span><span class="k">}</span><span class="s2">/</span><span class="k">${</span><span class="nv">PCA_loadings</span><span class="k">}</span><span class="s2">"</span>
  <span class="nv">PCA_AF</span><span class="o">=</span><span class="s2">"</span><span class="k">${</span><span class="nv">PCA_projection</span><span class="k">}</span><span class="s2">/</span><span class="k">${</span><span class="nv">PCA_af</span><span class="k">}</span><span class="s2">"</span>
<span class="c">################################################################################</span>
<span class="c"># Location of imputed genotype files</span>
<span class="c"># [Recommended]</span>
<span class="c"># PLINK 2 binary format: a prefix (with directories) of .pgen/.pvar/.psam files</span>
  <span class="nv">PFILE</span><span class="o">=</span><span class="s2">"</span><span class="k">${</span><span class="nv">prefix</span><span class="k">}</span><span class="s2">/work/INTERVAL"</span>
<span class="c"># [Acceptable]</span>
<span class="c"># PLINK 1 binary format: a prefix of .bed/.bim/.fam files</span>
  <span class="nv">BFILE</span><span class="o">=</span><span class="s2">""</span>
<span class="c">################################################################################</span>

  <span class="k">function </span>error_exit<span class="o">()</span> <span class="o">{</span>
    <span class="nb">echo</span> <span class="s2">"</span><span class="k">${</span><span class="nv">1</span><span class="k">:-</span><span class="s2">"Unknown Error"</span><span class="k">}</span><span class="s2">"</span> 1&gt;&amp;2
    <span class="nb">exit </span>1
  <span class="o">}</span>

<span class="c"># Input checks</span>
  <span class="k">if</span> <span class="o">[[</span> <span class="nt">-z</span> <span class="s2">"</span><span class="k">${</span><span class="nv">STUDY_NAME</span><span class="k">}</span><span class="s2">"</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then
    </span>error_exit <span class="s2">"Please specify </span><span class="se">\$</span><span class="s2">STUDY_NAME."</span>
  <span class="k">fi

  if</span> <span class="o">[[</span> <span class="nt">-z</span> <span class="s2">"</span><span class="k">${</span><span class="nv">ANALYST_LAST_NAME</span><span class="k">}</span><span class="s2">"</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then
    </span>error_exit <span class="s2">"Please specify </span><span class="se">\$</span><span class="s2">ANALYST_LAST_NAME."</span>
  <span class="k">fi

  if</span> <span class="o">[[</span> <span class="nt">-z</span> <span class="s2">"</span><span class="k">${</span><span class="nv">PCA_LOADINGS</span><span class="k">}</span><span class="s2">"</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then
    </span>error_exit <span class="s2">"Please specify </span><span class="se">\$</span><span class="s2">PCA_LOADINGS."</span>
  <span class="k">fi

  if</span> <span class="o">[[</span> <span class="nt">-z</span> <span class="s2">"</span><span class="k">${</span><span class="nv">PCA_AF</span><span class="k">}</span><span class="s2">"</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then
    </span>error_exit <span class="s2">"Please specify </span><span class="se">\$</span><span class="s2">PCA_AF."</span>
  <span class="k">fi

  if</span> <span class="o">[[</span> <span class="nt">-n</span> <span class="s2">"</span><span class="k">${</span><span class="nv">PFILE</span><span class="k">}</span><span class="s2">"</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then
    </span><span class="nv">input_command</span><span class="o">=</span><span class="s2">"--pfile </span><span class="k">${</span><span class="nv">PFILE</span><span class="k">}</span><span class="s2">"</span>
  <span class="k">elif</span> <span class="o">[[</span> <span class="nt">-n</span> <span class="s2">"</span><span class="k">${</span><span class="nv">BFILE</span><span class="k">}</span><span class="s2">"</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then
    </span><span class="nv">input_command</span><span class="o">=</span><span class="s2">"--bfile </span><span class="k">${</span><span class="nv">BFILE</span><span class="k">}</span><span class="s2">"</span>
  <span class="k">else
    </span>error_exit <span class="s2">"Either </span><span class="se">\$</span><span class="s2">PFILE or </span><span class="se">\$</span><span class="s2">BFILE should be specified"</span>
  <span class="k">fi</span>

<span class="c"># Run plink2 --score</span>
  plink2 <span class="se">\</span>
    <span class="k">${</span><span class="nv">input_command</span><span class="k">}</span> <span class="se">\</span>
    <span class="nt">--score</span> <span class="k">${</span><span class="nv">PCA_LOADINGS</span><span class="k">}</span> <span class="se">\</span>
            variance-standardize <span class="se">\</span>
            <span class="nv">cols</span><span class="o">=</span><span class="nt">-scoreavgs</span>,+scoresums <span class="se">\</span>
            list-variants <span class="se">\</span>
            header-read <span class="se">\</span>
    <span class="nt">--score-col-nums</span> 3-12 <span class="se">\</span>
    <span class="nt">--read-freq</span> <span class="k">${</span><span class="nv">PCA_AF</span><span class="k">}</span> <span class="se">\</span>
    <span class="nt">--out</span> <span class="k">${</span><span class="nv">OUTNAME</span><span class="k">}</span>

<span class="c"># The score file does not have FID (=0)</span>
  <span class="nb">awk</span> <span class="s1">'
  {
    if (NR==1) $1="#FID IID"
    else $1=0" "$1
    print
  }'</span> <span class="k">${</span><span class="nv">OUTNAME</span><span class="k">}</span>.sscore <span class="o">&gt;</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/snpid.sscore
  <span class="nb">ln</span> <span class="nt">-sf</span> <span class="k">${</span><span class="nv">OUTNAME</span><span class="k">}</span>.sscore.vars <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/snpid.sscore.vars

<span class="c"># Our PCs are PC_# rather than PC#</span>
<span class="c"># The phenotype and covariate file missed FID (=0) column</span>
  <span class="nb">awk</span> <span class="s1">'
  {
    if (NR==1)
    {
      $1="FID I"$1
      gsub(/PC_/,"PC",$0)
    }
    else $1=0" " $1
  };1'</span> <span class="k">${</span><span class="nv">dir</span><span class="k">}</span>/work/INTERVAL-covid.txt | <span class="se">\</span>
  <span class="nb">tr</span> <span class="s1">' '</span> <span class="s1">'\t'</span> <span class="o">&gt;</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/snpid.txt
  <span class="nb">cut</span> <span class="nt">-f1-3</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/snpid.txt <span class="o">&gt;</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/snpid.pheno
  <span class="nb">cut</span> <span class="nt">-f3</span> <span class="nt">--complement</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/snpid.txt <span class="o">&gt;</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/snpid.covars

  stata <span class="nt">-b</span> <span class="k">do </span>ethnic.do

  Rscript <span class="k">${</span><span class="nv">PCA_projection</span><span class="k">}</span>/plot_projected_pc.R <span class="se">\</span>
        <span class="nt">--sscore</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/snpid.sscore <span class="se">\</span>
        <span class="nt">--phenotype-file</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/snpid.pheno <span class="se">\</span>
        <span class="nt">--phenotype-col</span> SARS_CoV <span class="se">\</span>
        <span class="nt">--covariate-file</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>/work/snpid.covars <span class="se">\</span>
        <span class="nt">--pc-prefix</span> PC <span class="se">\</span>
        <span class="nt">--pc-num</span> 20 <span class="se">\</span>
        <span class="nt">--ancestry-file</span> ethnic.txt <span class="se">\</span>
        <span class="nt">--ancestry-col</span> ethnic <span class="se">\</span>
        <span class="nt">--study</span> <span class="k">${</span><span class="nv">STUDY_NAME</span><span class="k">}</span> <span class="se">\</span>
        <span class="nt">--out</span> <span class="k">${</span><span class="nv">OUTNAME</span><span class="k">}</span>
<span class="o">}</span>

extract_data<span class="p">;</span>twist<span class="p">;</span>project_pc
</code></pre> </div></div>

<h3 id="update_bgipy">update_bgi.py</h3>

<p>The magic <code class="language-plaintext highlighter-rouge notranslate">update_bgi.py</code> is as follows</p>

<div class="language-python highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="kn">import</span> <span class="nn">sqlite3</span>

<span class="c1"># mapping chromosome string (incluing 01-09, 23, and X) and correct string
</span><span class="n">CHROM_MAPPING_STR</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[(</span><span class="s">'0'</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">)]</span> <span class="o">+</span>
                         <span class="p">[(</span><span class="s">'X'</span><span class="p">,</span> <span class="s">'X'</span><span class="p">)])</span>
<span class="n">csvfile</span><span class="o">=</span><span class="n">os</span><span class="p">.</span><span class="n">environ</span><span class="p">[</span><span class="s">'csvfile'</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="n">conn</span> <span class="o">=</span> <span class="n">sqlite3</span><span class="p">.</span><span class="n">connect</span><span class="p">(</span><span class="n">args</span><span class="p">.</span><span class="n">bgi</span><span class="p">)</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">conn</span><span class="p">.</span><span class="n">cursor</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_sql</span><span class="p">(</span><span class="s">"select * from Variant"</span><span class="p">,</span> <span class="n">conn</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="n">flush</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">df</span><span class="p">.</span><span class="n">rsid</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="nb">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s">"{}:{}:{}:{}"</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">CHROM_MAPPING_STR</span><span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">5</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="n">flush</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">df</span><span class="p">.</span><span class="n">to_sql</span><span class="p">(</span><span class="s">"Variant"</span><span class="p">,</span> <span class="n">conn</span><span class="p">,</span> <span class="n">if_exists</span><span class="o">=</span><span class="s">"replace"</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">csvfile</span> <span class="o">!=</span> <span class="s">''</span><span class="p">:</span>
       <span class="n">df</span><span class="p">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">csvfile</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
    <span class="n">conn</span><span class="p">.</span><span class="n">close</span><span class="p">()</span>

<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">'__main__'</span><span class="p">:</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="p">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
    <span class="n">parser</span><span class="p">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">'--bgi'</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="p">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="n">main</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
</code></pre> </div></div>

<h3 id="bgisql">bgi.sql</h3>

<p>One may wonder the information regarding <code class="language-plaintext highlighter-rouge notranslate">.bgen.bgi</code> at all, which can be viewed as follows,</p>

<div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>sqlite3 INTERVAL.bgen.bgi &lt; bgi.sql
</code></pre> </div></div>

<p>while <code class="language-plaintext highlighter-rouge notranslate">bgi.sql</code> has the following lines</p>

<pre><code class="language-sqlite3">.tables
.separator "\t"
.header on
.output metadata.txt
select * from metadata;
.output Variant.txt
select * from Variant;
</code></pre> 

<h3 id="ethnicdo">ethnic.do</h3>

<p>The Stata program generates files containing FID, IID and ethnicity</p>

<pre><code class="language-stata">gzuse work/INTERVAL-omics-covid.dta.gz
sort identifier
gzsave INTERVAL-omics-covid, replace
insheet using 20201201/INTERVALdata_01DEC2020.csv, case clear
sort identifier
gzmerge identifier using INTERVAL-omics-covid.dta.gz
tabulate ethnicPulse SARS_CoV, all
gen str20 ethnic=ethnicPulse
replace ethnic="EUR" if inlist(ethnicPulse,"Eng/W/Scot/NI/Brit","White Irish")==1
replace ethnic="EAS" if inlist(ethnicPulse,"Asian- Bangladeshi","Asian- Indian","Asian- Pakistani","Chinese")==1
replace ethnic="MID" if ethnicPulse=="Arab"
gen ethnic_NA=ethnic
replace ethnic_NA="NA" if inlist(ethnic,"EUR","EAS","MID")==0
tab ethnic SARS_CoV, all
tab ethnic_NA SARS_CoV, all
gen FID=0
rename ID IID
outsheet FID IID ethnic using ethnic.txt if IID!=., noquote replace
outsheet FID IID ethnic_NA using ethnic_NA.txt if IID!=., noquote replace
rm INTERVAL-omics-covid.dta.gz
</code></pre> 

            </div>
        </div>
        <div class="navigation-bottom d-flex flex-justify-between py-3" role="navigation" aria-label="footer navigation">
            <div class="prev"><a href="/applications/pandoc.html" class="btn" title="pandoc" accesskey="p" rel="prev">
                        <i class="fa fa-arrow-circle-left"></i> Previous
                    </a></div>
            <div class="next"><a href="/applications/plink.html" class="btn" title="PLINK" accesskey="n" rel="next">
                        Next <i class="fa fa-arrow-circle-right"></i>
                    </a></div>
        </div><hr>

<div class="copyright text-center text-gray" role="contentinfo">
    <i class="fa fa-copyright"></i>
    <span class="time">2021,</span>
    <a class="text-gray" href="https://github.com/cambridge-ceu" rel="noreferrer" target="_blank">cambridge-ceu</a>
    Revision
    <a class="text-gray" href="https://github.com/cambridge-ceu/csd3/commit/2109d8f596b4cd432c7a5937f50b592827e6194f" title="2109d8f596b4cd432c7a5937f50b592827e6194f" rel="noreferrer" target="_blank">2109d8f</a>
    
        <br>
        <div class="generator">
            Built with
            <a href="https://pages.github.com" rel="noreferrer" target="_blank" title="github-pages v209">GitHub Pages</a>
            using a
            <a href="https://github.com/rundocs/jekyll-rtd-theme" rel="noreferrer" target="_blank" title="jekyll-rtd-theme v2.0.10">theme</a>
            provided by
            <a href="https://rundocs.io" rel="noreferrer" target="_blank">RunDocs</a>.
        </div>
    
</div>

    </div>
</div>
<div class="addons-wrap d-flex flex-column overflow-y-auto">
    <div class="status d-flex flex-justify-between p-2">
        <div class="title p-1">
            <i class="fa fa-book"></i> csd3
        </div>
        <div class="branch p-1">
            <span class="name">
                gh-pages
            </span>
            <i class="fa fa-caret-down"></i>
        </div>
    </div>
    <div class="addons d-flex flex-column height-full p-2 d-none">
        <dl>
            <dt>GitHub</dt>
            <dd>
                <a href="https://github.com/cambridge-ceu/csd3" title="Stars: ">
                    <i class="fa fa-github"></i>
                    Homepage
                </a>
            </dd>
            <dd>
                <a href="" title="Open issues: ">
                    <i class="fa fa-question-circle-o"></i>
                    Issues
                </a>
            </dd>
            <dd>
                <a href="https://github.com/cambridge-ceu/csd3/zipball/gh-pages" title="Size:  Kb">
                    <i class="fa fa-download"></i>
                    Download
                </a>
            </dd>
        </dl>
        <hr>
        <div class="license f6 pb-2">
            This
            <a href="/" title="csd3">Software</a>
            is under the terms of
            <a href="https://github.com/cambridge-ceu/csd3">The Unlicense</a>.
        </div>
    </div>
</div>
<script src="/assets/js/jquery.min.js"></script><script src="/assets/js/theme.js"></script>


</body>
</html>
