
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Usage notes &#8212; CSD3 0.1 documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="usage-notes">
<h1>Usage notes<a class="headerlink" href="#usage-notes" title="Permalink to this headline">¶</a></h1>
<p>This document contains information for the following software:</p>
<p>akt,
ABCtoolbox,
ANNOVAR,
AEGIS,
CaVEMaN,
bedops,
burden_testing,
R/brms,
DosageConverter,
HESS,
R,
R/gnn,
gsutil,
gwas2vcf,
KentUtils,
lapack,
R/LDlinkR,
MetaXcan,
R/PhenoScanner,
polyphen-2,
poppler,
PRSice,
PRSoS,
pspp,
qpdf,
R/rgdal,
R/rgeos,
R/Rhdf5lib,
R/rJava,
R/rjags,
R/rstan,
R/SAIGE,
smr,
R/snpnet,
VEP,
R/xlsx</p>
<p>Whenever appropriate, it is assumed that the destination of software installation is ${HPC_WORK}, e.g.,
via <code class="docutils literal notranslate"><span class="pre">export</span> <span class="pre">HPC_WORK=/rds/user/$USER/hpc-work</span></code> in .bashrc.</p>
<div class="section" id="akt">
<h2>akt<a class="headerlink" href="#akt" title="Permalink to this headline">¶</a></h2>
<p>Web: https://github.com/Illumina/akt</p>
<p>This is Illumina’s ancestry and kinship toolkit and the installation is canonical.</p>
</div>
<div class="section" id="abctoolbox">
<h2>ABCtoolbox<a class="headerlink" href="#abctoolbox" title="Permalink to this headline">¶</a></h2>
<p>Web page: https://bitbucket.org/wegmannlab/abctoolbox/wiki/Home</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git clone --depth <span class="m">1</span> https://bitbucket.org/wegmannlab/abctoolbox.git
<span class="c1"># g++ -O3 -o ABCtoolbox *.cpp</span>
module load openmpi/3.1.4-gcc-7.2.0
g++ -O3 -o ABCtoolbox *.cpp -DUSE_OMP -fopenmp
</pre></div>
</div>
<p>Note that the installation guide has misspecified the repository.</p>
</div>
<div class="section" id="annovar">
<h2>ANNOVAR<a class="headerlink" href="#annovar" title="Permalink to this headline">¶</a></h2>
<p>Web page: http://annovar.openbioinformatics.org/en/latest/</p>
<p>Registered from http://download.openbioinformatics.org/annovar_download_form.php with the following information,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">User</span> <span class="n">Name</span><span class="p">:</span> 
<span class="n">User</span> <span class="n">Email</span><span class="p">:</span> 
<span class="n">User</span> <span class="n">Institution</span><span class="p">:</span> <span class="n">University</span> <span class="n">of</span> <span class="n">Cambridge</span>
</pre></div>
</div>
<p>then copy the link received from email and issue commands from csd3,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="si">${</span><span class="nv">HPC_WORK</span><span class="si">}</span>
wget http://www.openbioinformatics.org/annovar/download/.../annovar.latest.tar.gz
tar xvfz annovar.latest.tar.gz
ls *pl <span class="p">|</span> sed <span class="s1">&#39;s/*//g&#39;</span> <span class="p">|</span> parallel -C<span class="s1">&#39; &#39;</span> <span class="s1">&#39;ln -sf ${HPC_WORK}/annovar/{} ${HPC_WORK}/bin/{}&#39;</span>
</pre></div>
</div>
<p>Additionally, one can download the ENSEMBL genes and whole-genome FASTA files to
humandb/hg19_seq for CCDS/GENCODE annotation.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> annovar
<span class="c1"># ENSEMBL genes</span>
annotate_variation.pl -buildver hg19 -downdb -webfrom annovar ensGene
annotate_variation.pl -build hg19 -out ex1 -dbtype ensGene example/ex1.avinput humandb/
<span class="c1"># reference genome in FASTA</span>
annotate_variation.pl -downdb -build hg19 seq humandb/hg19_seq/
<span class="c1"># CCDS genes</span>
annotate_variation.pl -downdb -build hg19 ccdsGene humandb
retrieve_seq_from_fasta.pl humandb/hg19_ccdsGene.txt -seqdir humandb/hg19_seq -format refGene <span class="se">\</span>
                           -outfile humandb/hg19_ccdsGeneMrna.fa
annotate_variation.pl -build hg19 -out ex1 -dbtype ccdsGene example/ex1.avinput humandb/
<span class="c1"># GENCODE genes</span>
annotate_variation.pl -downdb wgEncodeGencodeBasicV19 humandb/ -build hg19
retrieve_seq_from_fasta.pl -format genericGene -seqdir humandb/hg19_seq/ <span class="se">\</span>
                           -outfile humandb/hg19_wgEncodeGencodeBasicV19Mrna.fa humandb/hg19_wgEncodeGencodeBasicV19.txt
annotate_variation.pl -build hg19 -out ex1 -dbtype wgEncodeGencodeBasicV19 example/ex1.avinput humandb/
</pre></div>
</div>
<p>We can have region-based annotation as in http://annovar.openbioinformatics.org/en/latest/user-guide/region/,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Conserved genomic elements</span>
annotate_variation.pl -build hg19 -downdb phastConsElements46way humandb/
annotate_variation.pl -regionanno -build hg19 -out ex1 -dbtype phastConsElements46way example/ex1.avinput humandb/
annotate_variation.pl -regionanno -build hg19 -out ex1 -dbtype phastConsElements46way example/ex1.avinput humandb/ -normscore_threshold <span class="m">400</span>
<span class="c1"># Transcription factor binding sites</span>
annotate_variation.pl -build hg19 -downdb tfbsConsSites humandb/
annotate_variation.pl -regionanno -build hg19 -out ex1 -dbtype tfbsConsSites example/ex1.avinput humandb/
<span class="c1"># Cytogenetic band</span>
annotate_variation.pl -build hg19 -downdb cytoBand humandb/
annotate_variation.pl -regionanno -build hg19 -out ex1 -dbtype cytoBand example/ex1.avinput humandb/
<span class="c1"># microRNA, snoRNA</span>
annotate_variation.pl -build hg19 -downdb wgRna humandb/
annotate_variation.pl -regionanno -build hg19 -out ex1 -dbtype wgRna example/ex1.avinput humandb/
<span class="c1"># previously reported structural variants</span>
annotate_variation.pl -build hg19 -downdb dgvMerged humandb/
annotate_variation.pl -regionanno -build hg19 -out ex1 -dbtype dgvMerged example/ex1.avinput humandb/
<span class="c1"># published GWAS</span>
annotate_variation.pl -build hg19 -downdb gwasCatalog humandb/
annotate_variation.pl -regionanno -build hg19 -out ex1 -dbtype gwasCatalog example/ex1.avinput humandb/
</pre></div>
</div>
<p>Here is a more sophisticated example contrasting <code class="docutils literal notranslate"><span class="pre">table_annovar.pl</span></code> with <code class="docutils literal notranslate"><span class="pre">annotate_variation.pl</span></code>,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="o">[</span> ! -d <span class="nb">test</span> <span class="o">]</span><span class="p">;</span> <span class="k">then</span> mkdir test<span class="p">;</span> <span class="k">fi</span>
annotate_variation.pl -buildver hg19 -downdb -webfrom annovar refGene test/
annotate_variation.pl -buildver hg19 -downdb -webfrom annovar exac03 test/
annotate_variation.pl -buildver hg19 -downdb -webfrom annovar avsnp147 test/
annotate_variation.pl -buildver hg19 -downdb -webfrom annovar dbnsfp30a test/

annotate_variation.pl -geneanno -dbtype refGene -buildver hg19 example/ex1.avinput test/
annotate_variation.pl -filter -dbtype exac03 -buildver hg19 example/ex1.avinput test/

table_annovar.pl example/ex1.avinput test/ -buildver hg19 -out myanno <span class="se">\</span>
     -remove -protocol refGene,cytoBand,exac03,avsnp147,dbnsfp30a,gwasCatalog -operation g,r,f,f,f,r <span class="se">\</span>
     -nastring . -csvout -polish -xref example/gene_xref.txt
table_annovar.pl example/ex1.avinput test/ -buildver hg19 --outfile ex1 <span class="se">\</span>
     -protocol ensGene,refGene,ccdsGene,wgEncodeGencodeBasicV19,cytoBand,exac03,avsnp147,dbnsfp30a,gwasCatalog <span class="se">\</span>
     -operation g,g,g,g,r,f,f,f,r <span class="se">\</span>
     -csvout -polish -remove -nastring .
</pre></div>
</div>
<p>The second <code class="docutils literal notranslate"><span class="pre">table_annovar.pl</span></code> above works after symbolic links to relevant databases at humandb/ were created at test/.</p>
<p>It is handy to create a VCF file to be used by VEP (see below),</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>convert2annovar.pl -format annovar2vcf example/ex1.avinput &gt; ex1.vcf
vep -i ex1.vcf -o ex1.vcfoutput --offline --force_overwrite
</pre></div>
</div>
</div>
<div class="section" id="aegis">
<h2>AEGIS<a class="headerlink" href="#aegis" title="Permalink to this headline">¶</a></h2>
<p>Web: http://aegis.stanford.edu/</p>
<p>The installation follows the documentation but requires slight change,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># default Python 3.6</span>
module load python

virtualenv py36
<span class="nb">source</span> py36/bin/activate
git clone https://github.com/junjiezhujason/aegis.git
<span class="nb">cd</span> aegis

<span class="c1"># pip3 install -r requirements.txt</span>
cat requirements.txt <span class="p">|</span> <span class="se">\</span>
grep -v MarkupSafe <span class="p">|</span> <span class="se">\</span>
xargs pip3 install

<span class="c1"># setup</span>
<span class="c1">#!/usr/bin/bash</span>
<span class="nb">export</span> <span class="nv">PROJECT_PATH</span><span class="o">=</span>/home/<span class="nv">$USER</span>/aegis/data
<span class="nb">cd</span> <span class="si">${</span><span class="nv">PROJECT_PATH</span><span class="si">}</span>
<span class="c1"># wget -qO- http://stanford.edu/~jjzhu/fileshare/aegis/local_20180719.tar.gz | tar -xvzf local_20180719.tar.gz</span>
<span class="nb">cd</span> -
python3 main.py --lite --port <span class="m">5001</span> --folder <span class="si">${</span><span class="nv">PROJECT_PATH</span><span class="si">}</span> <span class="p">&amp;</span>
python3 main.py --port <span class="m">5002</span> --folder <span class="si">${</span><span class="nv">PROJECT_PATH</span><span class="si">}</span> <span class="p">&amp;</span>
firefox http://localhost:5002 <span class="p">&amp;</span>
</pre></div>
</div>
<p>as it appears that MarkupSafe could not be installed according to the predefined order.</p>
</div>
<div class="section" id="caveman">
<h2>CaVEMaN<a class="headerlink" href="#caveman" title="Permalink to this headline">¶</a></h2>
<p>Web, https://github.com/funpopgen/CaVEMaN/</p>
<p>Causal Variant Evidence Mapping with Non-parametric resampling (CaVEMaN) is one of the three fine-mapping software which provided data for GTEx v8.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># binary distribution</span>
https://github.com/funpopgen/CaVEMaN/releases/download/v1.01/CaVEMaN
<span class="c1"># build from source</span>
git clone https://github.com/funpopgen/CaVEMaN <span class="o">&amp;&amp;</span> <span class="nb">cd</span> CaVEMaN
module load xz
wget -qO- https://github.com/ldc-developers/ldc/releases/download/v1.24.0/ldc2-1.24.0-linux-x86_64.tar.xz <span class="p">|</span> <span class="se">\</span>
tar Jxf -
module load gsl/2.1
<span class="c1"># modify makefile </span>
LDC :<span class="o">=</span> <span class="si">${</span><span class="nv">PWD</span><span class="si">}</span>/ldc2-1.24.0-linux-x86_64/bin/ldc2
DMD :<span class="o">=</span> <span class="si">${</span><span class="nv">PWD</span><span class="si">}</span>/ldc2-1.24.0-linux-x86_64/bin/dmd
GSL :<span class="o">=</span> /usr/local/Cluster-Apps/gsl/2.1
make
</pre></div>
</div>
<p>Note that the D compiler setup actually requires gsl/2.1.</p>
</div>
<div class="section" id="bedops">
<h2>bedops<a class="headerlink" href="#bedops" title="Permalink to this headline">¶</a></h2>
<p>Web, https://bedops.readthedocs.io/en/latest/</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget -qO- https://github.com/bedops/bedops/releases/download/v2.4.39/bedops_linux_x86_64-v2.4.39.tar.bz2 <span class="p">|</span> <span class="se">\</span>
tar xfj -
</pre></div>
</div>
<p>would extract the executables into the bin/ directory.</p>
</div>
<div class="section" id="burden-testing">
<h2>burden_testing<a class="headerlink" href="#burden-testing" title="Permalink to this headline">¶</a></h2>
<p>Web: https://github.com/hmgu-itg/burden_testing/</p>
<p>To set it up, these are the necessary steps,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module avail singularity
module load singularity/current
<span class="c1"># download burten 1.6.5 from https://drive.google.com/file/d/1gelkV-wYydUWqmpc6ITPxE2O9cwyWfIJ/view</span>
singularity <span class="nb">exec</span> burden_testing_latest <span class="nb">help</span>
<span class="c1"># LINSIGHT</span>
<span class="c1"># https://github.com/CshlSiepelLab/LINSIGHT/</span>
wget http://compgen.cshl.edu/LINSIGHT/LINSIGHT.bw
<span class="c1"># CADD</span>
wget https://krishna.gs.washington.edu/download/CADD/v1.6/GRCh38/whole_genome_SNVs.tsv.gz
<span class="c1"># Eigen</span>
annotate_variation.pl -build hg19 -webfrom annovar -downdb eigen .
<span class="c1"># rclone, https://rclone.org/downloads/</span>
wget -qO- https://downloads.rclone.org/v1.53.1/rclone-v1.53.1-linux-amd64.zip <span class="p">|</span> jar xv
</pre></div>
</div>
<p>See https://github.com/hmgu-itg/burden_testing/wiki for more. Note Eigen is unavailable but hg19 version can be downloaded via ANNOVAR.</p>
</div>
<div class="section" id="r-brms">
<h2>R/brms<a class="headerlink" href="#r-brms" title="Permalink to this headline">¶</a></h2>
<p>It is available from CRAN, so it can be installed with <code class="docutils literal notranslate"><span class="pre">install.packages()</span></code>. Here is a documentation example,</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span>  <span class="nf">library</span><span class="p">(</span><span class="n">brms</span><span class="p">)</span>
<span class="c1"># Survival regression modeling the time between the first</span>
<span class="c1"># and second recurrence of an infection in kidney patients.</span>
  <span class="n">fit3</span> <span class="o">&lt;-</span> <span class="nf">brm</span><span class="p">(</span><span class="n">time</span> <span class="o">|</span> <span class="nf">cens</span><span class="p">(</span><span class="n">censored</span><span class="p">)</span> <span class="o">~</span> <span class="n">age</span> <span class="o">*</span> <span class="n">sex</span> <span class="o">+</span> <span class="n">disease</span> <span class="o">+</span> <span class="p">(</span><span class="m">1</span><span class="o">|</span><span class="n">patient</span><span class="p">),</span>
              <span class="n">data</span> <span class="o">=</span> <span class="n">kidney</span><span class="p">,</span> <span class="n">family</span> <span class="o">=</span> <span class="nf">lognormal</span><span class="p">())</span>
  <span class="nf">summary</span><span class="p">(</span><span class="n">fit3</span><span class="p">)</span>
  <span class="nf">plot</span><span class="p">(</span><span class="n">fit3</span><span class="p">,</span> <span class="n">ask</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
  <span class="nf">plot</span><span class="p">(</span><span class="nf">conditional_effects</span><span class="p">(</span><span class="n">fit3</span><span class="p">),</span> <span class="n">ask</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
  <span class="n">fit3cox</span> <span class="o">&lt;-</span> <span class="nf">brm</span><span class="p">(</span><span class="n">time</span> <span class="o">|</span> <span class="nf">cens</span><span class="p">(</span><span class="n">censored</span><span class="p">)</span> <span class="o">~</span> <span class="n">age</span> <span class="o">*</span> <span class="n">sex</span> <span class="o">+</span> <span class="n">disease</span> <span class="o">+</span> <span class="p">(</span><span class="m">1</span><span class="o">|</span><span class="n">patient</span><span class="p">),</span>
                 <span class="n">data</span> <span class="o">=</span> <span class="n">kidney</span><span class="p">,</span> <span class="n">family</span> <span class="o">=</span> <span class="nf">family</span><span class="p">(</span><span class="s">&#39;cox&#39;</span><span class="p">))</span>
  <span class="nf">summary</span><span class="p">(</span><span class="n">fit3cox</span><span class="p">)</span>
  <span class="nf">plot</span><span class="p">(</span><span class="n">fit3cox</span><span class="p">,</span> <span class="n">ask</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</div>
<p>In case of compilng error, it is necessary to use the developmental version of rstan (2.21.3 at the time of writing), see below.</p>
</div>
<div class="section" id="dosageconverter">
<h2>DosageConverter<a class="headerlink" href="#dosageconverter" title="Permalink to this headline">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1">## assuming you use hpc-work/ with a subdirectory called bin/</span>
<span class="nb">cd</span> /rds/user/<span class="nv">$USER</span>/hpc-work/
git clone https://github.com/Santy-8128/DosageConvertor
<span class="nb">cd</span> DosageConverter
pip install cget --user
module load cmake-3.8.1-gcc-4.8.5-zz55m7x
./install.sh
<span class="nb">cd</span> /rds/user/<span class="nv">$USER</span>/hpc-work/bin/
ln -s /rds/user/<span class="nv">$USER</span>/hpc-work/DosageConvertor/release-build/DosageConvertor
<span class="c1">## testing</span>
DosageConvertor  --vcfDose  test/TestDataImputedVCF.dose.vcf.gz <span class="se">\</span>
                 --info     test/TestDataImputedVCF.info <span class="se">\</span>
                 --prefix   <span class="nb">test</span> <span class="se">\</span>
                 --type     mach

gunzip -c test.mach.dose.gz <span class="p">|</span> wc -l

DosageConvertor  --vcfDose  test/TestDataImputedVCF.dose.vcf.gz <span class="se">\</span>
                 --info     test/TestDataImputedVCF.info <span class="se">\</span>
                 --prefix   <span class="nb">test</span> <span class="se">\</span>
                 --type     plink

gunzip -c test.plink.dosage.gz <span class="p">|</span> wc -l
</pre></div>
</div>
<p>so the MaCH dosage file is individual x genotype whereas PLINK dosage file is genotype x individual.</p>
</div>
<div class="section" id="hess">
<h2>HESS<a class="headerlink" href="#hess" title="Permalink to this headline">¶</a></h2>
<p>This section is extracted from https://github.com/jinghuazhao/software-notes.</p>
<p>HESS (Heritability Estimation from Summary Statistics) is now available from https://github.com/huwenboshi/hess and has a web page at</p>
<p>https://huwenboshi.github.io/hess-0.5/#hess</p>
<p>To prepare for the software, one can proceeds with</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python -m pip install pysnptools --user
</pre></div>
</div>
<p>Now we set up for analysis of height</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash</span>

<span class="nb">export</span> <span class="nv">HEIGHT</span><span class="o">=</span>https://portals.broadinstitute.org/collaboration/giant/images/0/01/GIANT_HEIGHT_Wood_et_al_2014_publicrelease_HapMapCeuFreq.txt.gz

wget -qO- <span class="nv">$HEIGHT</span> <span class="p">|</span> <span class="se">\</span>
awk <span class="s1">&#39;NR&gt;1&#39;</span> <span class="p">|</span> <span class="se">\</span>
sort -k1,1 <span class="p">|</span> <span class="se">\</span>
join -13 -21 snp150.txt - <span class="p">|</span> <span class="se">\</span>
awk <span class="s1">&#39;($9!=&quot;X&quot; &amp;&amp; $9!=&quot;Y&quot; &amp;&amp; $9!=&quot;Un&quot;){if(NR==1) print &quot;SNP CHR BP A1 A2 Z N&quot;; else print $1,$2,$3,$4,$5,$7/$8,$10}&#39;</span> &gt; height.tsv.gz

<span class="c1">#  SNP - rs ID of the SNP (e.g. rs62442).</span>
<span class="c1">#  CHR - Chromosome number of the SNP. This should be a number between 1 and 22.</span>
<span class="c1">#  BP - Base pair position of the SNP.</span>
<span class="c1">#  A1 - Effect allele of the SNP. The sign of the Z-score is with respect to this allele.</span>
<span class="c1">#  A2 - The other allele of the SNP.</span>
<span class="c1">#  Z - The Z-score of the SNP.</span>
<span class="c1">#  N - Sample size of the SNP.</span>
</pre></div>
</div>
<p>where snp150.txt from UCSC is described at the SUMSTATS repository, <a class="reference external" href="https://github.com/jinghuazhao/SUMSTATS">https://github.com/jinghuazhao/SUMSTATS</a>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> chrom in <span class="k">$(</span>seq <span class="m">22</span><span class="k">)</span>
<span class="k">do</span>
    python hess.py <span class="se">\</span>
        --local-hsqg height <span class="se">\</span>
        --chrom <span class="nv">$chrom</span> <span class="se">\</span>
        --bfile 1kg_eur_1pct/1kg_eur_1pct_chr<span class="si">${</span><span class="nv">chrom</span><span class="si">}</span> <span class="se">\</span>
        --partition nygcresearch-ldetect-data-ac125e47bf7f/EUR/fourier_ls-chr<span class="si">${</span><span class="nv">chrom</span><span class="si">}</span>.bed <span class="se">\</span>
        --out step1
<span class="k">done</span>
python hess.py --prefix step1 --reinflate-lambda-gc <span class="m">1</span> --tot-hsqg <span class="m">0</span>.8 <span class="m">0</span>.2 --out step2
</pre></div>
</div>
<p>It is preferable to use <code class="docutils literal notranslate"><span class="pre">miniconda</span></code> since it associates with faster libraries.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load miniconda2-4.3.14-gcc-5.4.0-xjtq53h
conda install pandas
</pre></div>
</div>
</div>
<div class="section" id="r-gnn">
<h2>R/gnn<a class="headerlink" href="#r-gnn" title="Permalink to this headline">¶</a></h2>
<p>It requires libgsl, so</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load gsl/2.4
</pre></div>
</div>
</div>
<div class="section" id="gsutil">
<h2>gsutil<a class="headerlink" href="#gsutil" title="Permalink to this headline">¶</a></h2>
<p>Web site: https://cloud.google.com/storage/docs/gsutil_install#linux</p>
<p>the authentification is achieved via Google SDK (e.g. <a class="reference external" href="https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-sdk-301.0.0-linux-x86_64.tar.gz">301.0.0</a>)</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-sdk-301.0.0-linux-x86_64.tar.gz
tar xfz google-cloud-sdk-301.0.0-linux-x86_64.tar.gz
<span class="nb">cd</span> google-cloud-sdk
./install.sh
<span class="c1"># to gain a bit flexibility by not using the awkward browser (Konqueror) at CSD3</span>
gcloud auto login --no-launch-browser
<span class="c1"># to set PROJECT_ID=divine-aegis-278909</span>
gcloud config <span class="nb">set</span> divine-aegis-278909
</pre></div>
</div>
<p>followed by <code class="docutils literal notranslate"><span class="pre">gsutil</span> <span class="pre">config</span></code> where <code class="docutils literal notranslate"><span class="pre">gsutil</span></code> is installed via <code class="docutils literal notranslate"><span class="pre">virtualenv</span></code> as follows,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load python/3.7
virtualenv py37
<span class="nb">source</span> py37/bin/activate
pip install <span class="nv">gsutil</span><span class="o">==</span><span class="m">4</span>.50
<span class="nb">cd</span> <span class="si">${</span><span class="nv">dir</span><span class="si">}</span>/py37/bin
gsutil ls
<span class="c1"># gsutil cp test gs::/covid19-hg-upload-bugbank</span>
gsutil cp <span class="nb">test</span> gs://covid19-hg-upload-uk--blood-donors-cohort
</pre></div>
</div>
<p>Less useful is the usual way to install.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget https://storage.googleapis.com/pub/gsutil.tar.gz
tar xvfz gsutil.tar.gz -C ..
<span class="nb">cd</span> ../gsutil
pip install <span class="nv">pyasn1</span><span class="o">==</span><span class="m">0</span>.4.8  --user
python setup.py install --prefix<span class="o">=</span><span class="nv">$HPC_WORK</span>
</pre></div>
</div>
</div>
<div class="section" id="gwas2vcf">
<h2>gwas2vcf<a class="headerlink" href="#gwas2vcf" title="Permalink to this headline">¶</a></h2>
<p>Web: https://github.com/MRCIEU/gwas2vcf</p>
<p>The required setup script is as follows,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load jdk-8u141-b15-gcc-5.4.0-p4aaopt
module load gatk
module load python/3.7
git clone https://github.com/MRCIEU/gwas2vcf
<span class="nb">cd</span> gwas2vcf
python -m venv env
<span class="nb">source</span> env/bin/activate
pip install -r requirements.txt
pip install git+git://github.com/bioinformed/vgraph@v1.4.0#egg<span class="o">=</span>vgraph
ln -s ~/rds/public_databases/GRCh37_reference_fasta/human_g1k_v37.fasta
ln -s ~/rds/public_databases/GRCh37_reference_fasta/human_g1k_v37.fasta.fai
python -m pytest -v <span class="nb">test</span>
python main.py -h

<span class="c1">## these are not necessarily required and GRC37 is here, ~/rds/public_databases/GRCh37_reference_fasta/</span>
<span class="c1"># GRCh36/hg18/b36</span>
wget http://fileserve.mrcieu.ac.uk/ref/2.8/b36/human_b36_both.fasta
wget http://fileserve.mrcieu.ac.uk/ref/2.8/b36/human_b36_both.fasta.fai

<span class="c1"># GRCh37/hg19/b37</span>
wget http://fileserve.mrcieu.ac.uk/ref/2.8/b37/human_g1k_v37.fasta
wget http://fileserve.mrcieu.ac.uk/ref/2.8/b37/human_g1k_v37.fasta.fai

<span class="c1"># GRCh38/hg38/b38</span>
wget https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta
wget https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta.fai

<span class="c1">## ENSEMBL and Reactome</span>

<span class="c1"># download ensembl-to-position mapping and sort by ensgId</span>
curl ftp://ftp.ensembl.org/pub/grch37/release-87/gff3/homo_sapiens/Homo_sapiens.GRCh37.87.gff3.gz <span class="p">|</span> <span class="se">\</span>
gzip -dc <span class="p">|</span> <span class="se">\</span>
awk -F<span class="s2">&quot;\t|:|;|=&quot;</span> <span class="s1">&#39;$3==&quot;gene&quot; &amp;&amp; $1 &gt;= 1 &amp;&amp; $1 &lt;= 22 {print $11&quot;\t&quot;$1&quot;\t&quot;$4&quot;\t&quot;$5}&#39;</span> <span class="p">|</span> <span class="se">\</span>
sort -k1,1 &gt; Ensembl2Position.sorted.txt

<span class="c1"># download ensembl-to-pathway mapping and sort by ensgId</span>
curl https://reactome.org/download/current/Ensembl2Reactome.txt <span class="p">|</span> <span class="se">\</span>
grep <span class="s2">&quot;Homo sapiens&quot;</span> <span class="p">|</span> <span class="se">\</span>
cut -s -f1,2 <span class="p">|</span> <span class="se">\</span>
sort -k1,1 &gt; Ensembl2Reactome.sorted.txt

<span class="c1"># merge tables by ensgId, sort, compress and index</span>
join <span class="se">\</span>
-t <span class="s1">$&#39;\t&#39;</span> <span class="se">\</span>
--check-order <span class="se">\</span>
Ensembl2Position.sorted.txt <span class="se">\</span>
Ensembl2Reactome.sorted.txt <span class="p">|</span> <span class="se">\</span>
awk -F<span class="s2">&quot;\t&quot;</span> <span class="s1">&#39;{print $2&quot;\t&quot;$3-1&quot;\t&quot;$4&quot;\t&quot;$5}&#39;</span> <span class="p">|</span> <span class="se">\</span>
sort -k1,1V -k2,2n -k3,3n <span class="p">|</span> <span class="se">\</span>
bgzip -c &gt; reactome.bed.gz
tabix -p bed reactome.bed.gz

<span class="c1"># sort, compress and index ensembl-to-position mapping</span>
awk -F<span class="s2">&quot;\t&quot;</span> <span class="s1">&#39;{print $2&quot;\t&quot;$3-1&quot;\t&quot;$4&quot;\t&quot;$1}&#39;</span> Ensembl2Position.sorted.txt <span class="p">|</span> <span class="se">\</span>
sort -k1,1V -k2,2n -k3,3n  <span class="p">|</span> <span class="se">\</span>
bgzip -c &gt; ensembl.bed.gz
tabix -p bed ensembl.bed.gz
</pre></div>
</div>
<p>Example scripts for GWAS VCF operations,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>  module load jdk-8u141-b15-gcc-5.4.0-p4aaopt
  module load gatk
  module load python/3.7
  module load picard/2.9.2
  <span class="nb">export</span> <span class="nv">ensembl</span><span class="o">=</span><span class="si">${</span><span class="nv">HPC_WORK</span><span class="si">}</span>/gwas2vcf/ensembl
  <span class="nb">export</span> <span class="nv">reference</span><span class="o">=</span><span class="si">${</span><span class="nv">HPC_WORK</span><span class="si">}</span>/gwas2vcf/human_g1k_v37
  <span class="nb">export</span> <span class="nv">input</span><span class="o">=</span><span class="si">${</span><span class="nv">INF</span><span class="si">}</span>/METAL/vcf/<span class="si">${</span><span class="nv">prot</span><span class="si">}</span>
  <span class="nb">export</span> <span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">INF</span><span class="si">}</span>/work/<span class="si">${</span><span class="nv">prot</span><span class="si">}</span>-b38
<span class="c1"># extract a region</span>
  bcftools filter -r <span class="m">1</span>:1000000-2000000 -o <span class="si">${</span><span class="nv">input</span><span class="si">}</span>-r.vcf.gz <span class="si">${</span><span class="nv">input</span><span class="si">}</span>.vcf.gz
<span class="c1"># filter on p values</span>
  bcftools filter -i <span class="s1">&#39;FORMAT/LP &gt; 7.3&#39;</span> -o <span class="si">${</span><span class="nv">input</span><span class="si">}</span>-7.3.vcf <span class="si">${</span><span class="nv">input</span><span class="si">}</span>.vcf.gz
<span class="c1"># annotate GWAS-VCF</span>
  bcftools annotate -a <span class="si">${</span><span class="nv">ensembl</span><span class="si">}</span>.bed.gz -c CHROM,FROM,TO,ENSG_ID <span class="se">\</span>
                    -h &lt;<span class="o">(</span><span class="nb">echo</span> <span class="s1">&#39;##INFO=&lt;ID=ENSG_ID,Number=.,Type=String,Description=&quot;Ensembl gene ID&quot;&gt;&#39;</span><span class="o">)</span> <span class="se">\</span>
                    -o <span class="si">${</span><span class="nv">output</span><span class="si">}</span>.vcf.gz -O z -l ENSG_ID:unique <span class="si">${</span><span class="nv">input</span><span class="si">}</span>.vcf.gz
<span class="c1"># GWAS Catalogue</span>
  bcftools query -e <span class="s1">&#39;ID == &quot;.&quot;&#39;</span> -f <span class="s1">&#39;%ID\t[%LP]\t%CHROM\t%POS\t%ALT\t%REF\t%AF\t[%ES\t%SE]\n&#39;</span> <span class="si">${</span><span class="nv">input</span><span class="si">}</span>.vcf.gz <span class="p">|</span> <span class="se">\</span>
  awk <span class="s1">&#39;BEGIN {</span>
<span class="s1">  print &quot;variant_id\tp_value\tchromosome\tbase_pair_location\teffect_allele\tother_allele\teffect_allele_frequency\tbeta\tstandard_error&quot;</span>
<span class="s1">  }; {OFS=&quot;\t&quot;; if ($2==0) $2=1; else if ($2==999) $2=0; else $2=10^-$2; print}&#39;</span> <span class="p">|</span> <span class="se">\</span>
  gzip -f &gt; <span class="si">${</span><span class="nv">input</span><span class="si">}</span>.tsv.gz
<span class="c1"># liftOver</span>
  java -jar /usr/local/Cluster-Apps/picard/2.9.2/picard.jar CreateSequenceDictionary <span class="nv">R</span><span class="o">=</span><span class="si">${</span><span class="nv">reference</span><span class="si">}</span>.fasta <span class="nv">O</span><span class="o">=</span><span class="si">${</span><span class="nv">reference</span><span class="si">}</span>.dict
  gatk LiftoverVcf --INPUT <span class="si">${</span><span class="nv">input</span><span class="si">}</span>.vcf.gz --OUTPUT <span class="si">${</span><span class="nv">output</span><span class="si">}</span>.vcf.gz <span class="se">\</span>
                   --REJECT <span class="si">${</span><span class="nv">input</span><span class="si">}</span>-rejected.vcf.gz --CHAIN <span class="si">${</span><span class="nv">HPC_WORK</span><span class="si">}</span>/bin/hg19ToHg38.over.chain <span class="se">\</span>
                   --REFERENCE_SEQUENCE <span class="si">${</span><span class="nv">reference</span><span class="si">}</span>.fasta --RECOVER_SWAPPED_REF_ALT <span class="nb">false</span>
<span class="c1"># Validation</span>
  gatk ValidateVariants -V <span class="si">${</span><span class="nv">output</span><span class="si">}</span>.vcf.gz -R <span class="si">${</span><span class="nv">reference</span><span class="si">}</span>.fasta --validation-type-to-exclude ALLELES
</pre></div>
</div>
<p>To use the latest dbSNP information, these steps can be helpful,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># hg19</span>
<span class="c1"># wget https://ftp.ncbi.nih.gov/snp/archive/b154/VCF/GCF_000001405.25.gz -O snp154_GCF_000001405.25.gz</span>
<span class="c1"># wget https://ftp.ncbi.nih.gov/snp/archive/b154/VCF/GCF_000001405.25.gz.tbi -O snp154_GCF_000001405.25.gz.tbi</span>
gunzip -c snp154_GCF_000001405.25.gz &gt; snp154_GCF_000001405.25
<span class="c1"># alias file for chomosomes</span>
<span class="c1"># https://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/chromToUcsc</span>

<span class="c1"># One can carry on with the familiar chr1, ..., etc.</span>
<span class="c1"># https://github.com/vkkodali/cthreepo</span>
cthreepo <span class="se">\</span>
    --infile snp154_GCF_000001405.25 <span class="se">\</span>
    --id_from rs <span class="se">\</span>
    --id_to uc <span class="se">\</span>
    --format vcf <span class="se">\</span>
    --mapfile h37 <span class="se">\</span>
    --outfile snp154_hg19.vcf

bgzip -f snp154_hg19.vcf
tabix -f snp154_hg19.vcf.gz

<span class="c1"># hg38</span>
cthreepo --infile snp154_GCF_000001405.38 --id_from rs --id_to uc --format vcf --mapfile h38 --outfile snp154_hg38.vcf
bgzip -f snp154_hg38.vcf
tabix -f snp154_hg38.vcf.gz
</pre></div>
</div>
</div>
<div class="section" id="kentutils">
<h2>KentUtils<a class="headerlink" href="#kentutils" title="Permalink to this headline">¶</a></h2>
<p>liftOver from UCSC is here, https://genome.ucsc.edu/cgi-bin/hgLiftOver.</p>
<p>This is part of the Kent utilities in module <code class="docutils literal notranslate"><span class="pre">kentutils-302.1-gcc-5.4.0-kbiujaa</span></code> nevertheless without the appropriate chain file.</p>
<p>To download the latest utilitiess, try <code class="docutils literal notranslate"><span class="pre">rsync</span> <span class="pre">-aP</span> <span class="pre">rsync://hgdownload.soe.ucsc.edu/genome/admin/exe/linux.x86_64/</span> <span class="pre">./</span></code>.</p>
</div>
<div class="section" id="lapack">
<h2>lapack<a class="headerlink" href="#lapack" title="Permalink to this headline">¶</a></h2>
<p>The sequence is as follows,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget http://www.netlib.org/lapack/lapack-3.9.0.tar.gz
tar xvfz lapack-3.9.0.tar.gz
<span class="nb">cd</span> lapack-3.9.0
mkdir build
<span class="nb">cd</span> build
<span class="c1">## ccmake .</span>
cmake ..
make
make install
</pre></div>
</div>
<p>Note <code class="docutils literal notranslate"><span class="pre">ccmake</span></code> is commented as it does not work properly on csd3. By default these build static libraries into build/lib/. It is also possible to build the static libraries in lapack-3.9.0/ by making a copy of <code class="docutils literal notranslate"><span class="pre">make.inc.example</span></code> as <code class="docutils literal notranslate"><span class="pre">make.inc</span></code> and compile <code class="docutils literal notranslate"><span class="pre">libblas.a</span></code>,<code class="docutils literal notranslate"><span class="pre">liblapack.a</span></code> as well as <code class="docutils literal notranslate"><span class="pre">libcblas.a</span></code> and <code class="docutils literal notranslate"><span class="pre">liblapacke.a</span></code> from BLAS/ and LAPACKE/ directories.</p>
<p>To enable the shared libraries, proceed with</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cmake -DCMAKE_INSTALL_PREFIX<span class="o">=</span><span class="si">${</span><span class="nv">HPC_WORK</span><span class="si">}</span> -DCMAKE_BUILD_TYPE<span class="o">=</span>RELEASE -DBUILD_SHARED_LIBS<span class="o">=</span>ON -DLAPACKE<span class="o">=</span>ON -DCBLAS<span class="o">=</span>ON ..
</pre></div>
</div>
</div>
<div class="section" id="r-ldlinkr">
<h2>R/LDlinkR<a class="headerlink" href="#r-ldlinkr" title="Permalink to this headline">¶</a></h2>
<p>Issue <code class="docutils literal notranslate"><span class="pre">install.packages(&quot;LDlinkR&quot;)</span></code> from R but requires registration at <a class="reference external" href="https://ldlink.nci.nih.gov/?tab=apiaccess">https://ldlink.nci.nih.gov/?tab=apiaccess</a>.</p>
</div>
<div class="section" id="r-phenoscanner">
<h2>R/PhenoScanner<a class="headerlink" href="#r-phenoscanner" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>The R package works as follows,</p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>install.packages<span class="o">(</span><span class="s2">&quot;devtools&quot;</span><span class="o">)</span>
library<span class="o">(</span>devtools<span class="o">)</span>
install_github<span class="o">(</span><span class="s2">&quot;phenoscanner/phenoscanner&quot;</span><span class="o">)</span>
library<span class="o">(</span>phenoscanner<span class="o">)</span>
example<span class="o">(</span>phenoscanner<span class="o">)</span>
</pre></div>
</div>
<ol class="simple">
<li><p>When the query list is long, the call is made by chunks, e.g.,</p></li>
</ol>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">options</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="m">500</span><span class="p">)</span>
<span class="nf">require</span><span class="p">(</span><span class="n">phenoscanner</span><span class="p">)</span>
<span class="n">rsid</span> <span class="o">&lt;-</span> <span class="nf">scan</span><span class="p">(</span><span class="s">&quot;swath-ms-invn.snp&quot;</span><span class="p">,</span><span class="n">what</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
<span class="n">batches</span> <span class="o">&lt;-</span> <span class="nf">split</span><span class="p">(</span><span class="n">rsid</span><span class="p">,</span> <span class="nf">ceiling</span><span class="p">(</span><span class="nf">seq_along</span><span class="p">(</span><span class="n">rsid</span><span class="p">)</span><span class="o">/</span><span class="m">100</span><span class="p">))</span>
<span class="n">s</span> <span class="o">&lt;-</span> <span class="n">t</span> <span class="o">&lt;-</span> <span class="nf">list</span><span class="p">()</span>
<span class="nf">for</span><span class="p">(</span><span class="n">i</span> <span class="n">in</span> <span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">batches</span><span class="p">))</span>
<span class="p">{</span>
  <span class="n">q</span> <span class="o">&lt;-</span> <span class="nf">phenoscanner</span><span class="p">(</span><span class="n">snpquery</span><span class="o">=</span><span class="n">batches</span><span class="p">[[</span><span class="n">i</span><span class="p">]],</span> <span class="n">catalogue</span><span class="o">=</span><span class="s">&quot;pQTL&quot;</span><span class="p">,</span> <span class="n">proxies</span> <span class="o">=</span> <span class="s">&quot;EUR&quot;</span><span class="p">,</span> <span class="n">pvalue</span> <span class="o">=</span> <span class="m">1e-07</span><span class="p">,</span> <span class="n">r2</span><span class="o">=</span> <span class="m">0.6</span><span class="p">,</span> <span class="n">build</span><span class="o">=</span><span class="m">37</span><span class="p">)</span>
  <span class="n">s</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="nf">with</span><span class="p">(</span><span class="n">q</span><span class="p">,</span><span class="n">snps</span><span class="p">)</span>
  <span class="n">t</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="nf">with</span><span class="p">(</span><span class="n">q</span><span class="p">,</span><span class="n">results</span><span class="p">)</span>
<span class="p">}</span>
<span class="n">snps</span> <span class="o">&lt;-</span> <span class="nf">do.call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span><span class="n">s</span><span class="p">)</span>
<span class="n">results</span> <span class="o">&lt;-</span> <span class="nf">do.call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span><span class="n">t</span><span class="p">)</span>
<span class="n">r</span> <span class="o">&lt;-</span> <span class="nf">list</span><span class="p">(</span><span class="n">snps</span><span class="o">=</span><span class="n">snps</span><span class="p">,</span><span class="n">results</span><span class="o">=</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<p>i.e., each chunk has 100 SNPs and chunks are combined manually.</p>
<ol class="simple">
<li><p>The command-line interface (CLI)</p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load ceuadmin/phenoscanner
phenoscanner --help
phenoscanner --snp<span class="o">=</span>rs704 -c All -x EUR -r <span class="m">0</span>.8
phenoscanner -s T -c All -x EUR -p <span class="m">0</span>.0000001 --r2 <span class="m">0</span>.6 -i INF1.merge.snp -o INF1
</pre></div>
</div>
<p>Note that <code class="docutils literal notranslate"><span class="pre">module</span> <span class="pre">load</span> <span class="pre">phenoscanner</span></code> is enabled from ~/.bashrc:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>export MODULEPATH=${MODULEPATH}:/usr/local/Cluster-Config/modulefiles/ceuadmin/
</pre></div>
</div>
<p>via <code class="docutils literal notranslate"><span class="pre">source</span> <span class="pre">~/.bashrc</span></code> or a new login.</p>
<ol class="simple">
<li><p>Work with R 4.x.x. Section 3 above would fail under R 4.x.x; to get around, make a copy of phenoscanner according to</p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load ceuadmin/phenoscanner
which phenoscanner
<span class="c1"># /rds/project/jmmh2/rds-jmmh2-projects/phenoscanner/mrcatalogue/mrcatalogue/phenoscanner_v2/phenoscanner</span>
</pre></div>
</div>
<p>and edit the header to call packages at the default R_LIBS location,</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">#!/rds/user/jhz22/hpc-work/bin/Rscript</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="n">getopt</span><span class="p">))</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="n">optparse</span><span class="p">))</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="n">DBI</span><span class="p">))</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="n">RMySQL</span><span class="p">))</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="n">reshape2</span><span class="p">))</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="n">plyr</span><span class="p">))</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="n">stringi</span><span class="p">))</span>
</pre></div>
</div>
<p>then deposit this to a directory on the search path and invoke,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load gcc/6
phenoscanner -h
</pre></div>
</div>
</div>
<div class="section" id="metaxcan">
<h2>MetaXcan<a class="headerlink" href="#metaxcan" title="Permalink to this headline">¶</a></h2>
<p>Web: https://github.com/hakyimlab/MetaXcan</p>
<p>Installation and a test run,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git clone https://github.com/hakyimlab/MetaXcan
<span class="nb">cd</span> MetaXcan/software
<span class="c1"># download sample_data.tar.gz</span>
tar -xzvpf sample_data.tar.gz
module load miniconda3/4.5.1
conda env create -f <span class="nv">$HPC_WORK</span>/MetaXcan/software/conda_env.yaml
conda activate imlabtools
conda install pandas
python setup.py install --user

./SPrediXcan.py <span class="se">\</span>
--model_db_path data/DGN-WB_0.5.db <span class="se">\</span>
--covariance data/covariance.DGN-WB_0.5.txt.gz <span class="se">\</span>
--gwas_folder data/GWAS <span class="se">\</span>
--gwas_file_pattern <span class="s2">&quot;.*gz&quot;</span> <span class="se">\</span>
--snp_column SNP <span class="se">\</span>
--effect_allele_column A1 <span class="se">\</span>
--non_effect_allele_column A2 <span class="se">\</span>
--beta_column BETA <span class="se">\</span>
--pvalue_column P <span class="se">\</span>
--output_file results/test.csv
</pre></div>
</div>
</div>
<div class="section" id="polyphen-2">
<h2>polyphen-2<a class="headerlink" href="#polyphen-2" title="Permalink to this headline">¶</a></h2>
<p>Official page: <a class="reference external" href="http://genetics.bwh.harvard.edu/pph2/dokuwiki/start">http://genetics.bwh.harvard.edu/pph2/dokuwiki/start</a>.</p>
<p>The setup can be furnished as follows,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HPC_WORK</span>
wget -qO- http://genetics.bwh.harvard.edu/pph2/dokuwiki/_media/polyphen-2.2.2r405c.tar.gz <span class="p">|</span> tar xfz
wget -qO- ftp://genetics.bwh.harvard.edu/pph2/bundled/polyphen-2.2.2-databases-2011_12.tar.bz2 <span class="p">|</span> tar xjf
wget -qO- ftp://genetics.bwh.harvard.edu/pph2/bundled/polyphen-2.2.2-alignments-mlc-2011_12.tar.bz2 <span class="p">|</span> tar xjf
wget -qO- ftp://genetics.bwh.harvard.edu/pph2/bundled/polyphen-2.2.2-alignments-multiz-2009_10.tar.bz2 <span class="p">|</span> tar xjf
ls  <span class="p">|</span> sed <span class="s1">&#39;s/\*//g&#39;</span> <span class="p">|</span> parallel -C<span class="s1">&#39; &#39;</span> <span class="s1">&#39;ln -sf $HPC_WORK/polyphen-2.2.2/bin/{} $HPC_WORK/bin/{}&#39;</span>
<span class="nb">cd</span> polyphen-2.2.2
<span class="c1"># set up BLAST/nrdb/PDB as decribed below</span>
<span class="nb">cd</span> src
make
make install
<span class="nb">cd</span> -
configure
<span class="nb">cd</span> bin
rsync -aP rsync://hgdownload.soe.ucsc.edu/genome/admin/exe/linux.x86_64/ ./
<span class="nb">cd</span> -
</pre></div>
</div>
<p>The MLC/MULTIZ databases need to be extracted to $HOME and symbolically linked if the number of files exceed 1 million
(limit on RDS). Then these are necessary,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HPC_WORK</span>/polyphen-2.2.2
ln -s <span class="nv">$HOME</span>/polyphen-2.2.2/precompiled
<span class="nb">cd</span> ucsc/hg19/multiz
ln -s <span class="nv">$HOME</span>/polyphen-2.2.2/ucsc/hg19/multiz/precomputed
</pre></div>
</div>
<p>The availability of MLC/MULTIZ databases make the annotation considerably faster.</p>
<p>The command <code class="docutils literal notranslate"><span class="pre">configure</span></code> creates files at config/ which can be changed maunaually. There is also
<a class="reference external" href="http://genetics.bwh.harvard.edu/pph2/dokuwiki/_media/hg0720.pdf">user’s guide</a>. The line <code class="docutils literal notranslate"><span class="pre">rsync</span></code> obtains
programs such as <code class="docutils literal notranslate"><span class="pre">twoBitToFa</span></code> as required by the example below.</p>
<p>BLAST and nrdb can be set up as follows,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>rmdir blast
ln -sf /usr/local/Cluster-Apps/blast/2.4.0 blast
<span class="nb">cd</span> nrdb
wget -qO- ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/uniref/uniref100/uniref100.fasta.gz <span class="p">|</span> <span class="se">\</span>
gunzip -c &gt; uniref100.fasta
../update/format_defline.pl uniref100.fasta &gt;uniref100-formatted.fasta
../blast/bin/makeblastdb -in uniref100-formatted.fasta -dbtype prot -out uniref100 -parse_seqids
rm -f uniref100.fasta uniref100-formatted.fasta
</pre></div>
</div>
<p>and for PDB</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>rsync -rltv --delete-after --port<span class="o">=</span><span class="m">33444</span> <span class="se">\</span>
      rsync.wwpdb.org::ftp/data/structures/divided/pdb/ wwpdb/divided/pdb/
rsync -rltv --delete-after --port<span class="o">=</span><span class="m">33444</span> <span class="se">\</span>
      rsync.wwpdb.org::ftp/data/structures/all/pdb/ wwpdb/all/pdb/
</pre></div>
</div>
<p>Our test is then,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HPC_WORK</span>/polyphen-2.2.2
run_pph.pl sets/test.input <span class="m">1</span>&gt;test.pph.output <span class="m">2</span>&gt;test.pph.log
run_weka.pl test.pph.output &gt;test.humdiv.output
run_weka.pl -l models/HumVar.UniRef100.NBd.f11.model test.pph.output &gt;test.humvar.output

sdiff test.humdiv.output sets/test.humdiv.output
sdiff test.humvar.output sets/test.humvar.output
</pre></div>
</div>
<p>Now we turn to an genomic SNPs query examples with snps.pph.list containing the following line,</p>
<blockquote>
<div><p>chr1:154426970 A/C</p>
</div></blockquote>
<p>to be called by <code class="docutils literal notranslate"><span class="pre">mapsnps.pl</span></code> and others.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mapsnps.pl -g hg19 -m -U -y snps.pph.input snps.pph.list <span class="m">1</span>&gt;snps.pph.features <span class="m">2</span>&gt;snps.log
run_pph.pl snps.pph.input <span class="m">1</span>&gt;snps.pph.output <span class="m">2</span>&gt;snps.pph.log
run_weka.pl snps.pph.output &gt;snps.humdiv.output
run_weka.pl -l models/HumVar.UniRef100.NBd.f11.model snps.pph.output &gt;snps.humvar.output
</pre></div>
</div>
<p>for <a class="reference external" href="files/snps.pph.input">.pph.input</a>, <a class="reference external" href="files/snps.pph.features">.pph.features</a>, <a class="reference external" href="files/snps.pph.output">.pph.output</a>, <a class="reference external" href="files/snps.humvar.output">.humvar.output</a> and <a class="reference external" href="files/snps.humdiv.output">.humdiv.output</a>.</p>
</div>
<div class="section" id="poppler">
<h2>poppler<a class="headerlink" href="#poppler" title="Permalink to this headline">¶</a></h2>
<p>Official page: <a class="reference external" href="https://poppler.freedesktop.org/">https://poppler.freedesktop.org/</a>.</p>
<p>We work on the latest version, 0.84.0.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load xz/5.2.2
wget https://poppler.freedesktop.org/poppler-0.84.0.tar.xz
tar xf poppler-0.84.0.tar.xz
<span class="nb">cd</span> poppler-0.84.0
mkdir build
<span class="nb">cd</span> build
module load gcc/5
module load openjpeg-2.1-gcc-5.4.0-myd2p3o
cmake ..
make
make install
</pre></div>
</div>
<p>Note it is necessary to change prefix, cc and c++ to gcc and g++ in line with gcc/5, e.g.,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>CMAKE_INSTALL_PREFIX:PATH=/rds/user/$USER/hpc-work
CMAKE_C_COMPILER:FILEPATH=/usr/local/software/master/gcc/5/bin/gcc
CMAKE_CXX_COMPILER:FILEPATH=/usr/local/software/master/gcc/5/bin/g++
</pre></div>
</div>
<p>which could be done by editing CMakeCache.txt and/or calling <code class="docutils literal notranslate"><span class="pre">ccmake</span> <span class="pre">.</span></code>. After these we have
the followihg utilities:
pdfattach, pdfdetach,  pdffonts,  pdfimages,  pdfinfo,  pdfseparate,  pdftocairo,  pdftohtml,  pdftoppm,  pdftops,  pdftotext,  pdfunite,
and following libraries: libpoppler.so, libpoppler-cpp.so, libpoppler-glibc.so, libpoppler-qt5.so.</p>
<p>Note that <code class="docutils literal notranslate"><span class="pre">poppler-cpp.pc</span></code> is installed at <code class="docutils literal notranslate"><span class="pre">hpc-work/lib64/pkgconfig</span></code> and preferably in the search path with</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">export</span> <span class="nv">PKG_CONFIG_PATH</span><span class="o">=</span><span class="nv">$HPC_WORK</span>/lib/pkgconfig:<span class="nv">$HPC_WORK</span>/lib64/pkgconfig:<span class="nv">$PKG_CONFIG_PATH</span>
</pre></div>
</div>
<p>after which we could install R/pdftools from CRAN.</p>
<p>Note also the packages <code class="docutils literal notranslate"><span class="pre">R/Rpoppler</span></code> and <code class="docutils literal notranslate"><span class="pre">qpdf</span></code> from CRAN.</p>
</div>
<div class="section" id="prsice">
<h2>PRSice<a class="headerlink" href="#prsice" title="Permalink to this headline">¶</a></h2>
<p>Web page: https://github.com/choishingwan/PRSice and http://www.prsice.info/.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HPC_WORK</span>
git clone https://github.com/choishingwan/PRSice
<span class="nb">cd</span> PRSice
mkdir build
<span class="nb">cd</span> build
module load cmake/3.9
cmake ..
make
ln -sf <span class="nv">$HPC_WORK</span>/PRSice/bin/PRSice <span class="nv">$HPC_WORK</span>/bin/PRSice
wget https://github.com/choishingwan/PRS-Tutorial/raw/master/resources/GIANT.height.gz
wget -qO- https://github.com/choishingwan/PRS-Tutorial/raw/master/resources/EUR.zip <span class="p">|</span> jar xv
</pre></div>
</div>
<p>The last two commands download/unpack the documentation example, which is described here, https://choishingwan.github.io/PRS-Tutorial/, whose scripts are partly extracted <a class="reference external" href="files/pgs.sh">here</a>.</p>
</div>
<div class="section" id="prsos">
<h2>PRSoS<a class="headerlink" href="#prsos" title="Permalink to this headline">¶</a></h2>
<p>Web page: https://github.com/MeaneyLab/PRSoS.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git clone https://github.com/MeaneyLab/PRSoS.git
<span class="nb">cd</span> PRSoS
pip install –r requirements.txt
~/.local/bin/spark-submit --master local<span class="o">[</span>*<span class="o">]</span> PRSoS.py examples/example.vcf examples/gwasfile.txt test_output
</pre></div>
</div>
</div>
<div class="section" id="pspp">
<h2>pspp<a class="headerlink" href="#pspp" title="Permalink to this headline">¶</a></h2>
<p>Official page: <a class="reference external" href="https://www.gnu.org/software/pspp/">https://www.gnu.org/software/pspp/</a>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load pspp/1.2.0
pspp example.sps
</pre></div>
</div>
<p>where <a class="reference external" href="files/example.sps">example.sps</a> is the documentation example. Nevertheless <code class="docutils literal notranslate"><span class="pre">psppire</span></code> is not yet functioning.</p>
<p>However, it is possible to compile it directly by using</p>
<ul class="simple">
<li><p>gtksourceview 4.0.3 (4.4.0 is more demanding with Python 3.5, meson, Vala, etc.) and use PKG_CONFIG_PATH when appropriate</p></li>
<li><p>spread-sheet-widget-0.3</p></li>
<li><p>fribidi-1.0.8</p></li>
<li><p>GTKSOURVIEW_CFLAGS and GTKSOURVIEW_LIBS in the configuration.</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">export</span> <span class="nv">PREFIX</span><span class="o">=</span>/rds/user/<span class="nv">$USER</span>/hpc-work
<span class="nb">export</span> <span class="nv">GTKSOURCEVIEW_CFLAGS</span><span class="o">=</span>-I<span class="si">${</span><span class="nv">PREFIX</span><span class="si">}</span>/includegtksourceview-4
<span class="nb">export</span> <span class="nv">GTKSOURCEVIEW_LIBS</span><span class="o">=</span><span class="s2">&quot;-L</span><span class="si">${</span><span class="nv">PREFIX</span><span class="si">}</span><span class="s2">/lib -lgtksourceview-4&quot;</span>
./configure --prefix<span class="o">=</span><span class="si">${</span><span class="nv">PREFIX</span><span class="si">}</span>
make
make install
</pre></div>
</div>
<p>note that it is necessary to comment on the statement <code class="docutils literal notranslate"><span class="pre">kludge</span> <span class="pre">=</span> <span class="pre">gtk_source_view_get_type</span> <span class="pre">();</span></code> from <code class="docutils literal notranslate"><span class="pre">src/ui/gui/widgets.c</span></code>
and to remove the <code class="docutils literal notranslate"><span class="pre">PREFIX=</span></code> specification in the Perl part of compiling, i.e,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>cd perl-module
/usr/bin/perl Makefile.PL PREFIX=/rds/user/$USER/hpc-work OPTIMIZE=&quot;-g -O2 -I/rds-d4/user/$USER/hpc-work/include/fribidi -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng15 -I/usr/include/uuid -I/usr/include/libdrm -I/usr/include/pango-1.0 -I/usr/include/harfbuzz  &quot;
</pre></div>
</div>
<p>Now we can execute <a class="reference external" href="files/plot.sps">plot.sps</a></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>psppire plot.ps <span class="p">&amp;</span>
</pre></div>
</div>
<p>More documentation examples are in the <a class="reference external" href="files/examples">examples</a> directory.</p>
</div>
<div class="section" id="qpdf">
<h2>qpdf<a class="headerlink" href="#qpdf" title="Permalink to this headline">¶</a></h2>
<p>Web page: https://sourceforge.net/projects/qpdf/.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HPC_WORK</span>
wget -qO- https://sourceforge.net/projects/qpdf/files/qpdf/9.1.0/qpdf-9.1.0.tar.gz <span class="p">|</span> <span class="se">\</span>
tar xvfz -
<span class="nb">cd</span> qpdf-9.1.0
./configure --prefix<span class="o">=</span><span class="nv">$HPC_WORK</span>
make
make install
</pre></div>
</div>
</div>
<div class="section" id="r">
<h2>R<a class="headerlink" href="#r" title="Permalink to this headline">¶</a></h2>
<p>To compile all the PDF documentations, load texlive.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load gcc/6
module load pcre/8.38
module load texlive
wget https://cran.r-project.org/src/base/R-4/R-4.0.0.tar.gz
tar xvfz R-4.0.0.tar.gz
<span class="nb">cd</span> R-4.0.0
<span class="nb">export</span> <span class="nv">prefix</span><span class="o">=</span>/rds-d4/user/<span class="nv">$USER</span>/hpc-work
./configure --prefix<span class="o">=</span><span class="si">${</span><span class="nv">prefix</span><span class="si">}</span> <span class="se">\</span>
            --with-pcre1 <span class="se">\</span>
            --enable-R-shlib <span class="nv">CPPFLAGS</span><span class="o">=</span>-I<span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>/include <span class="nv">LDFLAGS</span><span class="o">=</span>-L<span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>/lib
make
make install
</pre></div>
</div>
<p>With this setup, <code class="docutils literal notranslate"><span class="pre">R</span> <span class="pre">CMD</span> <span class="pre">check</span> <span class="pre">--as-cran</span></code> for a CRAN package check can be run smoothly.</p>
<p>Package reinstallation could be done with <code class="docutils literal notranslate"><span class="pre">update.packages(checkBuilt</span> <span class="pre">=</span> <span class="pre">TRUE,</span> <span class="pre">ask</span> <span class="pre">=</span> <span class="pre">FALSE)</span></code>.</p>
</div>
<div class="section" id="r-rgdal">
<h2>R/rgdal<a class="headerlink" href="#r-rgdal" title="Permalink to this headline">¶</a></h2>
<p>We have .R/Makevars as follows,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">CC</span><span class="o">=</span>gcc
<span class="nv">CXX</span><span class="o">=</span>g++ -std<span class="o">=</span>gnu++11
<span class="nv">PKG_CXXFLAGS</span><span class="o">=</span> -std<span class="o">=</span>c++11
<span class="nv">CFLAGS</span> <span class="o">=</span> -std<span class="o">=</span>c99 -I/usr/include -g -O2 -Wall -pedantic -mtune<span class="o">=</span>native -Wno-ignored-attributes -Wno-deprecated-declarations -Wno-parentheses -Wimplicit-function-declaration
<span class="nv">CXXFLAGS</span> <span class="o">=</span> -std<span class="o">=</span>c++11
</pre></div>
</div>
<p>the proceed with</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load gcc/5
module load geos-3.6.2-gcc-5.4.0-vejexvy
module load gdal-2.3.1-gcc-5.4.0-m7j7nw6
module load proj-5.0.1-gcc-5.4.0-cpqxtzr
Rscript -e <span class="s2">&quot;install.packages(&#39;sf&#39;)&quot;</span>
<span class="c1"># R 3.6.3 also requires</span>
module load json-c-0.13.1-gcc-5.4.0-ffamohj
module load libgeotiff-1.4.2-gcc-5.4.0-2emzhxh
module load libpng-1.6.29-gcc-5.4.0-3qwhidp
module load cfitsio-3.450-gcc-5.4.0-colpo6h
module load zlib/1.2.11
module load mpich-3.2-gcc-5.4.0-idlluti
Rscript -e <span class="s2">&quot;install.packages(&#39;rgdal&#39;)&quot;</span>
</pre></div>
</div>
<p>Under R 3.6.3, there are complaints about <code class="docutils literal notranslate"><span class="pre">-std=c++11</span></code> when installing <code class="docutils literal notranslate"><span class="pre">sf</span></code> but one can gets around with</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load geos-3.6.2-gcc-5.4.0-vejexvy
module load gcc/6
Rscript -e <span class="s2">&quot;install.packages(&#39;sf&#39;)&quot;</span>
</pre></div>
</div>
<p>Finally, gdal could also be installed with proj 6 available,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load geos-3.6.2-gcc-5.4.0-vejexvy
./configure --with-proj<span class="o">=</span><span class="nv">$HPC_WORK</span> --without-sqlite3 --prefix<span class="o">=</span><span class="nv">$HPC_WORK</span>
</pre></div>
</div>
<p>Then <code class="docutils literal notranslate"><span class="pre">proj_api.h</span></code> should have a statement</p>
<div class="highlight-c notranslate"><div class="highlight"><pre><span></span><span class="cp">#define ACCEPT_USE_OF_DEPRECATED_PROJ_API_H 1</span>
</pre></div>
</div>
</div>
<div class="section" id="r-rgeos">
<h2>R/rgeos<a class="headerlink" href="#r-rgeos" title="Permalink to this headline">¶</a></h2>
<p>This requires geos to be loaded,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load geos-3.6.2-gcc-5.4.0-vejexvy
</pre></div>
</div>
</div>
<div class="section" id="r-rhdf5lib">
<h2>R/Rhdf5lib<a class="headerlink" href="#r-rhdf5lib" title="Permalink to this headline">¶</a></h2>
<p>This is useful for hdf5 file handling, but BiocManager::install() gives error <code class="docutils literal notranslate"><span class="pre">cp:</span> <span class="pre">cannot</span> <span class="pre">stat</span> <span class="pre">‘hdf5/c++/src/.libs/libhdf5_cpp.a’:</span> <span class="pre">No</span> <span class="pre">such</span> <span class="pre">file</span> <span class="pre">or</span> <span class="pre">directory</span> </code> so we proceed manually,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load gcc/6
<span class="nb">cd</span> <span class="nv">$HOME</span>
wget https://bioconductor.org/packages/3.11/bioc/src/contrib/Rhdf5lib_1.10.1.tar.gz
tar xfz Rhdf5lib_1.10.1.tar.gz
<span class="nb">cd</span> Rhdf5lib
./configure
mv configure configure.sav
<span class="nb">cd</span> src
tar xfz hdf5small_cxx_hl_1.10.6.tar.gz
<span class="nb">cd</span> hdf5
./configure --prefix<span class="o">=</span><span class="nv">$HPC_WORK</span> --enable-build-all --enable-cxx <span class="nv">CFLAGS</span><span class="o">=</span>-fPIC
make
make install
mv configure configure.sav
<span class="nb">cd</span> <span class="nv">$HOME</span>
R CMD INSTALL Rhdf5lib
</pre></div>
</div>
</div>
<div class="section" id="r-rjava">
<h2>R/rJava<a class="headerlink" href="#r-rjava" title="Permalink to this headline">¶</a></h2>
<p>One may see the messages</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>...
checking whether JNI programs can be compiled... configure: error: Cannot compile a simple JNI program. See config.log for details.
...
ERROR: configuration failed for package ‘rJava’
</pre></div>
</div>
<p>so quit R and run</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>R CMD javareconf
Rscript -e <span class="s1">&#39;install.packages(&quot;rJava&quot;)&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="r-rjags">
<h2>R/rjags<a class="headerlink" href="#r-rjags" title="Permalink to this headline">¶</a></h2>
<p>Web page: <a class="reference external" href="https://sourceforge.net/projects/mcmc-jags/files/rjags/">https://sourceforge.net/projects/mcmc-jags/files/rjags/</a>.</p>
<p>It is known for sometime for its difficulty to install; here is what was done</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Cardio</span>
<span class="nb">export</span> <span class="nv">PKG_CONFIG_PATH</span><span class="o">=</span>/scratch/<span class="nv">$USER</span>/lib/pkgconfig

R CMD INSTALL rjags_4-6.tar.gz --configure-args<span class="o">=</span><span class="s1">&#39;CPPFLAGS=&quot;-fPIC&quot; LDFLAGS=&quot;-L/scratch/$USER/lib -ljags&quot;</span>
<span class="s1">--with-jags-prefix=/scratch/$USER</span>
<span class="s1">--with-jags-libdir=/scratch/$USER/lib</span>
<span class="s1">--with-jags-includedir=/scratch/$USER/include&#39;</span>

<span class="c1"># csd3</span>
<span class="nb">export</span> <span class="nv">hpcwork</span><span class="o">=</span>/rds-d4/user/<span class="nv">$USER</span>/hpc-work
<span class="nb">export</span> <span class="nv">PKG_CONFIG_PATH</span><span class="o">=</span><span class="si">${</span><span class="nv">hpcwork</span><span class="si">}</span>/lib/pkgconfig

wget https://cran.r-project.org/src/contrib/rjags_4-10.tar.gz
R CMD INSTALL rjags_4-10.tar.gz --configure-args<span class="o">=</span><span class="s1">&#39;CPPFLAGS=&quot;-fPIC&quot; LDFLAGS=&quot;-L${hpcwork}/lib -ljags&quot;</span>
<span class="s1">--with-jags-prefix=${hpcwork}</span>
<span class="s1">--with-jags-libdir=${hpcwork}/lib</span>
<span class="s1">--with-jags-includedir=${hpcwork}/include&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="r-rstan">
<h2>R/rstan<a class="headerlink" href="#r-rstan" title="Permalink to this headline">¶</a></h2>
<p>Official page: <a class="reference external" href="https://mc-stan.org/users/interfaces/rstan">https://mc-stan.org/users/interfaces/rstan</a> and also <a class="reference external" href="https://cran.r-project.org/package=rstan">https://cran.r-project.org/package=rstan</a>.</p>
<p>It is necessary to have <code class="docutils literal notranslate"><span class="pre">¬/.R/Makevars</span></code> the following lines,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CXX14</span> <span class="o">=</span> <span class="n">g</span><span class="o">++</span> <span class="o">-</span><span class="n">fPIC</span> <span class="o">-</span><span class="n">std</span><span class="o">=</span><span class="n">gnu</span><span class="o">++</span><span class="mi">11</span> <span class="o">-</span><span class="n">fext</span><span class="o">-</span><span class="n">numeric</span><span class="o">-</span><span class="n">literals</span>
</pre></div>
</div>
<p>to deal with <code class="docutils literal notranslate"><span class="pre">error:</span> <span class="pre">unable</span> <span class="pre">to</span> <span class="pre">find</span> <span class="pre">numeric</span> <span class="pre">literal</span> <span class="pre">operator</span> <span class="pre">'operator&quot;&quot;Q'</span></code> but</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CXX14</span> <span class="o">=</span> <span class="n">g</span><span class="o">++</span> <span class="o">-</span><span class="n">std</span><span class="o">=</span><span class="n">c</span><span class="o">++</span><span class="mi">1</span><span class="n">y</span> <span class="o">-</span><span class="n">fPIC</span>
</pre></div>
</div>
<p>to do away with the error message ``C++14 standard requested but CXX14 is not defined`.</p>
<p>In case <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code> installed with <code class="docutils literal notranslate"><span class="pre">gcc</span> <span class="pre">5.2.0</span></code> it is also necessary to preceed with <code class="docutils literal notranslate"><span class="pre">module</span> <span class="pre">load</span> <span class="pre">gcc/5</span></code>.</p>
<p>For the developmental version, try <code class="docutils literal notranslate"><span class="pre">remotes::install_github(&quot;stan-dev/rstan&quot;,</span> <span class="pre">ref</span> <span class="pre">=</span> <span class="pre">&quot;develop&quot;,</span> <span class="pre">subdir</span> <span class="pre">=</span> <span class="pre">&quot;rstan/rstan&quot;)</span></code>.</p>
</div>
<div class="section" id="r-saige-0-36-6-and-0-39-2">
<h2>R/SAIGE 0.36.6 and 0.39.2<a class="headerlink" href="#r-saige-0-36-6-and-0-39-2" title="Permalink to this headline">¶</a></h2>
<p>Full name: Scalable and Accurate Implementation of GEneralized mixed model (SAIGE)</p>
<p>GitHub page: <a class="reference external" href="https://github.com/weizhouUMICH/SAIGE">https://github.com/weizhouUMICH/SAIGE</a>.</p>
<p>The following is based on source from GitHub (so with the possibility to git pull),</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load cmake/3.9 gcc/5
module load python/2.7
virtualenv py27
<span class="nb">source</span> py27/bin/activate
pip install cget
git clone https://github.com/weizhouUMICH/SAIGE
R CMD INSTALL SAIGE
</pre></div>
</div>
<p>Now we see <code class="docutils literal notranslate"><span class="pre">.../SAIGE.so:</span> <span class="pre">undefined</span> <span class="pre">symbol:</span> <span class="pre">sgecon_</span></code>. One can get away with it by renaming <code class="docutils literal notranslate"><span class="pre">configure</span></code> to <code class="docutils literal notranslate"><span class="pre">configure.sav</span></code> (so avoid repeated downloads) and amend the last <code class="docutils literal notranslate"><span class="pre">g++</span> <span class="pre">...</span> <span class="pre">-o</span> <span class="pre">SAIGE.so</span></code> with <code class="docutils literal notranslate"><span class="pre">-L$HPC_WORK/lib64</span> <span class="pre">-llapack</span></code> and then rerun <code class="docutils literal notranslate"><span class="pre">R</span> <span class="pre">CMD</span> <span class="pre">INSTALL</span> <span class="pre">SAIGE</span></code>. After successful installation, we can try <code class="docutils literal notranslate"><span class="pre">cd</span> <span class="pre">SAIGE/extdata;</span> <span class="pre">bash</span> <span class="pre">cmd.sh</span></code>.</p>
<p>One of the third party software is <code class="docutils literal notranslate"><span class="pre">bgenix</span></code> (BE careful with a buggy <code class="docutils literal notranslate"><span class="pre">cat-bgen</span></code>!), whose <code class="docutils literal notranslate"><span class="pre">wscript</span></code> uses Python 2 syntax so it is necessary to stick to python/2.7 explicitly since gcc/5 automatically loads python 3.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>cd SAIGE
cd thirdParty
cd bgen
./waf configure --prefix=$HPC_WORK
./waf
./waf install
build/test/unit/test_bgen
build/apps/bgenix -g example/example.16bits.bgen -list
cd ../../..
</pre></div>
</div>
<p>See <a class="reference external" href="https://github.com/weizhouUMICH/SAIGE/issues/98">https://github.com/weizhouUMICH/SAIGE/issues/98</a>.</p>
<p>For the latest version 0.39.2 which deals with the chromosome X ploidy, the following steps are necessary</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>R -e <span class="s2">&quot;devtools::install_github(&#39;leeshawn/MetaSKAT&#39;)&quot;</span>
R -e <span class="s2">&quot;devtools::install_github(&#39;leeshawn/SPAtest&#39;)&quot;</span>
git clone --depth <span class="m">1</span> -b <span class="m">0</span>.39.2 https://github.com/weizhouUMICH/SAIGE
R CMD INSTALL SAIGE
</pre></div>
</div>
<p>which first installs MetaSKAT 0.80 also at CRAN but SPAtest 3.1.2 instead of 3.0.2 from CRAN.</p>
<p>A by-product of this is LEMMA,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget -qO- https://github.com/mkerin/LEMMA/archive/v1.0.2.tar.gz <span class="p">|</span> <span class="se">\</span>
tar xvfz <span class="p">|</span> <span class="se">\</span>
<span class="nb">cd</span> LEMMA-1.0.2
cmake -S . -B build <span class="se">\</span>
      -DBGEN_ROOT<span class="o">=</span><span class="si">${</span><span class="nv">HOME</span><span class="si">}</span>/SAIGE/thirdParty/bgen <span class="se">\</span>
      -DBOOST_ROOT<span class="o">=</span><span class="si">${</span><span class="nv">HOME</span><span class="si">}</span>/SAIGE/thirdParty/bgen/3rd_party/boost_1_55_0
<span class="nb">cd</span> build
make
</pre></div>
</div>
</div>
<div class="section" id="smr">
<h2>SMR<a class="headerlink" href="#smr" title="Permalink to this headline">¶</a></h2>
<p>Web: https://cnsgenomics.com/software/smr/</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget https://cnsgenomics.com/software/smr/download/smr_1.03_src.zip
unzip -v smr_1.03_src.zip
mkdir smr
<span class="nb">cd</span> smr
unzip ../smr_1.03_src.zip
rm -rf __MACOSX/
mv smr_1.03_src/ ../..
<span class="nb">cd</span> ..
rmdir smr
<span class="nb">cd</span> ../smr_1.03_src/
module load eigen/latest
module load zlib/1.2.8
make
./smr_linux
</pre></div>
</div>
</div>
<div class="section" id="r-snpnet">
<h2>R/snpnet<a class="headerlink" href="#r-snpnet" title="Permalink to this headline">¶</a></h2>
<p>GitHub page: <a class="reference external" href="https://github.com/rivas-lab/snpnet">https://github.com/rivas-lab/snpnet</a>.</p>
<p>A number of software needs to be set up with the current version.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load lz4-1.8.1.2-intel-17.0.4-celw56p
wget https://github.com/facebook/zstd/releases/download/v1.4.4/zstd-1.4.4.tar.gz
tar xfz zstd-1.4.4.tar.gz
<span class="nb">cd</span> zstd-1.4.4
make
make install <span class="nv">prefix</span><span class="o">=</span><span class="nv">$HPC_WORK</span>
<span class="c1"># gcc/6 is required for pgenlibr</span>
module load gcc/6
</pre></div>
</div>
<p>File <code class="docutils literal notranslate"><span class="pre">dotCall64/Makevars</span></code> needs to be modified, but can be difficult (e.g., reinstallation of gettext, https://ftp.gnu.org/gnu/gettext/gettext-0.20.tar.gz), then</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>PKG_CFLAGS = $(SHLIB_OPENMP_CFLAGS) -I../inst/include/ -DDOTCAL64_PRIVATE -I$HPC_WORK/include
PKG_LIBS = $(SHLIB_OPENMP_CFLAGS) -L$HPC_WORK/lib -lintl
</pre></div>
</div>
<p>Finally, we can proceed</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">devtools</span><span class="o">::</span><span class="nf">install_github</span><span class="p">(</span><span class="s">&quot;junyangq/glmnetPlus&quot;</span><span class="p">)</span>
<span class="n">devtools</span><span class="o">::</span><span class="nf">install_github</span><span class="p">(</span><span class="s">&quot;chrchang/plink-ng&quot;</span><span class="p">,</span> <span class="n">subdir</span><span class="o">=</span><span class="s">&quot;/2.0/cindex&quot;</span><span class="p">)</span>
<span class="n">devtools</span><span class="o">::</span><span class="nf">install_github</span><span class="p">(</span><span class="s">&quot;chrchang/plink-ng&quot;</span><span class="p">,</span> <span class="n">subdir</span><span class="o">=</span><span class="s">&quot;/2.0/pgenlibr&quot;</span><span class="p">)</span>
<span class="n">devtools</span><span class="o">::</span><span class="nf">install_github</span><span class="p">(</span><span class="s">&quot;rivas-lab/snpnet&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="r-sojo">
<h2>R/sojo<a class="headerlink" href="#r-sojo" title="Permalink to this headline">¶</a></h2>
<p>GitHub page: <a class="reference external" href="https://github.com/zhenin/sojo">https://github.com/zhenin/sojo</a>.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;sojo&quot;</span><span class="p">,</span> <span class="n">repos</span> <span class="o">=</span> <span class="s">&quot;http://R-Forge.R-project.org&quot;</span><span class="p">)</span>
<span class="c1"># Swedish twin registry</span>
<span class="nf">download.file</span><span class="p">(</span><span class="s">&quot;https://www.dropbox.com/s/ty1udfhx5ohauh8/LD_chr22.rda?raw=1&quot;</span><span class="p">,</span> 
<span class="n">destfile</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="nf">find.package</span><span class="p">(</span><span class="s">&#39;sojo&#39;</span><span class="p">),</span> <span class="s">&quot;/LD_chr22.rda&quot;</span><span class="p">))</span>
<span class="nf">load</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="nf">find.package</span><span class="p">(</span><span class="s">&#39;sojo&#39;</span><span class="p">),</span> <span class="s">&quot;/LD_chr22.rda&quot;</span><span class="p">))</span>
<span class="c1"># 1000Genomes</span>
<span class="nf">download.file</span><span class="p">(</span><span class="s">&quot;https://data.broadinstitute.org/alkesgroup/LDSCORE/1000G_Phase3_plinkfiles.tgz&quot;</span><span class="p">,</span> 
<span class="n">destfile</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="nf">find.package</span><span class="p">(</span><span class="s">&#39;sojo&#39;</span><span class="p">),</span> <span class="s">&quot;/1000G_Phase3_plinkfiles.tgz&quot;</span><span class="p">))</span>
<span class="nf">untar</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="nf">find.package</span><span class="p">(</span><span class="s">&#39;sojo&#39;</span><span class="p">),</span> <span class="s">&quot;/1000G_Phase3_plinkfiles.tgz&quot;</span><span class="p">),</span><span class="n">exdir</span><span class="o">=</span><span class="nf">find.package</span><span class="p">(</span><span class="s">&#39;sojo&#39;</span><span class="p">))</span>
<span class="nf">require</span><span class="p">(</span><span class="n">sojo</span><span class="p">)</span>
<span class="nf">data</span><span class="p">(</span><span class="n">sum.stat.discovery</span><span class="p">)</span>
<span class="n">hpc_work</span> <span class="o">&lt;-</span> <span class="nf">Sys.getenv</span><span class="p">(</span><span class="s">&quot;HPC_WORK&quot;</span><span class="p">)</span>
<span class="n">path.plink</span> <span class="o">&lt;-</span> <span class="nf">paste0</span><span class="p">(</span><span class="n">hpc_work</span><span class="p">,</span><span class="s">&quot;/bin/plink&quot;</span><span class="p">)</span>
<span class="n">path.1kG</span> <span class="o">&lt;-</span> <span class="nf">paste0</span><span class="p">(</span><span class="nf">find.package</span><span class="p">(</span><span class="s">&#39;sojo&#39;</span><span class="p">),</span><span class="s">&quot;/1000G_EUR_Phase3_plink&quot;</span><span class="p">)</span>
<span class="n">snps</span> <span class="o">&lt;-</span> <span class="n">sum.stat.discovery</span><span class="o">$</span><span class="n">SNP</span>
<span class="nf">write.table</span><span class="p">(</span><span class="n">snps</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="n">snps</span><span class="p">[</span><span class="m">1</span><span class="p">],</span><span class="s">&quot;_snp_list.txt&quot;</span><span class="p">),</span> <span class="n">quote</span> <span class="o">=</span> <span class="bp">F</span><span class="p">,</span> <span class="n">row.names</span> <span class="o">=</span> <span class="bp">F</span><span class="p">,</span> <span class="n">col.names</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
<span class="n">chr</span> <span class="o">&lt;-</span> <span class="m">22</span>
<span class="nf">system</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">path.plink</span><span class="p">,</span><span class="s">&quot; -bfile &quot;</span><span class="p">,</span> <span class="n">path.1kG</span><span class="p">,</span><span class="s">&quot;/1000G.EUR.QC.&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot; --r square --extract &quot;</span><span class="p">,</span> <span class="n">snps</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="s">&quot;_snp_list.txt --out &quot;</span><span class="p">,</span> <span class="n">snps</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="s">&quot; --noweb&quot;</span><span class="p">))</span>
<span class="nf">system</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">path.plink</span><span class="p">,</span><span class="s">&quot; -bfile &quot;</span><span class="p">,</span> <span class="n">path.1kG</span><span class="p">,</span><span class="s">&quot;/1000G.EUR.QC.&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot; --freq --extract &quot;</span><span class="p">,</span> <span class="n">snps</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="s">&quot;_snp_list.txt --out &quot;</span><span class="p">,</span> <span class="n">snps</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="s">&quot; --noweb&quot;</span><span class="p">))</span>
<span class="n">LD_1kG</span> <span class="o">&lt;-</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="nf">read.table</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">snps</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="s">&quot;.ld&quot;</span><span class="p">)))</span>
<span class="n">maf_1kG</span> <span class="o">&lt;-</span> <span class="nf">read.table</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">snps</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="s">&quot;.frq&quot;</span><span class="p">),</span> <span class="n">header</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>
<span class="n">snp_ref_1kG</span> <span class="o">&lt;-</span> <span class="n">maf_1kG</span><span class="p">[,</span><span class="s">&quot;A2&quot;</span><span class="p">]</span>
<span class="nf">names</span><span class="p">(</span><span class="n">snp_ref_1kG</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">maf_1kG</span><span class="p">[,</span><span class="s">&quot;SNP&quot;</span><span class="p">]</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">LD_1kG</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">LD_1kG</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">maf_1kG</span><span class="o">$</span><span class="n">SNP</span>
<span class="n">res</span> <span class="o">&lt;-</span> <span class="nf">sojo</span><span class="p">(</span><span class="n">sum.stat.discovery</span><span class="p">,</span> <span class="n">LD_ref</span> <span class="o">=</span> <span class="n">LD_mat</span><span class="p">,</span> <span class="n">snp_ref</span> <span class="o">=</span> <span class="n">snp_ref</span><span class="p">,</span> <span class="n">nvar</span> <span class="o">=</span> <span class="m">20</span><span class="p">)</span>
<span class="nf">matplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">lambda.v</span><span class="p">),</span> <span class="nf">t</span><span class="p">(</span><span class="nf">as.matrix</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">beta.mat</span><span class="p">)),</span> <span class="n">lty</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">type</span> <span class="o">=</span> <span class="s">&quot;l&quot;</span><span class="p">,</span> 
    <span class="n">xlab</span> <span class="o">=</span> <span class="nf">expression</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">log</span><span class="p">,</span> <span class="s">&quot; &quot;</span><span class="p">,</span> <span class="n">lambda</span><span class="p">)),</span> <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;Coefficients&quot;</span><span class="p">,</span> <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;Summary-level LASSO&quot;</span><span class="p">)</span>
<span class="nf">data</span><span class="p">(</span><span class="n">sum.stat.validation</span><span class="p">)</span>
<span class="n">res.valid</span> <span class="o">&lt;-</span> <span class="nf">sojo</span><span class="p">(</span><span class="n">sum.stat.discovery</span><span class="p">,</span> <span class="n">sum.stat.validation</span><span class="p">,</span> <span class="n">LD_ref</span> <span class="o">=</span> <span class="n">LD_mat</span><span class="p">,</span> <span class="n">snp_ref</span> <span class="o">=</span> <span class="n">snp_ref</span><span class="p">,</span> <span class="n">nvar</span> <span class="o">=</span> <span class="m">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="vep">
<h2>VEP<a class="headerlink" href="#vep" title="Permalink to this headline">¶</a></h2>
<p>Official page: <a class="reference external" href="https://www.ensembl.org/info/docs/tools/vep/index.html">https://www.ensembl.org/info/docs/tools/vep/index.html</a>
(<a class="reference external" href="https://www.ensembl.org/Tools/VEP">web interface</a>).</p>
<p>Detailed instructions for installation are available from here,</p>
<p>http://www.ensembl.org/info/docs/tools/vep/script/vep_download.html#installer.</p>
<p>There are several possible ways to install under csd3: GitHub, R and docker.</p>
<div class="section" id="github">
<h3>— GitHub —<a class="headerlink" href="#github" title="Permalink to this headline">¶</a></h3>
<p>GitHub Page: <a class="reference external" href="https://github.com/Ensembl/ensembl-vep">https://github.com/Ensembl/ensembl-vep</a>.</p>
<p>The ease with this option lies with GitHub in that updates can simply be made with <code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">pull</span></code> to an exisiting release.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HPC_WORK</span>
git clone https://github.com/Ensembl/ensembl-vep.git
<span class="nb">cd</span> ensembl-vep
mkdir .vep
ln -sf <span class="nv">$HPC_WORK</span>/ensembl-vep/.vep <span class="nv">$HOME</span>/.vep
module load htslib/1.4
perl INSTALL.pl
<span class="c1"># set up symbolic links to the executables</span>
<span class="k">for</span> f in convert_cache.pl filter_vep haplo variant_recoder vep<span class="p">;</span> 
    <span class="k">do</span> ln -sf <span class="nv">$HPC_WORK</span>/ensembl-vep/<span class="nv">$f</span> <span class="nv">$HPC_WORK</span>/bin/<span class="nv">$f</span><span class="p">;</span> <span class="k">done</span>
vep -i examples/homo_sapiens_GRCh37.vcf -o examples/homo_sapiens_GRCh37.txt <span class="se">\</span>
    --force_overwrite --offline --pick --symbol
</pre></div>
</div>
<p>Note in particular that by default, the cache files will be installed at $HOME which would exceed the quota (&lt;40GB) of an ordinary user,
and as before the destination was redirected. The setup above facilitates storage of cache files, FASTA files and plugins.</p>
<blockquote>
<div><p>The FASTA file should be automatically detected by the VEP when using –cache or –offline.
If it is not, use “–fasta $HOME/.vep/homo_sapiens/98_GRCh37/Homo_sapiens.GRCh37.75.dna.primary_assembly.fa”</p>
</div></blockquote>
<p>Without the htslib/1.4 module, the <code class="docutils literal notranslate"><span class="pre">--NO_HTSLIB</span></code> option is needed but “Cannot use format gff without Bio::DB::HTS::Tabix module installed”.
Bio::DB:HTS is in https://github.com/Ensembl/Bio-DB-HTS and change can be made to the <code class="docutils literal notranslate"><span class="pre">Makefile</span></code> of htslibs for a desired location, to be
used by <code class="docutils literal notranslate"><span class="pre">Build.PL</span></code> via its command line parameters.</p>
<p>It is notable that VEP accepts compress (.gz) input. It is worthwhile to check for the VEP plugins: <a class="reference external" href="https://github.com/Ensembl/VEP_plugins">https://github.com/Ensembl/VEP_plugins</a>. For instance, to enable the PolyPhen_SIFT plugin we first generate the database.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vep --database --port <span class="m">3337</span> --format guess --plugin PolyPhen_SIFT,create_db<span class="o">=</span><span class="m">1</span>
</pre></div>
</div>
<p>One may wish to skip the comments (lines started with ##) in processing of the output, e.g., in R,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">export</span> <span class="nv">skips</span><span class="o">=</span><span class="k">$(</span>grep <span class="s1">&#39;##&#39;</span> examples/homo_sapiens_GRCh37.txt <span class="p">|</span> wc -l<span class="k">)</span>
R --no-save -q <span class="s">&lt;&lt;END</span>
<span class="s">  vo &lt;- read.delim(&quot;examples/homo_sapiens_GRCh37.txt&quot;,skip=as.numeric(Sys.getenv(&quot;skips&quot;)))</span>
<span class="s">  head(vo)</span>
<span class="s">END</span>
</pre></div>
</div>
<p>allowing for variable number of lines given various command-line options to be skipped to have</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>  <span class="n">X</span><span class="o">.</span><span class="n">Uploaded_variation</span>    <span class="n">Location</span> <span class="n">Allele</span>            <span class="n">Gene</span>         <span class="n">Feature</span>
<span class="mi">1</span>          <span class="n">rs116645811</span> <span class="mi">21</span><span class="p">:</span><span class="mi">26960070</span>      <span class="n">A</span> <span class="n">ENSG00000154719</span> <span class="n">ENST00000307301</span>
<span class="mi">2</span>            <span class="n">rs1135638</span> <span class="mi">21</span><span class="p">:</span><span class="mi">26965148</span>      <span class="n">A</span> <span class="n">ENSG00000154719</span> <span class="n">ENST00000307301</span>
<span class="mi">3</span>              <span class="n">rs10576</span> <span class="mi">21</span><span class="p">:</span><span class="mi">26965172</span>      <span class="n">C</span> <span class="n">ENSG00000154719</span> <span class="n">ENST00000307301</span>
<span class="mi">4</span>            <span class="n">rs1057885</span> <span class="mi">21</span><span class="p">:</span><span class="mi">26965205</span>      <span class="n">C</span> <span class="n">ENSG00000154719</span> <span class="n">ENST00000307301</span>
<span class="mi">5</span>          <span class="n">rs116331755</span> <span class="mi">21</span><span class="p">:</span><span class="mi">26976144</span>      <span class="n">G</span> <span class="n">ENSG00000154719</span> <span class="n">ENST00000307301</span>
<span class="mi">6</span>            <span class="n">rs7278168</span> <span class="mi">21</span><span class="p">:</span><span class="mi">26976222</span>      <span class="n">T</span> <span class="n">ENSG00000154719</span> <span class="n">ENST00000307301</span>
  <span class="n">Feature_type</span>        <span class="n">Consequence</span> <span class="n">cDNA_position</span> <span class="n">CDS_position</span> <span class="n">Protein_position</span>
<span class="mi">1</span>   <span class="n">Transcript</span>   <span class="n">missense_variant</span>          <span class="mi">1043</span>         <span class="mi">1001</span>              <span class="mi">334</span>
<span class="mi">2</span>   <span class="n">Transcript</span> <span class="n">synonymous_variant</span>           <span class="mi">939</span>          <span class="mi">897</span>              <span class="mi">299</span>
<span class="mi">3</span>   <span class="n">Transcript</span> <span class="n">synonymous_variant</span>           <span class="mi">915</span>          <span class="mi">873</span>              <span class="mi">291</span>
<span class="mi">4</span>   <span class="n">Transcript</span> <span class="n">synonymous_variant</span>           <span class="mi">882</span>          <span class="mi">840</span>              <span class="mi">280</span>
<span class="mi">5</span>   <span class="n">Transcript</span> <span class="n">synonymous_variant</span>           <span class="mi">426</span>          <span class="mi">384</span>              <span class="mi">128</span>
<span class="mi">6</span>   <span class="n">Transcript</span> <span class="n">synonymous_variant</span>           <span class="mi">348</span>          <span class="mi">306</span>              <span class="mi">102</span>
  <span class="n">Amino_acids</span>  <span class="n">Codons</span> <span class="n">Existing_variation</span>
<span class="mi">1</span>         <span class="n">T</span><span class="o">/</span><span class="n">M</span> <span class="n">aCg</span><span class="o">/</span><span class="n">aTg</span>                  <span class="o">-</span>
<span class="mi">2</span>           <span class="n">G</span> <span class="n">ggC</span><span class="o">/</span><span class="n">ggT</span>                  <span class="o">-</span>
<span class="mi">3</span>           <span class="n">P</span> <span class="n">ccA</span><span class="o">/</span><span class="n">ccG</span>                  <span class="o">-</span>
<span class="mi">4</span>           <span class="n">V</span> <span class="n">gtA</span><span class="o">/</span><span class="n">gtG</span>                  <span class="o">-</span>
<span class="mi">5</span>           <span class="n">L</span> <span class="n">ctT</span><span class="o">/</span><span class="n">ctC</span>                  <span class="o">-</span>
<span class="mi">6</span>           <span class="n">K</span> <span class="n">aaG</span><span class="o">/</span><span class="n">aaA</span>                  <span class="o">-</span>
                                                                     <span class="n">Extra</span>
<span class="mi">1</span> <span class="n">IMPACT</span><span class="o">=</span><span class="n">MODERATE</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=-</span><span class="mi">1</span><span class="p">;</span><span class="n">SYMBOL</span><span class="o">=</span><span class="n">MRPL39</span><span class="p">;</span><span class="n">SYMBOL_SOURCE</span><span class="o">=</span><span class="n">HGNC</span><span class="p">;</span><span class="n">HGNC_ID</span><span class="o">=</span><span class="mi">14027</span>
<span class="mi">2</span>      <span class="n">IMPACT</span><span class="o">=</span><span class="n">LOW</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=-</span><span class="mi">1</span><span class="p">;</span><span class="n">SYMBOL</span><span class="o">=</span><span class="n">MRPL39</span><span class="p">;</span><span class="n">SYMBOL_SOURCE</span><span class="o">=</span><span class="n">HGNC</span><span class="p">;</span><span class="n">HGNC_ID</span><span class="o">=</span><span class="mi">14027</span>
<span class="mi">3</span>      <span class="n">IMPACT</span><span class="o">=</span><span class="n">LOW</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=-</span><span class="mi">1</span><span class="p">;</span><span class="n">SYMBOL</span><span class="o">=</span><span class="n">MRPL39</span><span class="p">;</span><span class="n">SYMBOL_SOURCE</span><span class="o">=</span><span class="n">HGNC</span><span class="p">;</span><span class="n">HGNC_ID</span><span class="o">=</span><span class="mi">14027</span>
<span class="mi">4</span>      <span class="n">IMPACT</span><span class="o">=</span><span class="n">LOW</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=-</span><span class="mi">1</span><span class="p">;</span><span class="n">SYMBOL</span><span class="o">=</span><span class="n">MRPL39</span><span class="p">;</span><span class="n">SYMBOL_SOURCE</span><span class="o">=</span><span class="n">HGNC</span><span class="p">;</span><span class="n">HGNC_ID</span><span class="o">=</span><span class="mi">14027</span>
<span class="mi">5</span>      <span class="n">IMPACT</span><span class="o">=</span><span class="n">LOW</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=-</span><span class="mi">1</span><span class="p">;</span><span class="n">SYMBOL</span><span class="o">=</span><span class="n">MRPL39</span><span class="p">;</span><span class="n">SYMBOL_SOURCE</span><span class="o">=</span><span class="n">HGNC</span><span class="p">;</span><span class="n">HGNC_ID</span><span class="o">=</span><span class="mi">14027</span>
<span class="mi">6</span>      <span class="n">IMPACT</span><span class="o">=</span><span class="n">LOW</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=-</span><span class="mi">1</span><span class="p">;</span><span class="n">SYMBOL</span><span class="o">=</span><span class="n">MRPL39</span><span class="p">;</span><span class="n">SYMBOL_SOURCE</span><span class="o">=</span><span class="n">HGNC</span><span class="p">;</span><span class="n">HGNC_ID</span><span class="o">=</span><span class="mi">14027</span>
<span class="o">&gt;</span>
</pre></div>
</div>
<p>When a particular release really works well on the system, it is possible to install it, e.g.,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">git</span> <span class="n">checkout</span> <span class="n">release</span><span class="o">/</span><span class="mi">98</span>
<span class="n">perl</span> <span class="n">INSTALL</span><span class="o">.</span><span class="n">pl</span>
</pre></div>
</div>
<p>for release 98 instead of the latest from <code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">pull</span></code>.</p>
<p>It could be useful to filter VEP output, see https://www.ensembl.org/info/docs/tools/vep/script/vep_filter.html.</p>
</div>
<div class="section" id="nearest-gene">
<h3><strong>Nearest gene</strong><a class="headerlink" href="#nearest-gene" title="Permalink to this headline">¶</a></h3>
<p>This can produces error message</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">--------------------</span> <span class="n">EXCEPTION</span> <span class="o">--------------------</span>
<span class="n">MSG</span><span class="p">:</span> <span class="n">ERROR</span><span class="p">:</span> <span class="o">--</span><span class="n">nearest</span> <span class="n">requires</span> <span class="n">Set</span><span class="p">::</span><span class="n">IntervalTree</span> <span class="n">perl</span> <span class="n">module</span> <span class="n">to</span> <span class="n">be</span> <span class="n">installed</span>
</pre></div>
</div>
<p>and we get around with</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl -MCPAN -e shell
install Set::IntervalTree
</pre></div>
</div>
<p>which will enable <code class="docutils literal notranslate"><span class="pre">--nearest</span> <span class="pre">gene</span></code>.</p>
</div>
<div class="section" id="annotation-in-chunks">
<h3><strong>Annotation in chunks</strong><a class="headerlink" href="#annotation-in-chunks" title="Permalink to this headline">¶</a></h3>
<p>A toy example, following http://www.ensembl.org/info/docs/tools/vep/script/vep_other.html#faster, is given as follows,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> examples
bgzip -f homo_sapiens_GRCh37.vcf
tabix -Cf homo_sapiens_GRCh37.vcf.gz
tabix -h  homo_sapiens_GRCh37.vcf.gz <span class="m">22</span>:50616005-50616006 <span class="p">|</span> <span class="se">\</span>
vep --cache --fork <span class="m">4</span> --port <span class="m">3337</span> --format vcf -o - --tab --no_stats <span class="p">|</span> <span class="se">\</span>
grep -v <span class="s1">&#39;##&#39;</span>
<span class="nb">cd</span> -
</pre></div>
</div>
<p>from this we could propagate the idea for autosomes in chunks as follows,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">export</span> <span class="nv">chunk_size</span><span class="o">=</span><span class="m">10000</span>
seq <span class="m">22</span> <span class="p">|</span> <span class="se">\</span>
parallel -j1 --env ref -C<span class="s1">&#39; &#39;</span> <span class="s1">&#39;</span>
<span class="s1">  export n=$(awk &quot;END{print NR-1}&quot; ${ref}/impute_{}_interval.snpstats)</span>
<span class="s1">  export g=$(expr ${n} / ${chunk_size})</span>
<span class="s1">  export s=$(expr $n - $(( $g * $chunk_size)))</span>
<span class="s1">  (</span>
<span class="s1">    for i in $(seq ${g}); do</span>
<span class="s1">      (</span>
<span class="s1">        awk &quot;BEGIN{print \&quot;##fileformat=VCFv4.0\&quot;}&quot;</span>
<span class="s1">        awk -vOFS=&quot;\t&quot; &quot;BEGIN{print \&quot;#CHROM\&quot;,\&quot;POS\&quot;,\&quot;ID\&quot;,\&quot;REF\&quot;,\&quot;ALT\&quot;,\&quot;QUAL\&quot;,\&quot;FILTER\&quot;,\&quot;INFO\&quot;}&quot;</span>
<span class="s1">        (</span>
<span class="s1">          sed &quot;1d&quot; ${ref}/impute_{}_interval.snpstats | \</span>
<span class="s1">          awk -v i=${i} -v chunk_size=${chunk_size} -v OFS=&quot;\t&quot; &quot;NR==(i-1)*chunk_size+1,NR==i*chunk_size {</span>
<span class="s1">              if(\$1==\&quot;.\&quot;) \$1=\$3+0 \&quot;:\&quot; \$4 \&quot;_\&quot; \$5 \&quot;/\&quot; \$6; print \$3+0,\$4,\$1,\$5,\$6,\&quot;.\&quot;,\&quot;.\&quot;,\$19}&quot;</span>
<span class="s1">          if [ ${s} -gt 0 ] &amp;&amp;  [ ${i} -eq ${g} ]; then</span>
<span class="s1">             sed &quot;1d&quot; ${ref}/impute_{}_interval.snpstats | \</span>
<span class="s1">             awk -v i=${i} -v chunk_size=${chunk_size} -v OFS=&quot;\t&quot; -v n=${n} &quot;NR==i*chunk_size+1,NR==n {</span>
<span class="s1">                 if(\$1==\&quot;.\&quot;) \$1=\$3+0 \&quot;:\&quot; \$4 \&quot;_\&quot; \$5 \&quot;/\&quot; \$6; print \$3+0,\$4,\$1,\$5,\$6,\&quot;.\&quot;,\&quot;.\&quot;,\$19}&quot;</span>
<span class="s1">          fi</span>
<span class="s1">        )</span>
<span class="s1">      ) | \</span>
<span class="s1">      vep  --cache --offline --format vcf -o - --tab --pick --no_stats  \</span>
<span class="s1">           --species homo_sapiens --assembly GRCh37 --port 3337 | \</span>
<span class="s1">      (</span>
<span class="s1">        if [ ${i} -eq 1 ]; then cat; else grep -v &quot;#&quot;; fi</span>
<span class="s1">      )</span>
<span class="s1">    done</span>
<span class="s1">  ) | \</span>
<span class="s1">  gzip -f &gt; work/INTERVAL-{}.vep.gz</span>
<span class="s1">&#39;</span>
</pre></div>
</div>
<p>Note we use information from <code class="docutils literal notranslate"><span class="pre">.snpstats</span></code> files at location <code class="docutils literal notranslate"><span class="pre">ref</span></code> to build input in vcf format on the fly and feed into VEP.</p>
</div>
<div class="section" id="ensembl-synonym-translation">
<h3><strong>ENSEMBL-synonym translation</strong><a class="headerlink" href="#ensembl-synonym-translation" title="Permalink to this headline">¶</a></h3>
<p>The ENSEMBL-synonym translation is useful to check for the feature types – in the case of ENSG00000160712 (IL6R)
we found ENST00000368485 and ENST00000515190, we do</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget http://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/ensemblToGeneName.txt.gz
zgrep -e ENST00000368485 -e ENST00000515190 ensemblToGeneName.txt.gz
</pre></div>
</div>
<p>giving</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ENST00000368485</span> <span class="n">IL6R</span>
<span class="n">ENST00000515190</span> <span class="n">IL6R</span>
</pre></div>
</div>
<p>though this could also be furnished with R/biomaRt as follows,</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">biomaRt</span><span class="p">)</span>
<span class="n">ensembl</span> <span class="o">&lt;-</span> <span class="nf">useMart</span><span class="p">(</span><span class="s">&quot;ensembl&quot;</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="s">&quot;grch37.ensembl.org&quot;</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="s">&quot;/biomart/martservice&quot;</span><span class="p">)</span>
<span class="n">attr</span> <span class="o">&lt;-</span> <span class="nf">listAttributes</span><span class="p">(</span><span class="n">ensembl</span><span class="p">)</span>
<span class="n">g</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;ensembl_gene_id&#39;</span><span class="p">,</span> <span class="s">&#39;chromosome_name&#39;</span><span class="p">,</span> <span class="s">&#39;start_position&#39;</span><span class="p">,</span> <span class="s">&#39;end_position&#39;</span><span class="p">,</span> <span class="s">&#39;description&#39;</span><span class="p">,</span> <span class="s">&#39;hgnc_symbol&#39;</span><span class="p">)</span>
<span class="n">t</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;ensembl_transcript_id&#39;</span><span class="p">,</span> <span class="s">&#39;transcription_start_site&#39;</span><span class="p">,</span> <span class="s">&#39;transcript_start&#39;</span><span class="p">,</span> <span class="s">&#39;transcript_end&#39;</span><span class="p">)</span>
<span class="n">u</span> <span class="o">&lt;-</span> <span class="s">&quot;uniprotswissprot&quot;</span>
<span class="n">gtu</span> <span class="o">&lt;-</span> <span class="nf">getBM</span><span class="p">(</span><span class="n">attributes</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="n">g</span><span class="p">,</span><span class="n">t</span><span class="p">,</span><span class="n">u</span><span class="p">),</span> <span class="n">mart</span> <span class="o">=</span> <span class="n">ensembl</span><span class="p">)</span>
</pre></div>
</div>
<p>For ENSEMBL genes, R/grex is likely to work though it was developed for other purpose, e.g.,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>R -q -e <span class="s2">&quot;grex::grex(\&quot;ENSG00000160712\&quot;)&quot;</span>
</pre></div>
</div>
<p>giving</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>       <span class="n">ensembl_id</span> <span class="n">entrez_id</span> <span class="n">hgnc_symbol</span>              <span class="n">hgnc_name</span> <span class="n">cyto_loc</span>
<span class="mi">1</span> <span class="n">ENSG00000160712</span>      <span class="mi">3570</span>        <span class="n">IL6R</span> <span class="n">interleukin</span> <span class="mi">6</span> <span class="n">receptor</span>   <span class="mi">1</span><span class="n">q21</span><span class="o">.</span><span class="mi">3</span>
  <span class="n">uniprot_id</span>   <span class="n">gene_biotype</span>
<span class="mi">1</span>     <span class="n">A0N0L5</span> <span class="n">protein_coding</span>
</pre></div>
</div>
<p>The aligment of ENSG, ENST, ENSP is <code class="docutils literal notranslate"><span class="pre">ensGtp.txt.gz</span></code> at the same UCSC directory above.</p>
<p>Other possibilities include UCSC/ensembl MySQL server, e.g.,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mysql  --user<span class="o">=</span>genome --host<span class="o">=</span>genome-mysql.cse.ucsc.edu -A -D hg19 <span class="s">&lt;&lt;END</span>
<span class="s">select distinct G.gene,N.value from ensGtp as G, ensemblToGeneName as N where </span>
<span class="s">   G.transcript=N.name and</span>
<span class="s">   G.gene in (&quot;ENSG00000160712&quot;);</span>
<span class="s">END</span>
</pre></div>
</div>
<p>and</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mysql -h ensembldb.ensembl.org --port <span class="m">5306</span>  -u anonymous -D homo_sapiens_core_64_37 -A <span class="s">&lt;&lt;END</span>
<span class="s">select distinct</span>
<span class="s">   G.stable_id,</span>
<span class="s">   S.synonym</span>
<span class="s">from</span>
<span class="s">  gene_stable_id as G,</span>
<span class="s">  object_xref as OX,</span>
<span class="s">  external_synonym as S,</span>
<span class="s">  xref as X ,</span>
<span class="s">  external_db as D</span>
<span class="s">where</span>
<span class="s">  D.external_db_id=X.external_db_id and</span>
<span class="s">  X.xref_id=S.xref_id and</span>
<span class="s">  OX.xref_id=X.xref_id and</span>
<span class="s">  OX.ensembl_object_type=&quot;Gene&quot; and</span>
<span class="s">  G.gene_id=OX.ensembl_id and</span>
<span class="s">  G.stable_id in (&quot;ENSG00000160712&quot;);</span>
<span class="s">END</span>
</pre></div>
</div>
<p>according to https://www.biostars.org/p/14367/.</p>
<p>Perhaps it is the easiest to use <code class="docutils literal notranslate"><span class="pre">gprofiler2</span></code>, i.e.,</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">gprofiler2</span><span class="o">::</span><span class="nf">gconvert</span><span class="p">(</span><span class="s">&quot;ENSG00000164761&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>giving</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>  <span class="n">input_number</span> <span class="nb">input</span> <span class="n">target_number</span>          <span class="n">target</span> <span class="n">name</span>                                               <span class="n">description</span>                           <span class="n">namespace</span>
<span class="mi">1</span>            <span class="mi">1</span>  <span class="n">IL6R</span>           <span class="mf">1.1</span> <span class="n">ENSG00000160712</span> <span class="n">IL6R</span> <span class="n">interleukin</span> <span class="mi">6</span> <span class="n">receptor</span> <span class="p">[</span><span class="n">Source</span><span class="p">:</span><span class="n">HGNC</span> <span class="n">Symbol</span><span class="p">;</span><span class="n">Acc</span><span class="p">:</span><span class="n">HGNC</span><span class="p">:</span><span class="mi">6019</span><span class="p">]</span> <span class="n">ENTREZGENE</span><span class="p">,</span><span class="n">HGNC</span><span class="p">,</span><span class="n">UNIPROT_GN</span><span class="p">,</span><span class="n">WIKIGENE</span>
</pre></div>
</div>
</div>
<div class="section" id="clinvar">
<h3><strong>clinvar</strong><a class="headerlink" href="#clinvar" title="Permalink to this headline">¶</a></h3>
<p>The local installation enables considerable flexibilty, and the following example, using GRCh37 assembly, is based on
<a class="reference external" href="https://www.ensembl.org/info/docs/tools/vep/script/vep_custom.html#custom_options">https://www.ensembl.org/info/docs/tools/vep/script/vep_custom.html#custom_options</a>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compressed VCF file</span>
curl ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh37/clinvar.vcf.gz -o clinvar_GRCh37.vcf.gz
<span class="c1"># Index file</span>
curl ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh37/clinvar.vcf.gz.tbi -o clinvar_GRCh37.vcf.gz.tbi
</pre></div>
</div>
<p>Information is gathered from the header of the VCF file,</p>
<p>ClinVar Variation ID | Description
———————|———————————————————————————————————–
AF_ESP | allele frequencies from GO-ESP
AF_EXAC | allele frequencies from ExAC
AF_TGP | allele frequencies from TGP
ALLELEID | the ClinVar Allele ID
CLNDN | ClinVar’s preferred disease name for the concept specified by disease identifiers in CLNDISDB
CLNDNINCL | For included Variant : ClinVar’s preferred disease name for the concept specified by disease identifiers in CLNDISDB
CLNDISDB | Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN
CLNDISDBINCL | For included Variant: Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN
CLNHGVS | Top-level (primary assembly, alt, or patch) HGVS expression.
CLNREVSTAT | ClinVar review status for the Variation ID
CLNSIG | Clinical significance for this single variant
CLNSIGCONF | Conflicting clinical significance for this single variant
CLNSIGINCL | Clinical significance for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:clinical significance.
CLNVC | Variant type
CLNVCSO | Sequence Ontology id for variant type
CLNVI | the variant’s clinical sources reported as tag-value pairs of database and variant identifier
DBVARID | nsv accessions from dbVar for the variant
GENEINFO | Gene(s) for the variant reported as gene symbol:gene id. The gene symbol and id are delimited by a colon (:) and each pair is delimited by a vertical bar (|)
MC | comma separated list of molecular consequence in the form of Sequence Ontology ID|molecular_consequence
ORIGIN | Allele origin. One or more of the following values may be added: 0 - unknown; 1 - germline; 2 - somatic ; 4 - inherited; 8 - paternal; 16 - maternal; 32 - de-novo; 64 - biparental; 128 - uniparental; 256 - not-tested; 512 - tested-inconclusive; 1073741824 - other
RS | dbSNP ID (i.e. rs number)
SSR | Variant Suspect Reason Codes. One or more of the following values may be added: 0 - unspecified, 1 - Paralog, 2 - byEST, 4 - oldAlign, 8 - Para_EST, 16 - 1kg_failed, 1024 - other</p>
<p>We now query rs2228145,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vep --id <span class="s2">&quot;1 154426970 154426970 A/C 1&quot;</span> --species homo_sapiens -o rs2228145 --cache --offline --force_overwrite <span class="se">\</span>
    --assembly GRCh37 --custom clinvar_GRCh37.vcf.gz,ClinVar,vcf,exact,0,CLNSIG,CLNREVSTAT,CLNDN
</pre></div>
</div>
<p>which gives</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#Uploaded_variation	Location	Allele	Gene	Feature	Feature_type	Consequence	cDNA_position	CDS_position	Protein_position	Amino_acids	Codons	Existing_variation	Extra</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="mi">1</span><span class="p">:</span><span class="mi">154426970</span>	<span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">ENST00000344086</span>	<span class="n">Transcript</span>	<span class="n">intron_variant</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="n">IMPACT</span><span class="o">=</span><span class="n">MODIFIER</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=</span><span class="mi">1</span><span class="p">;</span><span class="n">ClinVar</span><span class="o">=</span><span class="mi">14660</span><span class="p">;</span><span class="n">ClinVar_CLNDN</span><span class="o">=</span><span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="p">;</span><span class="n">ClinVar_CLNREVSTAT</span><span class="o">=</span><span class="n">no_assertion_criteria_provided</span><span class="p">;</span><span class="n">ClinVar_CLNSIG</span><span class="o">=</span><span class="n">association</span><span class="p">;</span><span class="n">ClinVar_FILTER</span><span class="o">=.</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="mi">1</span><span class="p">:</span><span class="mi">154426970</span>	<span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">ENST00000368485</span>	<span class="n">Transcript</span>	<span class="n">missense_variant</span>	<span class="mi">1510</span>	<span class="mi">1073</span>	<span class="mi">358</span>	<span class="n">D</span><span class="o">/</span><span class="n">A</span>	<span class="n">gAt</span><span class="o">/</span><span class="n">gCt</span>	<span class="o">-</span>	<span class="n">IMPACT</span><span class="o">=</span><span class="n">MODERATE</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=</span><span class="mi">1</span><span class="p">;</span><span class="n">ClinVar</span><span class="o">=</span><span class="mi">14660</span><span class="p">;</span><span class="n">ClinVar_CLNDN</span><span class="o">=</span><span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="p">;</span><span class="n">ClinVar_CLNREVSTAT</span><span class="o">=</span><span class="n">no_assertion_criteria_provided</span><span class="p">;</span><span class="n">ClinVar_CLNSIG</span><span class="o">=</span><span class="n">association</span><span class="p">;</span><span class="n">ClinVar_FILTER</span><span class="o">=.</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="mi">1</span><span class="p">:</span><span class="mi">154426970</span>	<span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">ENST00000476006</span>	<span class="n">Transcript</span>	<span class="n">downstream_gene_variant</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="n">IMPACT</span><span class="o">=</span><span class="n">MODIFIER</span><span class="p">;</span><span class="n">DISTANCE</span><span class="o">=</span><span class="mi">4515</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=</span><span class="mi">1</span><span class="p">;</span><span class="n">FLAGS</span><span class="o">=</span><span class="n">cds_start_NF</span><span class="p">,</span><span class="n">cds_end_NF</span><span class="p">;</span><span class="n">ClinVar</span><span class="o">=</span><span class="mi">14660</span><span class="p">;</span><span class="n">ClinVar_CLNDN</span><span class="o">=</span><span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="p">;</span><span class="n">ClinVar_CLNREVSTAT</span><span class="o">=</span><span class="n">no_assertion_criteria_provided</span><span class="p">;</span><span class="n">ClinVar_CLNSIG</span><span class="o">=</span><span class="n">association</span><span class="p">;</span><span class="n">ClinVar_FILTER</span><span class="o">=.</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="mi">1</span><span class="p">:</span><span class="mi">154426970</span>	<span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">ENST00000502679</span>	<span class="n">Transcript</span>	<span class="n">non_coding_transcript_exon_variant</span>	<span class="mi">386</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="n">IMPACT</span><span class="o">=</span><span class="n">MODIFIER</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=</span><span class="mi">1</span><span class="p">;</span><span class="n">ClinVar</span><span class="o">=</span><span class="mi">14660</span><span class="p">;</span><span class="n">ClinVar_CLNDN</span><span class="o">=</span><span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="p">;</span><span class="n">ClinVar_CLNREVSTAT</span><span class="o">=</span><span class="n">no_assertion_criteria_provided</span><span class="p">;</span><span class="n">ClinVar_CLNSIG</span><span class="o">=</span><span class="n">association</span><span class="p">;</span><span class="n">ClinVar_FILTER</span><span class="o">=.</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="mi">1</span><span class="p">:</span><span class="mi">154426970</span>	<span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">ENST00000507256</span>	<span class="n">Transcript</span>	<span class="n">non_coding_transcript_exon_variant</span>	<span class="mi">271</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="o">-</span>	<span class="n">IMPACT</span><span class="o">=</span><span class="n">MODIFIER</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=</span><span class="mi">1</span><span class="p">;</span><span class="n">ClinVar</span><span class="o">=</span><span class="mi">14660</span><span class="p">;</span><span class="n">ClinVar_CLNDN</span><span class="o">=</span><span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="p">;</span><span class="n">ClinVar_CLNREVSTAT</span><span class="o">=</span><span class="n">no_assertion_criteria_provided</span><span class="p">;</span><span class="n">ClinVar_CLNSIG</span><span class="o">=</span><span class="n">association</span><span class="p">;</span><span class="n">ClinVar_FILTER</span><span class="o">=.</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="mi">1</span><span class="p">:</span><span class="mi">154426970</span>	<span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">ENST00000515190</span>	<span class="n">Transcript</span>	<span class="n">missense_variant</span>	<span class="mi">481</span>	<span class="mi">482</span>	<span class="mi">161</span>	<span class="n">D</span><span class="o">/</span><span class="n">A</span>	<span class="n">gAt</span><span class="o">/</span><span class="n">gCt</span>	<span class="o">-</span>	<span class="n">IMPACT</span><span class="o">=</span><span class="n">MODERATE</span><span class="p">;</span><span class="n">STRAND</span><span class="o">=</span><span class="mi">1</span><span class="p">;</span><span class="n">FLAGS</span><span class="o">=</span><span class="n">cds_start_NF</span><span class="p">,</span><span class="n">cds_end_NF</span><span class="p">;</span><span class="n">ClinVar</span><span class="o">=</span><span class="mi">14660</span><span class="p">;</span><span class="n">ClinVar_CLNDN</span><span class="o">=</span><span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="p">;</span><span class="n">ClinVar_CLNREVSTAT</span><span class="o">=</span><span class="n">no_assertion_criteria_provided</span><span class="p">;</span><span class="n">ClinVar_CLNSIG</span><span class="o">=</span><span class="n">association</span><span class="p">;</span><span class="n">ClinVar_FILTER</span><span class="o">=.</span>
</pre></div>
</div>
<p>A <a class="reference external" href="files/rs2228145_summary.html">HTML summary</a> (somehow the web browser may not display the embedded figures) is also available. The <code class="docutils literal notranslate"><span class="pre">Extra</span></code> column looks clumsy and we could add the <code class="docutils literal notranslate"><span class="pre">--tab</span></code> option to generate a tab-delimited output.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vep --id <span class="s2">&quot;1 154426970 154426970 A/C 1&quot;</span> --species homo_sapiens -o rs2228145 --cache --offline --force_overwrite <span class="se">\</span>
    --custom clinvar_GRCh37.vcf.gz,ClinVar,vcf,exact,0,CLNSIG,CLNREVSTAT,CLNDN --tab <span class="se">\</span>
    --fields Uploaded_variation,Gene,Consequence,ClinVar_CLNSIG,ClinVar_CLNREVSTAT,ClinVar_CLNDN
</pre></div>
</div>
<p>to give neatly</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#Uploaded_variation	Gene	Consequence	ClinVar_CLNSIG	ClinVar_CLNREVSTAT	ClinVar_CLNDN</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">intron_variant</span>	<span class="n">association</span>	<span class="n">no_assertion_criteria_provided</span>	<span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">missense_variant</span>	<span class="n">association</span>	<span class="n">no_assertion_criteria_provided</span>	<span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">downstream_gene_variant</span>	<span class="n">association</span>	<span class="n">no_assertion_criteria_provided</span>	<span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">non_coding_transcript_exon_variant</span>	<span class="n">association</span>	<span class="n">no_assertion_criteria_provided</span>	<span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">non_coding_transcript_exon_variant</span>	<span class="n">association</span>	<span class="n">no_assertion_criteria_provided</span>	<span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span>
<span class="mi">1_154426970</span><span class="n">_A</span><span class="o">/</span><span class="n">C</span>	<span class="n">ENSG00000160712</span>	<span class="n">missense_variant</span>	<span class="n">association</span>	<span class="n">no_assertion_criteria_provided</span>	<span class="n">Interleukin_6</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span><span class="o">|</span><span class="n">Soluble_interleukin</span><span class="o">-</span><span class="mi">6</span><span class="n">_receptor</span><span class="p">,</span><span class="n">_serum_level_of</span><span class="p">,</span><span class="n">_quantitative_trait_locus</span>
</pre></div>
</div>
</div>
<div class="section" id="dbnsfp">
<h3><strong>dbNSFP</strong><a class="headerlink" href="#dbnsfp" title="Permalink to this headline">¶</a></h3>
<p>Web page: <a class="reference external" href="https://sites.google.com/site/jpopgen/dbNSFP">https://sites.google.com/site/jpopgen/dbNSFP</a>.</p>
<p>This is set up as follows,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget ftp://dbnsfp:dbnsfp@dbnsfp.softgenetics.com/dbNSFP4.1a.zip
unzip dbNSFP4.1a.zip -d dbNSFP4.1a
<span class="nb">cd</span> dbNSFP4.1a
<span class="c1"># code for 4.0a</span>
<span class="c1"># zcat dbNSFP4.0a_variant.chr1.gz | head -n1 &gt; h</span>
<span class="c1"># zgrep -h -v ^#chr dbNSFP4.0a_variant.chr* | sort -k1,1 -k2,2n - | cat h - | bgzip -c &gt; dbNSFP4.0a.gz</span>
<span class="c1"># efficient version</span>
<span class="o">(</span>
  zcat dbNSFP4.1a_variant.chr1.gz <span class="p">|</span> head -n1
  <span class="nb">export</span> <span class="nv">prefix</span><span class="o">=</span>dbNSFP4.1a_variant.chr
  <span class="nb">echo</span> <span class="si">${</span><span class="nv">prefix</span><span class="si">}</span><span class="o">{</span><span class="m">1</span>..22<span class="o">}</span>.gz <span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>M.gz <span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>X.gz <span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>Y.gz <span class="p">|</span> <span class="se">\</span>
  xargs -I <span class="o">{}</span> bash -c <span class="s2">&quot;zcat {} | sed &#39;1d&#39;&quot;</span>
<span class="o">)</span> <span class="p">|</span> bgzip -c &gt; dbNSFP4.1a.gz
tabix -s <span class="m">1</span> -b <span class="m">2</span> -e <span class="m">2</span> dbNSFP4.1a.gz
<span class="nb">cd</span> -
vep -i examples/homo_sapiens_GRCh37.vcf -o <span class="nb">test</span> --cache --force_overwrite --offline <span class="se">\</span>
    --plugin dbNSFP,dbNSFP4.1a/dbNSFP4.1a.gz,LRT_score,FATHMM_score,MutationTaster_score
</pre></div>
</div>
<p>Since this release is frozen on Ensembl 94’s transcript set, one may prefer to use it independently via its Java programs, e.g.,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>java -jar search_dbNSFP41a.jar -i tryhg19.in -o tryhg19.out -v hg19
java -jar search_dbNSFP41a.jar -i tryhg38.in -o tryhg38.out
</pre></div>
</div>
</div>
<div class="section" id="loftee">
<h3>— loftee —<a class="headerlink" href="#loftee" title="Permalink to this headline">¶</a></h3>
<p>GitHub page: <a class="reference external" href="https://github.com/konradjk/loftee">https://github.com/konradjk/loftee</a>.</p>
<p>Reference: MacArthur DG et al (2012). A systematic survey of loss-of-function variants in human protein-coding genes. <em>Science</em> 335:823–828</p>
<p>It is actually part of the standard VEP plugins.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> loftee
<span class="c1"># human_ancestor_fa</span>
<span class="c1">## samtools --version gives 0.1.19</span>
wget http://www.broadinstitute.org/~konradk/loftee/human_ancestor.fa.rz
wget http://www.broadinstitute.org/~konradk/loftee/human_ancestor.fa.rz.fai
<span class="c1">## samtools --version gives 1.x</span>
wget https://s3.amazonaws.com/bcbio_nextgen/human_ancestor.fa.gz
wget https://s3.amazonaws.com/bcbio_nextgen/human_ancestor.fa.gz.fai
wget https://s3.amazonaws.com/bcbio_nextgen/human_ancestor.fa.gz.gzi
<span class="c1"># conservation_file</span>
wget https://personal.broadinstitute.org/konradk/loftee_data/GRCh37/phylocsf_gerp.sql.gz
wget https://www.broadinstitute.org/~konradk/loftee/phylocsf_data.tsv.gz
wget https://personal.broadinstitute.org/konradk/loftee_data/GRCh37/GERP_scores.final.sorted.txt.gz
wget https://personal.broadinstitute.org/konradk/loftee_data/GRCh37/GERP_scores.exons.txt.gz
<span class="c1"># annotation</span>
vep --id <span class="s2">&quot;1 154426970 154426970 A/C 1&quot;</span> --species homo_sapiens -o rs2228145 --cache --offline --force_overwrite <span class="se">\</span>
    --assembly GRCh37 --plugin LoF,loftee_path:.,human_ancestor_fa:human_ancestor.fa.gz,conservation_file:phylocsf_gerp.sql.gz
<span class="c1"># VEP documentation example</span>
vep --input_file homo_sapiens_GRCh37.vcf --output_file <span class="nb">test</span> --cache --dir_cache <span class="si">${</span><span class="nv">HPC_WORK</span><span class="si">}</span>/ensembl-vep/.vep --dir_plugins <span class="si">${</span><span class="nv">HPC_WORK</span><span class="si">}</span>/loftee --offline <span class="se">\</span>
    --pick --force_overwrite --species homo_sapiens --assembly GRCh37 <span class="se">\</span>
    --plugin LoF,loftee_path:.,human_ancestor_fa:human_ancestor.fa.gz,conservation_file:phylocsf_gerp.sql.gz
</pre></div>
</div>
<p>If offers implementation for parsing CSQ field but is also possible with R as described below. Note that if loftee_path uses an absolute path, that path should also be within PERL5LIB, e.g.,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>export PERL5LIB=$PERL5LIB:$HPC_WORK/loftee
</pre></div>
</div>
<p>is put in .bashrc. We see now all files are provided, and there are complaints over rz was not generated by bgzip, so we do</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># error</span>
samtools faidx human_ancestor.fa.rz
<span class="c1"># rework</span>
gunzip -c human_ancestor.fa.rz <span class="p">|</span> bgzip -c &gt; output.fa.gz
<span class="c1"># check and put it back</span>
ll human_ancestor.fa.rz output.fa.gz
mv output.fa.gz human_ancestor.fa.rz
<span class="c1"># now possible and check</span>
samtools faidx human_ancestor.fa.rz
ll human_ancestor.fa.*
</pre></div>
</div>
<p>You may also like to explore Hail, <a class="reference external" href="https://hail.is/">https://hail.is/</a> as with <a class="reference external" href="https://github.com/danking/hail-cloud-docs/blob/master/how-to-cloud.md">hail on cloud</a>. We proceed as follows,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load python/3.7 hadoop/2.7.7
virtualenv py37
<span class="nb">source</span> py37/bin/activate
pip install hail
pip install gnomAD
python
&gt;&gt;&gt; import hail as hl
&gt;&gt;&gt; <span class="nv">mt</span> <span class="o">=</span> hl.balding_nichols_model<span class="o">(</span><span class="nv">n_populations</span><span class="o">=</span><span class="m">3</span>, <span class="nv">n_samples</span><span class="o">=</span><span class="m">50</span>, <span class="nv">n_variants</span><span class="o">=</span><span class="m">100</span><span class="o">)</span>
&gt;&gt;&gt; mt.describe<span class="o">()</span>
&gt;&gt;&gt; mt.summarize<span class="o">()</span>
&gt;&gt;&gt; mt.count<span class="o">()</span>
&gt;&gt;&gt; mt.show<span class="o">()</span>
&gt;&gt;&gt; mt.write<span class="o">(</span><span class="s2">&quot;hail&quot;</span><span class="o">)</span>
&gt;&gt;&gt; <span class="nv">at</span> <span class="o">=</span> hl.import_vcf<span class="o">(</span><span class="s2">&quot;INTERVAL.vcf.bgz&quot;</span><span class="o">)</span>
&gt;&gt;&gt; hl.export_vcf<span class="o">(</span>at,<span class="s2">&quot;at.vcf.bgz&quot;</span><span class="o">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id1">
<h3>— R —<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>This is a wrapper and <code class="docutils literal notranslate"><span class="pre">the</span> <span class="pre">Ensembl</span> <span class="pre">VEP</span> <span class="pre">perl</span> <span class="pre">script</span> <span class="pre">must</span> <span class="pre">be</span> <span class="pre">installed</span> <span class="pre">in</span> <span class="pre">your</span> <span class="pre">path</span></code>. Expected to be slower than the <code class="docutils literal notranslate"><span class="pre">--offline</span></code> mode above, it is
relatively easy to set up,</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;ensemblVEP&quot;</span><span class="p">)</span>
<span class="nf">vignette</span><span class="p">(</span><span class="s">&quot;ensemblVEP&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ensemblVEP</span><span class="p">)</span>
<span class="n">file</span> <span class="o">&lt;-</span> <span class="nf">system.file</span><span class="p">(</span><span class="s">&quot;extdata&quot;</span><span class="p">,</span> <span class="s">&quot;ex2.vcf&quot;</span><span class="p">,</span> <span class="n">package</span><span class="o">=</span><span class="s">&quot;VariantAnnotation&quot;</span><span class="p">)</span>
<span class="n">param</span> <span class="o">&lt;-</span> <span class="nf">VEPFlags</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">vcf</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">check_existing</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">symbol</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">terms</span><span class="o">=</span><span class="s">&quot;SO&quot;</span><span class="p">,</span><span class="n">sift</span><span class="o">=</span><span class="s">&quot;b&quot;</span><span class="p">,</span><span class="n">polyphen</span><span class="o">=</span><span class="s">&quot;p&quot;</span><span class="p">))</span>
<span class="n">vep</span> <span class="o">&lt;-</span> <span class="nf">ensemblVEP</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">param</span><span class="p">)</span>
<span class="nf">info</span><span class="p">(</span><span class="n">vep</span><span class="p">)</span><span class="o">$</span><span class="n">CSQ</span>
<span class="n">csq</span> <span class="o">&lt;-</span> <span class="nf">parseCSQToGRanges</span><span class="p">(</span><span class="n">vep</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">csq</span><span class="p">,</span> <span class="m">2</span><span class="p">)</span>
<span class="c1"># clinvar</span>
<span class="n">param</span> <span class="o">&lt;-</span> <span class="nf">VEPFlags</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">vcf</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">output_file</span><span class="o">=</span><span class="s">&quot;clinvar.vcf&quot;</span><span class="p">,</span><span class="n">force_overwrite</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">assembly</span><span class="o">=</span><span class="s">&quot;GRCh37&quot;</span><span class="p">,</span><span class="n">port</span><span class="o">=</span><span class="m">3337</span><span class="p">,</span>
                  <span class="n">custom</span><span class="o">=</span><span class="s">&quot;clinvar_GRCh37.vcf.gz,ClinVar,vcf,exact,0,CLNSIG,CLNREVSTAT,CLNDN&quot;</span><span class="p">))</span>
<span class="nf">ensemblVEP</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">param</span><span class="p">)</span>
</pre></div>
</div>
<p>The second <code class="docutils literal notranslate"><span class="pre">ensemblVEP</span></code> obtains <code class="docutils literal notranslate"><span class="pre">clinvar.vcf</span></code> and <code class="docutils literal notranslate"><span class="pre">clinvar.vcf_summary.html</span></code>. Annotation is made to a VCF file, and returns data with unparsed ‘CSQ’.</p>
<p>The facility to parse the CSQ column of a VCF object is done for the documentation example.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># VCF output from the VEP web interface or the call above</span>
<span class="n">vep</span> <span class="o">&lt;-</span> <span class="s">&quot;INF1.merge.trans.vcf&quot;</span>
<span class="c1"># Parse into a GRanges and include the &#39;VCFRowID&#39; column.</span>
<span class="n">vcf</span> <span class="o">&lt;-</span> <span class="nf">readVcf</span><span class="p">(</span><span class="n">vep</span><span class="p">,</span> <span class="s">&quot;hg19&quot;</span><span class="p">)</span>
<span class="n">csq</span> <span class="o">&lt;-</span> <span class="nf">parseCSQToGRanges</span><span class="p">(</span><span class="n">vep</span><span class="p">,</span> <span class="n">VCFRowID</span><span class="o">=</span><span class="nf">rownames</span><span class="p">(</span><span class="n">vcf</span><span class="p">))</span>
<span class="nf">write.table</span><span class="p">(</span><span class="nf">mcols</span><span class="p">(</span><span class="n">csq</span><span class="p">),</span><span class="n">file</span><span class="o">=</span><span class="s">&quot;INF1.merge.trans.txt&quot;</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The dbNSFP counterpart is also possible</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;myvariant&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">VariantAnnotation</span><span class="p">)</span>
<span class="n">file</span> <span class="o">&lt;-</span> <span class="nf">system.file</span><span class="p">(</span><span class="s">&quot;extdata&quot;</span><span class="p">,</span> <span class="s">&quot;dbsnp_mini.vcf&quot;</span><span class="p">,</span> <span class="n">package</span><span class="o">=</span><span class="s">&quot;myvariant&quot;</span><span class="p">)</span>
<span class="n">vcf</span> <span class="o">&lt;-</span> <span class="nf">readVcf</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s">&quot;hg19&quot;</span><span class="p">)</span>
<span class="nf">rowRanges</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">myvariant</span><span class="p">)</span>
<span class="n">hgvs</span> <span class="o">&lt;-</span> <span class="nf">formatHgvs</span><span class="p">(</span><span class="n">vcf</span><span class="p">,</span> <span class="n">variant_type</span><span class="o">=</span><span class="s">&quot;snp&quot;</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">hgvs</span><span class="p">)</span>
<span class="nf">getVariants</span><span class="p">(</span><span class="n">hgvs</span><span class="p">)</span>
<span class="n">rsids</span> <span class="o">&lt;-</span> <span class="nf">paste</span><span class="p">(</span><span class="s">&quot;rs&quot;</span><span class="p">,</span> <span class="nf">info</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span><span class="o">$</span><span class="n">RS</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">rsids</span><span class="p">)</span>
<span class="n">res</span> <span class="o">&lt;-</span> <span class="nf">queryVariants</span><span class="p">(</span><span class="n">q</span><span class="o">=</span><span class="n">rsids</span><span class="p">,</span> <span class="n">scopes</span><span class="o">=</span><span class="s">&quot;dbsnp.rsid&quot;</span><span class="p">,</span> <span class="n">fields</span><span class="o">=</span><span class="s">&quot;all&quot;</span><span class="p">)</span>
<span class="n">fields</span> <span class="o">&lt;-</span> <span class="nf">names</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="n">cadd</span> <span class="o">&lt;-</span> <span class="nf">grep</span><span class="p">(</span><span class="s">&#39;cadd&#39;</span><span class="p">,</span><span class="n">fields</span><span class="p">)</span>
<span class="n">res</span><span class="p">[</span><span class="n">fields</span><span class="p">[</span><span class="n">cadd</span><span class="p">]]</span>
</pre></div>
</div>
<p>Note that the CSQ field could also be handled by bcftools split-vep plugin, see <a class="reference external" href="http://samtools.github.io/bcftools/howtos/plugin.split-vep.html">http://samtools.github.io/bcftools/howtos/plugin.split-vep.html</a>.</p>
</div>
<div class="section" id="docker">
<h3>— docker —<a class="headerlink" href="#docker" title="Permalink to this headline">¶</a></h3>
<p>See <code class="docutils literal notranslate"><span class="pre">docker/Dockerfile</span> </code> from the GitHub directory above, or https://github.com/Ensembl/ensembl-vep.</p>
</div>
<div class="section" id="virtual-machine">
<h3>— Virtual machine —<a class="headerlink" href="#virtual-machine" title="Permalink to this headline">¶</a></h3>
<p>See http://www.ensembl.org/info/data/virtual_machine.html which is possibly best for MicroSoft Windows and is not pursued here.</p>
<p>ENSEMBL-synonym translation (hg19) file</p>
</div>
</div>
<div class="section" id="r-xlsx">
<h2>R/xlsx<a class="headerlink" href="#r-xlsx" title="Permalink to this headline">¶</a></h2>
<p>When there is message,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Error: .onLoad failed in loadNamespace() for &#39;rJava&#39;, details:
  call: dyn.load(file, DLLpath = DLLpath, ...)
  error: unable to load shared object &#39;/rds/user/jhz22/hpc-work/R/rJava/libs/rJava.so&#39;:
  libjvm.so: cannot open shared object file: No such file or directory
Execution halted
ERROR: lazy loading failed for package ‘xlsx’
* removing ‘/rds/user/jhz22/hpc-work/R/xlsx’
* restoring previous ‘/rds/user/jhz22/hpc-work/R/xlsx’
</pre></div>
</div>
<p>then see description on <code class="docutils literal notranslate"><span class="pre">rJava</span></code> earlier.</p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">CSD3</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Jing Hua Zhao.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.4.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/usage.md.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>