<!DOCTYPE html> <html lang="en" dir="auto"> <head><meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=5, user-scalable=no"> <meta name="description" content="regenie Web: https://github.com/rgcgithub/regenie (documentation) It is the easiest to use the Centos 7 distribution, wget -qO- https://github.com/..."> <meta name="revised" content="b5f8beb971e1e7020fb16231556a1ec40359efd4"> <meta name="author" content="cambridge-ceu"> <meta name="generator" content="rundocs/jekyll-rtd-theme v2.0.10"><meta name="theme-color" content="#2980b9"> <title>regenie 路 csd3</title> <meta name="twitter:title" content="regenie 路 csd3"> <meta name="twitter:description" content="regenie Web: https://github.com/rgcgithub/regenie (documentation) It is the easiest to use the Centos 7 distribution, wget -qO- https://github.com/..."> <meta name="twitter:card" content="summary"> <meta name="twitter:site" content="@cambridge-ceu"> <meta name="twitter:url" content="/applications/regenie.html"> <meta name="twitter:creator" content="@rundocs/jekyll-rtd-theme v2.0.10"> <meta property="og:title" content="regenie 路 csd3"> <meta property="og:description" content="regenie Web: https://github.com/rgcgithub/regenie (documentation) It is the easiest to use the Centos 7 distribution, wget -qO- https://github.com/..."> <meta property="og:locale" content="en"> <meta property="og:url" content="/applications/regenie.html"> <meta property="og:type" content="article"> <meta property="article:author" content="cambridge-ceu"> <meta property="article:published_time" content="2021-06-28T15:26:15+01:00"> <meta property="article:modified_time" content="2021-06-28T15:26:15+01:00"> <script type="application/ld+json"> { "@context": "https://schema.org", "@type": "Article", "mainEntityOfPage": { "@type": "WebPage", "@id": "/applications/regenie.html" }, "headline": "regenie 路 csd3", "image": [], "author": { "@type": "Person", "name": "cambridge-ceu" }, "datePublished": "2021-06-28T15:26:15+01:00", "dateModified": "2021-06-28T15:26:15+01:00", "publisher": { "@type": "", "name": "cambridge-ceu", "logo": { "@type": "ImageObject", "url": "" } }, "description": "regenie Web: https://github.com/rgcgithub/regenie (documentation) It is the easiest to use the Centos 7 distribution, wget -qO- https://github.com/..." } </script> <link rel="dns-prefetch" href="https://cdn.jsdelivr.net"><link rel="prev" href="/applications/qpdf.html"><link rel="next" href="/applications/rgdal.html"><link rel="canonical" href="/applications/regenie.html"><link rel="icon" type="image/svg+xml" href="/assets/images/favicon.svg"><link rel="icon" type="image/png" href="/assets/images/favicon-16x16.png" sizes="16x16"> <link rel="icon" type="image/png" href="/assets/images/favicon-32x32.png" sizes="32x32"> <link rel="icon" type="image/png" href="/assets/images/favicon-96x96.png" sizes="96x96"><link rel="mask-icon" href="/assets/images/favicon.svg" color="#2980b9"><link rel="apple-touch-icon" href="/assets/images/apple-touch-icon-300x300.jpg"> <link rel="stylesheet" href="/assets/css/theme.css"><script> window.ui = { title: "csd3", baseurl: "", i18n: { search_results: "Search Results", search_results_found: "Search finished, found # page(s) matching the search query.", search_results_not_found: "Your search did not match any documents, please make sure that all characters are spelled correctly!" } }; </script> </head> <body class="container"><div class="sidebar-wrap overflow-hidden"> <div class="sidebar height-full overflow-y-scroll overflow-x-hidden"> <div class="header d-flex flex-column p-3 text-center"> <div class="title pb-1"> <a class="h4 no-underline py-1 px-2 rounded-1" href="/" title="A collection of CSD3 notes"> <i class="fa fa-home"></i> csd3 </a> </div> <span class="version"></span> <form class="search pt-2" action="/search.html" method="get" autocomplete="off"> <input class="form-control input-block input-sm" type="text" name="q" placeholder="Search docs..."> </form> </div> <div class="toctree py-2" data-spy="affix" role="navigation" aria-label="main navigation"> <ul> </ul> <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/systems/"> The system </a><ul> <li class="toc level-1 " data-sort="1" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/system.html">1. System</a> </li> <li class="toc level-1 " data-sort="2" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/login.html">2. Login</a> </li> <li class="toc level-1 " data-sort="3" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/directories.html">3. Directories</a> </li> <li class="toc level-1 " data-sort="4" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/policies.html">4. Policies</a> </li> <li class="toc level-1 " data-sort="5" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/email.html">5. email</a> </li> <li class="toc level-1 " data-sort="6" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/software.html">6. Software</a> </li> <li class="toc level-1 " data-sort="7" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/training.html">7. Training</a> </li> <li class="toc level-1 " data-sort="8" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/acknowledgement.html">8. Acknowledgement</a> </li> <li class="toc level-1 " data-sort="8" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/contacts.html">8. Contacts</a> </li></ul> <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/cardio/"> cardio </a><ul> </ul> <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/applications/"> Applications </a><ul> <li class="toc level-1 " data-sort="1" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/ABCtoolbox.html">1. ABCtoolbox</a> </li> <li class="toc level-1 " data-sort="2" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/AEGIS.html">2. AEGIS</a> </li> <li class="toc level-1 " data-sort="3" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/akt.html">3. akt</a> </li> <li class="toc level-1 " data-sort="4" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/ANNOVAR.html">4. ANNOVAR</a> </li> <li class="toc level-1 " data-sort="5" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/bcftools.html">5. bcftools</a> </li> <li class="toc level-1 " data-sort="6" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/bedops.html">6. bedops</a> </li> <li class="toc level-1 " data-sort="7" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/bgenix.html">7. bgenix</a> </li> <li class="toc level-1 " data-sort="8" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/brms.html">8. R/brms</a> </li> <li class="toc level-1 " data-sort="9" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/CaVEMaN.html">9. CaVEMaN</a> </li> <li class="toc level-1 " data-sort="10" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/citeproc.html">10. citeproc</a> </li> <li class="toc level-1 " data-sort="11" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/DosageConverter.html">11. DosageConverter</a> </li> <li class="toc level-1 " data-sort="12" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/EasyQC.html">12. EasyQC</a> </li> <li class="toc level-1 " data-sort="13" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/gnn.html">13. R/gnn</a> </li> <li class="toc level-1 " data-sort="14" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/gsutil.html">14. gsutil</a> </li> <li class="toc level-1 " data-sort="15" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/gwas2vcf.html">15. gwas2vcf</a> </li> <li class="toc level-1 " data-sort="16" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/hail.html">16. hail</a> </li> <li class="toc level-1 " data-sort="17" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/HESS.html">17. HESS</a> </li> <li class="toc level-1 " data-sort="18" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/KentUtils.html">18. KentUtils</a> </li> <li class="toc level-1 " data-sort="19" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/lapack.html">19. lapack</a> </li> <li class="toc level-1 " data-sort="20" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/LDlinkR.html">20. R/LDlinkR</a> </li> <li class="toc level-1 " data-sort="21" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/LEMMA.html">21. LEMMA</a> </li> <li class="toc level-1 " data-sort="22" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/magick.html">22. R/magick</a> </li> <li class="toc level-1 " data-sort="23" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/MetaXcan.html">23. MetaXcan</a> </li> <li class="toc level-1 " data-sort="24" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/pandoc.html">24. pandoc</a> </li> <li class="toc level-1 " data-sort="25" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/pca_projection.html">25. PCA projection</a> </li> <li class="toc level-1 " data-sort="26" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/PhenoScanner.html">26. R/PhenoScanner</a> </li> <li class="toc level-1 " data-sort="27" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/PKI.html">27. R/PKI</a> </li> <li class="toc level-1 " data-sort="28" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/plumber.html">28. R/plumber</a> </li> <li class="toc level-1 " data-sort="29" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/polyphen-2.html">29. polyphen-2</a> </li> <li class="toc level-1 " data-sort="30" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/poppler.html">30. poppler</a> </li> <li class="toc level-1 " data-sort="31" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/protolite.html">31. R/protolite</a> </li> <li class="toc level-1 " data-sort="32" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/PRSice.html">32. PRSice</a> </li> <li class="toc level-1 " data-sort="33" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/PRSoS.html">33. PRSoS</a> </li> <li class="toc level-1 " data-sort="34" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/pspp.html">34. pspp</a> </li> <li class="toc level-1 " data-sort="35" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/qpdf.html">35. qpdf</a> </li> <li class="toc level-1 current" data-sort="36" data-level="1"> <a class="d-flex flex-items-baseline current" href="/applications/regenie.html">36. regenie</a> </li> <li class="toc level-1 " data-sort="37" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/rgdal.html">37. R/rgdal</a> </li> <li class="toc level-1 " data-sort="38" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/rgeos.html">38. R/rgeos</a> </li> <li class="toc level-1 " data-sort="39" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/Rhdf5lib.html">39. R/Rhdf5lib</a> </li> <li class="toc level-1 " data-sort="40" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/rjags.html">40. R/rjags</a> </li> <li class="toc level-1 " data-sort="41" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/rJava.html">41. R/rJava</a> </li> <li class="toc level-1 " data-sort="42" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/R.html">42. R</a> </li> <li class="toc level-1 " data-sort="43" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/rstan.html">43. R/rstan</a> </li> <li class="toc level-1 " data-sort="44" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/RStudio.html">44. RStudio</a> </li> <li class="toc level-1 " data-sort="45" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/SAIGE.html">45. R/SAIGE 0.36.6 and 0.39.2</a> </li> <li class="toc level-1 " data-sort="46" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/sf.html">46. R/sf</a> </li> <li class="toc level-1 " data-sort="47" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/shapeit3.html">47. shapeit3</a> </li> <li class="toc level-1 " data-sort="48" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/SMR.html">48. SMR</a> </li> <li class="toc level-1 " data-sort="49" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/snakemake.html">49. snakemake</a> </li> <li class="toc level-1 " data-sort="50" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/snpnet.html">50. R/snpnet</a> </li> <li class="toc level-1 " data-sort="51" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/SNPTEST.html">51. SNPTEST</a> </li> <li class="toc level-1 " data-sort="52" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/sojo.html">52. R/sojo</a> </li> <li class="toc level-1 " data-sort="53" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/tabix.html">53. tabix</a> </li> <li class="toc level-1 " data-sort="54" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/VEP.html">54. VEP</a> </li> <li class="toc level-1 " data-sort="55" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/xlsx.html">55. R/xlsx</a> </li></ul> <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/writing/"> Writing Related </a><ul> <li class="toc level-1 " data-sort="1" data-level="1"> <a class="d-flex flex-items-baseline " href="/writing/front-matter.html">1. Front matter</a> </li> <li class="toc level-1 " data-sort="2" data-level="1"> <a class="d-flex flex-items-baseline " href="/writing/list.html">2. List directory</a> </li> <li class="toc level-1 " data-sort="3" data-level="1"> <a class="d-flex flex-items-baseline " href="/writing/mathjax.html">3. Mathjax</a> </li> <li class="toc level-1 " data-sort="4" data-level="1"> <a class="d-flex flex-items-baseline " href="/writing/admonitions.html">4. Admonition card</a> </li> <li class="toc level-1"> <a class="d-flex flex-items-baseline" href="/writing/test/"> Test Documentation </a><ul> <li class="toc level-2 " data-sort="1" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/markdown.html">1. Markdown Elements</a> </li> <li class="toc level-2 " data-sort="2" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/toasts.html">2. Toasts Card</a> </li> <li class="toc level-2 " data-sort="3" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/codes.html">3. Code Blocks</a> </li> <li class="toc level-2 " data-sort="4" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/mermaid.html">4. Mermaid Test</a> </li> <li class="toc level-2 " data-sort="5" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/emoji.html">5. Emoji Test</a> </li> <li class="toc level-2 " data-sort="6" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/gist.html">6. Gist Test</a> </li> <li class="toc level-2 " data-sort="7" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/avatar.html">7. Avatar Test</a> </li> <li class="toc level-2 " data-sort="8" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/mentions.html">8. Mentions Test</a> </li> <li class="toc level-2 " data-sort="9" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/fonts.html">9. Fonts Test</a> </li> <li class="toc level-2 " data-sort="10" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/utilities.html">10. Primer Utilities Test</a> </li></ul> </li></ul> </div> </div> </div> <div class="content-wrap"> <div class="header d-flex flex-justify-between p-2 hide-lg hide-xl" aria-label="top navigation"> <button id="toggle" aria-label="Toggle menu" class="btn-octicon p-2 m-0 text-white" type="button"> <i class="fa fa-bars"></i> </button> <div class="title flex-1 d-flex flex-justify-center"> <a class="h4 no-underline py-1 px-2 rounded-1" href="/">csd3</a> </div> </div> <div class="content p-3 p-sm-5"> <div class="navigation-top d-flex flex-justify-between"> <ul class="breadcrumb" role="navigation" aria-label="breadcrumbs navigation"> <li class="breadcrumb-item"> <a class="no-underline" href="/" title="/"> <i class="fa fa-home"></i> </a> </li><li class="breadcrumb-item" ><a href="/applications/">applications</a></li><li class="breadcrumb-item" aria-current="page">regenie.md</li></ul> <a class="edit" href="https://github.com/cambridge-ceu/csd3/edit/gh-pages/applications/regenie.md" title="Edit on GitHub" rel="noreferrer" target="_blank"> <i class="fa fa-edit"></i> </a> </div> <hr> <div role="main" itemscope="itemscope" itemtype="https://schema.org/Article"> <div class="markdown-body" itemprop="articleBody"> <h1 id="regenie">regenie</h1> <p>Web: <a href="https://github.com/rgcgithub/regenie">https://github.com/rgcgithub/regenie</a> (<a href="https://rgcgithub.github.io/regenie/">documentation</a>)</p> <p>It is the easiest to use the Centos 7 distribution,</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>wget <span class="nt">-qO-</span> https://github.com/rgcgithub/regenie/releases/download/v2.0.1-newest/regenie_v2.0.2.gz_x86_64_Centos7_mkl.zip | <span class="se">\</span>
<span class="nb">gunzip</span> <span class="nt">-c</span> <span class="o">&gt;</span> regenie_v2.0.2
<span class="nb">chmod</span> +x regenie_v2.0.2
regenie_v2.0.2 <span class="nt">--help</span>
</code></pre> </div></div> <p>The last command gives the following information (Why .gz in the banner?),</p> <div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>              |=============================|
              |      REGENIE v2.0.2.gz      |
              |=============================|

Copyright (c) 2020-2021 Joelle Mbatchou, Andrey Ziyatdinov and Jonathan Marchini.
Distributed under the MIT License.
Compiled with Boost Iostream library.
Using Intel MKL with Eigen.


Usage:
  regenie_v2.0.2 [OPTION...]

  -h, --help      print list of available options
      --helpFull  print list of all available options

 Main options:
      --step INT                specify if fitting null model (=1) or
                                association testing (=2)
      --bed PREFIX              prefix to PLINK .bed/.bim/.fam files
      --pgen PREFIX             prefix to PLINK2 .pgen/.pvar/.psam files
      --bgen FILE               BGEN file
      --sample FILE             sample file corresponding to BGEN file
      --ref-first               use the first allele as the reference for
                                BGEN or PLINK bed/bim/fam input format [default
                                assumes reference is last]
      --keep FILE               file listing samples to retain in the
                                analysis (no header; starts with FID IID)
      --remove FILE             file listing samples to remove from the
                                analysis (no header; starts with FID IID)
      --extract FILE            file with IDs of variants to retain in the
                                analysis
      --exclude FILE            file with IDs of variants to remove from the
                                analysis
  -p, --phenoFile FILE          phenotype file (header required starting with
                                FID IID)
      --phenoCol STRING         phenotype name in header (use for each
                                phenotype to keep; can use parameter expansion
                                {i:j})
      --phenoColList STRING,..,STRING
                                comma separated list of phenotype names to
                                keep (can use parameter expansion {i:j})
  -c, --covarFile FILE          covariate file (header required starting with
                                FID IID)
      --covarCol STRING         covariate name in header (use for each
                                covariate to keep; can use parameter expansion
                                {i:j})
      --covarColList STRING,..,STRING
                                comma separated list of covariate names to
                                keep (can use parameter expansion {i:j})
      --catCovarList STRING,..,STRING
                                comma separated list of categorical
                                covariates
  -o, --out PREFIX              prefix for output files
      --bt                      analyze phenotypes as binary
  -1, --cc12                    use control=1,case=2,missing=NA encoding for
                                binary traits
  -b, --bsize INT               size of genotype blocks
      --cv INT(=5)              number of cross validation (CV) folds
      --loocv                   use leave-one out cross validation (LOOCV)
      --l0 INT(=5)              number of ridge parameters to use when
                                fitting models within blocks [evenly spaced in
                                (0,1)]
      --l1 INT(=5)              number of ridge parameters to use when
                                fitting model across blocks [evenly spaced in (0,1)]
      --lowmem                  reduce memory usage by writing level 0
                                predictions to temporary files
      --lowmem-prefix PREFIX    prefix where to write the temporary files in
                                step 1 (default is to use prefix from --out)
      --split-l0 PREFIX,N       split level 0 across N jobs and set prefix of
                                output files
      --run-l0 FILE,K           run level 0 for job K in {1..N} using master
                                file created from '--split-l0'
      --run-l1 FILE             run level 1 using master file from
                                '--split-l0'
      --keep-l0                 avoid deleting the level 0 predictions
                                written on disk after fitting the level 1 models
      --strict                  remove all samples with missingness at any of
                                the traits
      --print-prs               also output polygenic predictions without
                                using LOCO (=whole genome PRS)
      --gz                      compress output files (gzip format)
      --threads INT             number of threads
      --pred FILE               file containing the list of predictions files
                                from step 1
      --ignore-pred             skip reading predictions from step 1
                                (equivalent to linear/logistic regression with only
                                covariates)
      --use-prs                 when using whole genome PRS step 1 output in
                                '--pred'
      --write-samples           write IDs of samples included for each trait
                                (only in step 2)
      --minMAC FLOAT(=5)        minimum minor allele count (MAC) for tested
                                variants
      --minINFO DOUBLE(=0)      minimum imputation info score (Impute/Mach
                                R^2) for tested variants
      --split                   split asssociation results into separate
                                files for each trait
      --firth                   use Firth correction for p-values less than
                                threshold
      --approx                  use approximation to Firth correction for
                                computational speedup
      --spa                     use Saddlepoint approximation (SPA) for
                                p-values less than threshold
      --pThresh FLOAT(=0.05)    P-value threshold below which to apply
                                Firth/SPA correction
      --chr STRING              specify chromosome to test in step 2 (use for
                                each chromosome)
      --chrList STRING,..,STRING
                                Comma separated list of chromosomes to test
                                in step 2
      --range CHR:MINPOS-MAXPOS
                                to specify a physical position window for
                                variants to test in step 2
      --test STRING             'dominant' or 'recessive' (default is
                                additive test)
      --interaction STRING      perform interaction testing with a
                                quantitative/categorical covariate
      --hc4-mac FLOAT(=100)     minor allele count (MAC) threshold below
                                which to use robust SE with HC4 instead of HC3 for
                                interaction testing
      --set-list FILE           file with sets definition
      --extract-sets FILE       file with IDs of sets to retain in the
                                analysis
      --exclude-sets FILE       file with IDs of sets to remove from the
                                analysis
      --extract-setlist FILE    comma separated list of sets to retain in the
                                analysis
      --exclude-setlist FILE    comma separated list of sets to remove from
                                the analysis
      --anno-file FILE          file with variant annotations
      --anno-labels FILE        file with labels to annotations
      --mask-def FILE           file with mask definitions
      --aaf-file FILE           file with AAF to use when building masks
      --aaf-bins FLOAT,..,FLOAT
                                comma separated list of AAF bins cutoffs for
                                building masks
      --build-mask STRING       rule to construct masks, can be 'max', 'sum'
                                or 'comphet' (default is max)
      --singleton-carrier       define singletons as variants with a single
                                carrier in the sample
      --write-mask              write masks in PLINK bed/bim/fam format
      --mask-lovo STRING        apply Leave-One-Variant-Out (LOVO) scheme
                                when building masks
                                (&lt;set_name&gt;,&lt;mask_name&gt;,&lt;aaf_cutoff&gt;)
      --mask-lodo STRING        apply Leave-One-Domain-Out (LODO) scheme when
                                building masks
                                (&lt;set_name&gt;,&lt;mask_name&gt;,&lt;aaf_cutoff&gt;)
      --skip-test               skip computing association tests after
                                building masks
      --check-burden-files      check annotation file, set list file and mask
                                file for consistency
      --strict-check-burden     to exit early if the annotation, set list and
                                mask definition files dont agree

For more information, use option '--help' or visit the website: https://rgcgithub.github.io/regenie/

</code></pre> </div></div> <h2 id="reference">Reference</h2> <p>Mbatchou, J., et al. (2020). "Computationally efficient whole genome regression for quantitative and binary traits." bioRxiv: 2020.2006.2019.162354. https://www.biorxiv.org/content/10.1101/2020.06.19.162354v1</p> </div> </div> <div class="navigation-bottom d-flex flex-justify-between py-3" role="navigation" aria-label="footer navigation"> <div class="prev"><a href="/applications/qpdf.html" class="btn" title="qpdf" accesskey="p" rel="prev"> <i class="fa fa-arrow-circle-left"></i> Previous </a></div> <div class="next"><a href="/applications/rgdal.html" class="btn" title="R/rgdal" accesskey="n" rel="next"> Next <i class="fa fa-arrow-circle-right"></i> </a></div> </div><hr> <div class="copyright text-center text-gray" role="contentinfo"> <i class="fa fa-copyright"></i> <span class="time">2021,</span> <a class="text-gray" href="https://github.com/cambridge-ceu" rel="noreferrer" target="_blank">cambridge-ceu</a> Revision <a class="text-gray" href="https://github.com/cambridge-ceu/csd3/commit/b5f8beb971e1e7020fb16231556a1ec40359efd4" title="b5f8beb971e1e7020fb16231556a1ec40359efd4" rel="noreferrer" target="_blank">b5f8beb</a> <br> <div class="generator"> Built with <a href="https://pages.github.com" rel="noreferrer" target="_blank" title="github-pages v209">GitHub Pages</a> using a <a href="https://github.com/rundocs/jekyll-rtd-theme" rel="noreferrer" target="_blank" title="rundocs/jekyll-rtd-theme v2.0.10">theme</a> provided by <a href="https://rundocs.io" rel="noreferrer" target="_blank">RunDocs</a>. </div> </div> </div> </div> <div class="addons-wrap d-flex flex-column overflow-y-auto"> <div class="status d-flex flex-justify-between p-2"> <div class="title p-1"> <i class="fa fa-book"></i> csd3 </div> <div class="branch p-1"> <span class="name"> gh-pages </span> <i class="fa fa-caret-down"></i> </div> </div> <div class="addons d-flex flex-column height-full p-2 d-none"> <dl> <dt>GitHub</dt> <dd> <a href="https://github.com/cambridge-ceu/csd3" title="Stars: "> <i class="fa fa-github"></i> Homepage </a> </dd> <dd> <a href="" title="Open issues: "> <i class="fa fa-question-circle-o"></i> Issues </a> </dd> <dd> <a href="https://github.com/cambridge-ceu/csd3/zipball/gh-pages" title="Size: Kb"> <i class="fa fa-download"></i> Download </a> </dd> </dl> <hr> <div class="license f6 pb-2"> This <a href="/" title="csd3">Software</a> is under the terms of <a href="https://github.com/cambridge-ceu/csd3">The Unlicense</a>. </div> </div> </div> <script src="/assets/js/jquery.min.js"></script><script src="/assets/js/theme.js"></script> </body> </html>