<!DOCTYPE html> <html lang="en" dir="auto"> <head><meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=5, user-scalable=no"> <meta name="description" content="VEP Official page: https://www.ensembl.org/info/docs/tools/vep/index.html (web interface). Detailed instructions for installation are available fro..."> <meta name="revised" content="b5f8beb971e1e7020fb16231556a1ec40359efd4"> <meta name="author" content="cambridge-ceu"> <meta name="generator" content="rundocs/jekyll-rtd-theme v2.0.10"><meta name="theme-color" content="#2980b9"> <title>VEP · csd3</title> <meta name="twitter:title" content="VEP · csd3"> <meta name="twitter:description" content="VEP Official page: https://www.ensembl.org/info/docs/tools/vep/index.html (web interface). Detailed instructions for installation are available fro..."> <meta name="twitter:card" content="summary"> <meta name="twitter:site" content="@cambridge-ceu"> <meta name="twitter:url" content="/applications/VEP.html"> <meta name="twitter:creator" content="@rundocs/jekyll-rtd-theme v2.0.10"> <meta property="og:title" content="VEP · csd3"> <meta property="og:description" content="VEP Official page: https://www.ensembl.org/info/docs/tools/vep/index.html (web interface). Detailed instructions for installation are available fro..."> <meta property="og:locale" content="en"> <meta property="og:url" content="/applications/VEP.html"> <meta property="og:type" content="article"> <meta property="article:author" content="cambridge-ceu"> <meta property="article:published_time" content="2021-06-28T15:26:15+01:00"> <meta property="article:modified_time" content="2021-06-28T15:26:15+01:00"> <script type="application/ld+json"> { "@context": "https://schema.org", "@type": "Article", "mainEntityOfPage": { "@type": "WebPage", "@id": "/applications/VEP.html" }, "headline": "VEP · csd3", "image": [], "author": { "@type": "Person", "name": "cambridge-ceu" }, "datePublished": "2021-06-28T15:26:15+01:00", "dateModified": "2021-06-28T15:26:15+01:00", "publisher": { "@type": "", "name": "cambridge-ceu", "logo": { "@type": "ImageObject", "url": "" } }, "description": "VEP Official page: https://www.ensembl.org/info/docs/tools/vep/index.html (web interface). Detailed instructions for installation are available fro..." } </script> <link rel="dns-prefetch" href="https://cdn.jsdelivr.net"><link rel="prev" href="/applications/tabix.html"><link rel="next" href="/applications/xlsx.html"><link rel="canonical" href="/applications/VEP.html"><link rel="icon" type="image/svg+xml" href="/assets/images/favicon.svg"><link rel="icon" type="image/png" href="/assets/images/favicon-16x16.png" sizes="16x16"> <link rel="icon" type="image/png" href="/assets/images/favicon-32x32.png" sizes="32x32"> <link rel="icon" type="image/png" href="/assets/images/favicon-96x96.png" sizes="96x96"><link rel="mask-icon" href="/assets/images/favicon.svg" color="#2980b9"><link rel="apple-touch-icon" href="/assets/images/apple-touch-icon-300x300.jpg"> <link rel="stylesheet" href="/assets/css/theme.css"><script> window.ui = { title: "csd3", baseurl: "", i18n: { search_results: "Search Results", search_results_found: "Search finished, found # page(s) matching the search query.", search_results_not_found: "Your search did not match any documents, please make sure that all characters are spelled correctly!" } }; </script> </head> <body class="container"><div class="sidebar-wrap overflow-hidden"> <div class="sidebar height-full overflow-y-scroll overflow-x-hidden"> <div class="header d-flex flex-column p-3 text-center"> <div class="title pb-1"> <a class="h4 no-underline py-1 px-2 rounded-1" href="/" title="A collection of CSD3 notes"> <i class="fa fa-home"></i> csd3 </a> </div> <span class="version"></span> <form class="search pt-2" action="/search.html" method="get" autocomplete="off"> <input class="form-control input-block input-sm" type="text" name="q" placeholder="Search docs..."> </form> </div> <div class="toctree py-2" data-spy="affix" role="navigation" aria-label="main navigation"> <ul> </ul> <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/systems/"> The system </a><ul> <li class="toc level-1 " data-sort="1" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/system.html">1. System</a> </li> <li class="toc level-1 " data-sort="2" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/login.html">2. Login</a> </li> <li class="toc level-1 " data-sort="3" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/directories.html">3. Directories</a> </li> <li class="toc level-1 " data-sort="4" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/policies.html">4. Policies</a> </li> <li class="toc level-1 " data-sort="5" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/email.html">5. email</a> </li> <li class="toc level-1 " data-sort="6" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/software.html">6. Software</a> </li> <li class="toc level-1 " data-sort="7" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/training.html">7. Training</a> </li> <li class="toc level-1 " data-sort="8" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/acknowledgement.html">8. Acknowledgement</a> </li> <li class="toc level-1 " data-sort="8" data-level="1"> <a class="d-flex flex-items-baseline " href="/systems/contacts.html">8. Contacts</a> </li></ul> <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/cardio/"> cardio </a><ul> </ul> <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/applications/"> Applications </a><ul> <li class="toc level-1 " data-sort="1" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/ABCtoolbox.html">1. ABCtoolbox</a> </li> <li class="toc level-1 " data-sort="2" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/AEGIS.html">2. AEGIS</a> </li> <li class="toc level-1 " data-sort="3" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/akt.html">3. akt</a> </li> <li class="toc level-1 " data-sort="4" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/ANNOVAR.html">4. ANNOVAR</a> </li> <li class="toc level-1 " data-sort="5" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/bcftools.html">5. bcftools</a> </li> <li class="toc level-1 " data-sort="6" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/bedops.html">6. bedops</a> </li> <li class="toc level-1 " data-sort="7" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/bgenix.html">7. bgenix</a> </li> <li class="toc level-1 " data-sort="8" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/brms.html">8. R/brms</a> </li> <li class="toc level-1 " data-sort="9" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/CaVEMaN.html">9. CaVEMaN</a> </li> <li class="toc level-1 " data-sort="10" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/citeproc.html">10. citeproc</a> </li> <li class="toc level-1 " data-sort="11" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/DosageConverter.html">11. DosageConverter</a> </li> <li class="toc level-1 " data-sort="12" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/EasyQC.html">12. EasyQC</a> </li> <li class="toc level-1 " data-sort="13" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/gnn.html">13. R/gnn</a> </li> <li class="toc level-1 " data-sort="14" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/gsutil.html">14. gsutil</a> </li> <li class="toc level-1 " data-sort="15" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/gwas2vcf.html">15. gwas2vcf</a> </li> <li class="toc level-1 " data-sort="16" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/hail.html">16. hail</a> </li> <li class="toc level-1 " data-sort="17" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/HESS.html">17. HESS</a> </li> <li class="toc level-1 " data-sort="18" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/KentUtils.html">18. KentUtils</a> </li> <li class="toc level-1 " data-sort="19" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/lapack.html">19. lapack</a> </li> <li class="toc level-1 " data-sort="20" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/LDlinkR.html">20. R/LDlinkR</a> </li> <li class="toc level-1 " data-sort="21" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/LEMMA.html">21. LEMMA</a> </li> <li class="toc level-1 " data-sort="22" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/magick.html">22. R/magick</a> </li> <li class="toc level-1 " data-sort="23" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/MetaXcan.html">23. MetaXcan</a> </li> <li class="toc level-1 " data-sort="24" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/pandoc.html">24. pandoc</a> </li> <li class="toc level-1 " data-sort="25" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/pca_projection.html">25. PCA projection</a> </li> <li class="toc level-1 " data-sort="26" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/PhenoScanner.html">26. R/PhenoScanner</a> </li> <li class="toc level-1 " data-sort="27" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/PKI.html">27. R/PKI</a> </li> <li class="toc level-1 " data-sort="28" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/plumber.html">28. R/plumber</a> </li> <li class="toc level-1 " data-sort="29" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/polyphen-2.html">29. polyphen-2</a> </li> <li class="toc level-1 " data-sort="30" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/poppler.html">30. poppler</a> </li> <li class="toc level-1 " data-sort="31" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/protolite.html">31. R/protolite</a> </li> <li class="toc level-1 " data-sort="32" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/PRSice.html">32. PRSice</a> </li> <li class="toc level-1 " data-sort="33" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/PRSoS.html">33. PRSoS</a> </li> <li class="toc level-1 " data-sort="34" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/pspp.html">34. pspp</a> </li> <li class="toc level-1 " data-sort="35" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/qpdf.html">35. qpdf</a> </li> <li class="toc level-1 " data-sort="36" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/regenie.html">36. regenie</a> </li> <li class="toc level-1 " data-sort="37" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/rgdal.html">37. R/rgdal</a> </li> <li class="toc level-1 " data-sort="38" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/rgeos.html">38. R/rgeos</a> </li> <li class="toc level-1 " data-sort="39" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/Rhdf5lib.html">39. R/Rhdf5lib</a> </li> <li class="toc level-1 " data-sort="40" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/rjags.html">40. R/rjags</a> </li> <li class="toc level-1 " data-sort="41" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/rJava.html">41. R/rJava</a> </li> <li class="toc level-1 " data-sort="42" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/R.html">42. R</a> </li> <li class="toc level-1 " data-sort="43" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/rstan.html">43. R/rstan</a> </li> <li class="toc level-1 " data-sort="44" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/RStudio.html">44. RStudio</a> </li> <li class="toc level-1 " data-sort="45" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/SAIGE.html">45. R/SAIGE 0.36.6 and 0.39.2</a> </li> <li class="toc level-1 " data-sort="46" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/sf.html">46. R/sf</a> </li> <li class="toc level-1 " data-sort="47" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/shapeit3.html">47. shapeit3</a> </li> <li class="toc level-1 " data-sort="48" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/SMR.html">48. SMR</a> </li> <li class="toc level-1 " data-sort="49" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/snakemake.html">49. snakemake</a> </li> <li class="toc level-1 " data-sort="50" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/snpnet.html">50. R/snpnet</a> </li> <li class="toc level-1 " data-sort="51" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/SNPTEST.html">51. SNPTEST</a> </li> <li class="toc level-1 " data-sort="52" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/sojo.html">52. R/sojo</a> </li> <li class="toc level-1 " data-sort="53" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/tabix.html">53. tabix</a> </li> <li class="toc level-1 current" data-sort="54" data-level="1"> <a class="d-flex flex-items-baseline current" href="/applications/VEP.html">54. VEP</a> </li> <li class="toc level-1 " data-sort="55" data-level="1"> <a class="d-flex flex-items-baseline " href="/applications/xlsx.html">55. R/xlsx</a> </li></ul> <a class="caption d-block text-uppercase no-wrap px-2 py-0" href="/writing/"> Writing Related </a><ul> <li class="toc level-1 " data-sort="1" data-level="1"> <a class="d-flex flex-items-baseline " href="/writing/front-matter.html">1. Front matter</a> </li> <li class="toc level-1 " data-sort="2" data-level="1"> <a class="d-flex flex-items-baseline " href="/writing/list.html">2. List directory</a> </li> <li class="toc level-1 " data-sort="3" data-level="1"> <a class="d-flex flex-items-baseline " href="/writing/mathjax.html">3. Mathjax</a> </li> <li class="toc level-1 " data-sort="4" data-level="1"> <a class="d-flex flex-items-baseline " href="/writing/admonitions.html">4. Admonition card</a> </li> <li class="toc level-1"> <a class="d-flex flex-items-baseline" href="/writing/test/"> Test Documentation </a><ul> <li class="toc level-2 " data-sort="1" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/markdown.html">1. Markdown Elements</a> </li> <li class="toc level-2 " data-sort="2" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/toasts.html">2. Toasts Card</a> </li> <li class="toc level-2 " data-sort="3" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/codes.html">3. Code Blocks</a> </li> <li class="toc level-2 " data-sort="4" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/mermaid.html">4. Mermaid Test</a> </li> <li class="toc level-2 " data-sort="5" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/emoji.html">5. Emoji Test</a> </li> <li class="toc level-2 " data-sort="6" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/gist.html">6. Gist Test</a> </li> <li class="toc level-2 " data-sort="7" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/avatar.html">7. Avatar Test</a> </li> <li class="toc level-2 " data-sort="8" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/mentions.html">8. Mentions Test</a> </li> <li class="toc level-2 " data-sort="9" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/fonts.html">9. Fonts Test</a> </li> <li class="toc level-2 " data-sort="10" data-level="2"> <a class="d-flex flex-items-baseline " href="/writing/test/utilities.html">10. Primer Utilities Test</a> </li></ul> </li></ul> </div> </div> </div> <div class="content-wrap"> <div class="header d-flex flex-justify-between p-2 hide-lg hide-xl" aria-label="top navigation"> <button id="toggle" aria-label="Toggle menu" class="btn-octicon p-2 m-0 text-white" type="button"> <i class="fa fa-bars"></i> </button> <div class="title flex-1 d-flex flex-justify-center"> <a class="h4 no-underline py-1 px-2 rounded-1" href="/">csd3</a> </div> </div> <div class="content p-3 p-sm-5"> <div class="navigation-top d-flex flex-justify-between"> <ul class="breadcrumb" role="navigation" aria-label="breadcrumbs navigation"> <li class="breadcrumb-item"> <a class="no-underline" href="/" title="/"> <i class="fa fa-home"></i> </a> </li><li class="breadcrumb-item" ><a href="/applications/">applications</a></li><li class="breadcrumb-item" aria-current="page">VEP.md</li></ul> <a class="edit" href="https://github.com/cambridge-ceu/csd3/edit/gh-pages/applications/VEP.md" title="Edit on GitHub" rel="noreferrer" target="_blank"> <i class="fa fa-edit"></i> </a> </div> <hr> <div role="main" itemscope="itemscope" itemtype="https://schema.org/Article"> <div class="markdown-body" itemprop="articleBody"> <h1 id="vep">VEP</h1> <p>Official page: <a href="https://www.ensembl.org/info/docs/tools/vep/index.html">https://www.ensembl.org/info/docs/tools/vep/index.html</a> (<a href="https://www.ensembl.org/Tools/VEP">web interface</a>).</p> <p>Detailed instructions for installation are available from here,</p> <p>http://www.ensembl.org/info/docs/tools/vep/script/vep_download.html#installer.</p> <p>There are several possible ways to install under csd3: GitHub, R and docker.</p> <h3 id="-github-">— GitHub —</h3> <p>GitHub Page: <a href="https://github.com/Ensembl/ensembl-vep">https://github.com/Ensembl/ensembl-vep</a>.</p> <p>The ease with this option lies with GitHub in that updates can simply be made with <code class="language-plaintext highlighter-rouge notranslate">git pull</code> to an exisiting release.</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="nb">cd</span> <span class="nv">$HPC_WORK</span>
git clone https://github.com/Ensembl/ensembl-vep.git
<span class="nb">cd </span>ensembl-vep
<span class="nb">mkdir</span> .vep
<span class="nb">ln</span> <span class="nt">-sf</span> <span class="nv">$HPC_WORK</span>/ensembl-vep/.vep <span class="nv">$HOME</span>/.vep
module load htslib/1.4
perl INSTALL.pl
<span class="c"># set up symbolic links to the executables</span>
<span class="k">for </span>f <span class="k">in </span>convert_cache.pl filter_vep haplo variant_recoder vep<span class="p">;</span>
    <span class="k">do </span><span class="nb">ln</span> <span class="nt">-sf</span> <span class="nv">$HPC_WORK</span>/ensembl-vep/<span class="nv">$f</span> <span class="nv">$HPC_WORK</span>/bin/<span class="nv">$f</span><span class="p">;</span> <span class="k">done
</span>vep <span class="nt">-i</span> examples/homo_sapiens_GRCh37.vcf <span class="nt">-o</span> examples/homo_sapiens_GRCh37.txt <span class="se">\</span>
    <span class="nt">--force_overwrite</span> <span class="nt">--offline</span> <span class="nt">--pick</span> <span class="nt">--symbol</span>
</code></pre> </div></div> <p>Note in particular that by default, the cache files will be installed at $HOME which would exceed the quota (&lt;40GB) of an ordinary user, and as before the destination was redirected. The setup above facilitates storage of cache files, FASTA files and plugins.</p> <blockquote> <p>The FASTA file should be automatically detected by the VEP when using –cache or –offline. If it is not, use "–fasta $HOME/.vep/homo_sapiens/98_GRCh37/Homo_sapiens.GRCh37.75.dna.primary_assembly.fa"</p> </blockquote> <p>Without the htslib/1.4 module, the <code class="language-plaintext highlighter-rouge notranslate">--NO_HTSLIB</code> option is needed but "Cannot use format gff without Bio::DB::HTS::Tabix module installed". Bio::DB:HTS is in https://github.com/Ensembl/Bio-DB-HTS and change can be made to the <code class="language-plaintext highlighter-rouge notranslate">Makefile</code> of htslibs for a desired location, to be used by <code class="language-plaintext highlighter-rouge notranslate">Build.PL</code> via its command line parameters.</p> <p>It is notable that VEP accepts compress (.gz) input. It is worthwhile to check for the VEP plugins: <a href="https://github.com/Ensembl/VEP_plugins">https://github.com/Ensembl/VEP_plugins</a>. For instance, to enable the PolyPhen_SIFT plugin we first generate the database.</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>vep <span class="nt">--database</span> <span class="nt">--port</span> 3337 <span class="nt">--format</span> guess <span class="nt">--plugin</span> PolyPhen_SIFT,create_db<span class="o">=</span>1
</code></pre> </div></div> <p>One may wish to skip the comments (lines started with ##) in processing of the output, e.g., in R,</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="nb">export </span><span class="nv">skips</span><span class="o">=</span><span class="si">$(</span><span class="nb">grep</span> <span class="s1">'##'</span> examples/homo_sapiens_GRCh37.txt | <span class="nb">wc</span> <span class="nt">-l</span><span class="si">)</span>
R <span class="nt">--no-save</span> <span class="nt">-q</span> <span class="o">&lt;&lt;</span><span class="no">END</span><span class="sh">
  vo &lt;- read.delim("examples/homo_sapiens_GRCh37.txt",skip=as.numeric(Sys.getenv("skips")))
  head(vo)
</span><span class="no">END
</span></code></pre> </div></div> <p>allowing for variable number of lines given various command-line options to be skipped to have</p> <div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>  X.Uploaded_variation    Location Allele            Gene         Feature
1          rs116645811 21:26960070      A ENSG00000154719 ENST00000307301
2            rs1135638 21:26965148      A ENSG00000154719 ENST00000307301
3              rs10576 21:26965172      C ENSG00000154719 ENST00000307301
4            rs1057885 21:26965205      C ENSG00000154719 ENST00000307301
5          rs116331755 21:26976144      G ENSG00000154719 ENST00000307301
6            rs7278168 21:26976222      T ENSG00000154719 ENST00000307301
  Feature_type        Consequence cDNA_position CDS_position Protein_position
1   Transcript   missense_variant          1043         1001              334
2   Transcript synonymous_variant           939          897              299
3   Transcript synonymous_variant           915          873              291
4   Transcript synonymous_variant           882          840              280
5   Transcript synonymous_variant           426          384              128
6   Transcript synonymous_variant           348          306              102
  Amino_acids  Codons Existing_variation
1         T/M aCg/aTg                  -
2           G ggC/ggT                  -
3           P ccA/ccG                  -
4           V gtA/gtG                  -
5           L ctT/ctC                  -
6           K aaG/aaA                  -
                                                                     Extra
1 IMPACT=MODERATE;STRAND=-1;SYMBOL=MRPL39;SYMBOL_SOURCE=HGNC;HGNC_ID=14027
2      IMPACT=LOW;STRAND=-1;SYMBOL=MRPL39;SYMBOL_SOURCE=HGNC;HGNC_ID=14027
3      IMPACT=LOW;STRAND=-1;SYMBOL=MRPL39;SYMBOL_SOURCE=HGNC;HGNC_ID=14027
4      IMPACT=LOW;STRAND=-1;SYMBOL=MRPL39;SYMBOL_SOURCE=HGNC;HGNC_ID=14027
5      IMPACT=LOW;STRAND=-1;SYMBOL=MRPL39;SYMBOL_SOURCE=HGNC;HGNC_ID=14027
6      IMPACT=LOW;STRAND=-1;SYMBOL=MRPL39;SYMBOL_SOURCE=HGNC;HGNC_ID=14027
&gt;
</code></pre> </div></div> <p>When a particular release really works well on the system, it is possible to install it, e.g.,</p> <div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>git checkout release/98
perl INSTALL.pl
</code></pre> </div></div> <p>for release 98 instead of the latest from <code class="language-plaintext highlighter-rouge notranslate">git pull</code>.</p> <p>It could be useful to filter VEP output, see https://www.ensembl.org/info/docs/tools/vep/script/vep_filter.html.</p> <h3 id="nearest-gene"><strong>Nearest gene</strong></h3> <p>This can produces error message</p> <div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>-------------------- EXCEPTION --------------------
MSG: ERROR: --nearest requires Set::IntervalTree perl module to be installed
</code></pre> </div></div> <p>and we get around with</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>perl <span class="nt">-MCPAN</span> <span class="nt">-e</span> shell
<span class="nb">install </span>Set::IntervalTree
</code></pre> </div></div> <p>which will enable <code class="language-plaintext highlighter-rouge notranslate">--nearest gene</code>.</p> <h3 id="annotation-in-chunks"><strong>Annotation in chunks</strong></h3> <p>A toy example, following http://www.ensembl.org/info/docs/tools/vep/script/vep_other.html#faster, is given as follows,</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="nb">cd </span>examples
bgzip <span class="nt">-f</span> homo_sapiens_GRCh37.vcf
tabix <span class="nt">-Cf</span> homo_sapiens_GRCh37.vcf.gz
tabix <span class="nt">-h</span>  homo_sapiens_GRCh37.vcf.gz 22:50616005-50616006 | <span class="se">\</span>
vep <span class="nt">--cache</span> <span class="nt">--fork</span> 4 <span class="nt">--port</span> 3337 <span class="nt">--format</span> vcf <span class="nt">-o</span> - <span class="nt">--tab</span> <span class="nt">--no_stats</span> | <span class="se">\</span>
<span class="nb">grep</span> <span class="nt">-v</span> <span class="s1">'##'</span>
<span class="nb">cd</span> -
</code></pre> </div></div> <p>from this we could propagate the idea for autosomes in chunks as follows,</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="nb">export </span><span class="nv">ref</span><span class="o">=</span>~/rds/post_qc_data/interval/reference_files/genetic/reference_files_genotyped_imputed/
<span class="nb">export </span><span class="nv">chunk_size</span><span class="o">=</span>10000
<span class="nb">seq </span>22 | <span class="se">\</span>
parallel <span class="nt">-j1</span> <span class="nt">--env</span> ref <span class="nt">-C</span><span class="s1">' '</span> <span class="s1">'
  export n=$(awk "END{print NR-1}" ${ref}/impute_{}_interval.snpstats)
  export g=$(expr ${n} / ${chunk_size})
  export s=$(expr $n - $(( $g * $chunk_size)))
  (
    for i in $(seq ${g}); do
      (
        awk "BEGIN{print \"##fileformat=VCFv4.0\"}"
        awk -vOFS="\t" "BEGIN{print \"#CHROM\",\"POS\",\"ID\",\"REF\",\"ALT\",\"QUAL\",\"FILTER\",\"INFO\"}"
        (
          sed "1d" ${ref}/impute_{}_interval.snpstats | \
          awk -v i=${i} -v chunk_size=${chunk_size} -v OFS="\t" "NR==(i-1)*chunk_size+1,NR==i*chunk_size {
              if(\$1==\".\") \$1=\$3+0 \":\" \$4 \"_\" \$5 \"/\" \$6; print \$3+0,\$4,\$1,\$5,\$6,\".\",\".\",\$19}"
          if [ ${s} -gt 0 ] &amp;&amp;  [ ${i} -eq ${g} ]; then
             sed "1d" ${ref}/impute_{}_interval.snpstats | \
             awk -v i=${i} -v chunk_size=${chunk_size} -v OFS="\t" -v n=${n} "NR==i*chunk_size+1,NR==n {
                 if(\$1==\".\") \$1=\$3+0 \":\" \$4 \"_\" \$5 \"/\" \$6; print \$3+0,\$4,\$1,\$5,\$6,\".\",\".\",\$19}"
          fi
        )
      ) | \
      vep  --cache --offline --format vcf -o - --tab --pick --no_stats  \
           --species homo_sapiens --assembly GRCh37 --port 3337 | \
      (
        if [ ${i} -eq 1 ]; then cat; else grep -v "#"; fi
      )
    done
  ) | \
  gzip -f &gt; work/INTERVAL-{}.vep.gz
'</span>
</code></pre> </div></div> <p>Note we use information from <code class="language-plaintext highlighter-rouge notranslate">.snpstats</code> files at location <code class="language-plaintext highlighter-rouge notranslate">ref</code> to build input in vcf format on the fly and feed into VEP. For instance <code class="language-plaintext highlighter-rouge notranslate">impute_22_interval.snpstats</code> contains the following lines,</p> <div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>SNPID	RSID	chromosome	position	A_allele	B_allele	minor_allele	major_allele	AA	AB	BB	AA_calls	AB_calls	BB_calls	MAF	HWE	missing	missing_calls	information
rs587697622	rs587697622	22	16050075	A	G	G	A	43058	1.2499	0	43058	1	0	1.4514e-05	-0	0	0	0.81003
rs587755077	rs587755077	22	16050115	G	A	A	G	42485	572.42	1.7625	42105	306	1	0.0066878	0.14341	3.5437e-09	0.015026	0.68672
rs587654921	rs587654921	22	16050213	C	T	T	C	42848	211.12	0.16248	42746	124	0	0.0024553	-0	1.4175e-09	0.0043893	0.67168
rs587712275	rs587712275	22	16050319	C	T	T	C	43022	36.998	0	43022	23	0	0.00042962	-0	0	0.00032514	0.69071
</code></pre> </div></div> <h3 id="ensembl-synonym-translation"><strong>ENSEMBL-synonym translation</strong></h3> <p>The ENSEMBL-synonym translation is useful to check for the feature types – in the case of ENSG00000160712 (IL6R) we found ENST00000368485 and ENST00000515190, we do</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>wget http://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/ensemblToGeneName.txt.gz
zgrep <span class="nt">-e</span> ENST00000368485 <span class="nt">-e</span> ENST00000515190 ensemblToGeneName.txt.gz
</code></pre> </div></div> <p>giving</p> <div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>ENST00000368485 IL6R
ENST00000515190 IL6R
</code></pre> </div></div> <p>though this could also be furnished with R/biomaRt as follows,</p> <div class="language-r highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">biomaRt</span><span class="p">)</span><span class="w">
</span><span class="n">ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">useMart</span><span class="p">(</span><span class="s2">"ensembl"</span><span class="p">,</span><span class="w"> </span><span class="n">dataset</span><span class="o">=</span><span class="s2">"hsapiens_gene_ensembl"</span><span class="p">,</span><span class="w"> </span><span class="n">host</span><span class="o">=</span><span class="s2">"grch37.ensembl.org"</span><span class="p">,</span><span class="w"> </span><span class="n">path</span><span class="o">=</span><span class="s2">"/biomart/martservice"</span><span class="p">)</span><span class="w">
</span><span class="n">attr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">listAttributes</span><span class="p">(</span><span class="n">ensembl</span><span class="p">)</span><span class="w">
</span><span class="n">g</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'ensembl_gene_id'</span><span class="p">,</span><span class="w"> </span><span class="s1">'chromosome_name'</span><span class="p">,</span><span class="w"> </span><span class="s1">'start_position'</span><span class="p">,</span><span class="w"> </span><span class="s1">'end_position'</span><span class="p">,</span><span class="w"> </span><span class="s1">'description'</span><span class="p">,</span><span class="w"> </span><span class="s1">'hgnc_symbol'</span><span class="p">)</span><span class="w">
</span><span class="n">t</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'ensembl_transcript_id'</span><span class="p">,</span><span class="w"> </span><span class="s1">'transcription_start_site'</span><span class="p">,</span><span class="w"> </span><span class="s1">'transcript_start'</span><span class="p">,</span><span class="w"> </span><span class="s1">'transcript_end'</span><span class="p">)</span><span class="w">
</span><span class="n">u</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"uniprotswissprot"</span><span class="w">
</span><span class="n">gtu</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">g</span><span class="p">,</span><span class="n">t</span><span class="p">,</span><span class="n">u</span><span class="p">),</span><span class="w"> </span><span class="n">mart</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ensembl</span><span class="p">)</span><span class="w">
</span></code></pre> </div></div> <p>For ENSEMBL genes, R/grex is likely to work though it was developed for other purpose, e.g.,</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>R <span class="nt">-q</span> <span class="nt">-e</span> <span class="s2">"grex::grex(</span><span class="se">\"</span><span class="s2">ENSG00000160712</span><span class="se">\"</span><span class="s2">)"</span>
</code></pre> </div></div> <p>giving</p> <div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>       ensembl_id entrez_id hgnc_symbol              hgnc_name cyto_loc
1 ENSG00000160712      3570        IL6R interleukin 6 receptor   1q21.3
  uniprot_id   gene_biotype
1     A0N0L5 protein_coding
</code></pre> </div></div> <p>The aligment of ENSG, ENST, ENSP is <code class="language-plaintext highlighter-rouge notranslate">ensGtp.txt.gz</code> at the same UCSC directory above.</p> <p>Other possibilities include UCSC/ensembl MySQL server, e.g.,</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>mysql  <span class="nt">--user</span><span class="o">=</span>genome <span class="nt">--host</span><span class="o">=</span>genome-mysql.cse.ucsc.edu <span class="nt">-A</span> <span class="nt">-D</span> hg19 <span class="o">&lt;&lt;</span><span class="no">END</span><span class="sh">
select distinct G.gene,N.value from ensGtp as G, ensemblToGeneName as N where
   G.transcript=N.name and
   G.gene in ("ENSG00000160712");
</span><span class="no">END
</span></code></pre> </div></div> <p>and</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>mysql <span class="nt">-h</span> ensembldb.ensembl.org <span class="nt">--port</span> 5306  <span class="nt">-u</span> anonymous <span class="nt">-D</span> homo_sapiens_core_64_37 <span class="nt">-A</span> <span class="o">&lt;&lt;</span><span class="no">END</span><span class="sh">
select distinct
   G.stable_id,
   S.synonym
from
  gene_stable_id as G,
  object_xref as OX,
  external_synonym as S,
  xref as X ,
  external_db as D
where
  D.external_db_id=X.external_db_id and
  X.xref_id=S.xref_id and
  OX.xref_id=X.xref_id and
  OX.ensembl_object_type="Gene" and
  G.gene_id=OX.ensembl_id and
  G.stable_id in ("ENSG00000160712");
</span><span class="no">END
</span></code></pre> </div></div> <p>according to https://www.biostars.org/p/14367/.</p> <p>Perhaps it is the easiest to use <code class="language-plaintext highlighter-rouge notranslate">gprofiler2</code>, i.e.,</p> <div class="language-r highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="n">gprofiler2</span><span class="o">::</span><span class="n">gconvert</span><span class="p">(</span><span class="s2">"ENSG00000164761"</span><span class="p">)</span><span class="w">
</span></code></pre> </div></div> <p>giving</p> <div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>  input_number input target_number          target name                                               description                           namespace
1            1  IL6R           1.1 ENSG00000160712 IL6R interleukin 6 receptor [Source:HGNC Symbol;Acc:HGNC:6019] ENTREZGENE,HGNC,UNIPROT_GN,WIKIGENE
</code></pre> </div></div> <h3 id="clinvar"><strong>clinvar</strong></h3> <p>The local installation enables considerable flexibilty, and the following example, using GRCh37 assembly, is based on <a href="https://www.ensembl.org/info/docs/tools/vep/script/vep_custom.html#custom_options">https://www.ensembl.org/info/docs/tools/vep/script/vep_custom.html#custom_options</a>.</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="c"># Compressed VCF file</span>
curl ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh37/clinvar.vcf.gz <span class="nt">-o</span> clinvar_GRCh37.vcf.gz
<span class="c"># Index file</span>
curl ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh37/clinvar.vcf.gz.tbi <span class="nt">-o</span> clinvar_GRCh37.vcf.gz.tbi
</code></pre> </div></div> <p>Information is gathered from the header of the VCF file,</p> <table> <thead> <tr> <th>ClinVar Variation ID</th> <th>Description</th> <th> </th> </tr> </thead> <tbody> <tr> <td>AF_ESP</td> <td>allele frequencies from GO-ESP</td> <td> </td> </tr> <tr> <td>AF_EXAC</td> <td>allele frequencies from ExAC</td> <td> </td> </tr> <tr> <td>AF_TGP</td> <td>allele frequencies from TGP</td> <td> </td> </tr> <tr> <td>ALLELEID</td> <td>the ClinVar Allele ID</td> <td> </td> </tr> <tr> <td>CLNDN</td> <td>ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB</td> <td> </td> </tr> <tr> <td>CLNDNINCL</td> <td>For included Variant : ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB</td> <td> </td> </tr> <tr> <td>CLNDISDB</td> <td>Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN</td> <td> </td> </tr> <tr> <td>CLNDISDBINCL</td> <td>For included Variant: Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN</td> <td> </td> </tr> <tr> <td>CLNHGVS</td> <td>Top-level (primary assembly, alt, or patch) HGVS expression.</td> <td> </td> </tr> <tr> <td>CLNREVSTAT</td> <td>ClinVar review status for the Variation ID</td> <td> </td> </tr> <tr> <td>CLNSIG</td> <td>Clinical significance for this single variant</td> <td> </td> </tr> <tr> <td>CLNSIGCONF</td> <td>Conflicting clinical significance for this single variant</td> <td> </td> </tr> <tr> <td>CLNSIGINCL</td> <td>Clinical significance for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:clinical significance.</td> <td> </td> </tr> <tr> <td>CLNVC</td> <td>Variant type</td> <td> </td> </tr> <tr> <td>CLNVCSO</td> <td>Sequence Ontology id for variant type</td> <td> </td> </tr> <tr> <td>CLNVI</td> <td>the variant's clinical sources reported as tag-value pairs of database and variant identifier</td> <td> </td> </tr> <tr> <td>DBVARID</td> <td>nsv accessions from dbVar for the variant</td> <td> </td> </tr> <tr> <td>GENEINFO</td> <td>Gene(s) for the variant reported as gene symbol:gene id. The gene symbol and id are delimited by a colon (:) and each pair is delimited by a vertical bar (</td> <td>)</td> </tr> <tr> <td>MC</td> <td>comma separated list of molecular consequence in the form of Sequence Ontology ID</td> <td>molecular_consequence</td> </tr> <tr> <td>ORIGIN</td> <td>Allele origin. One or more of the following values may be added: 0 - unknown; 1 - germline; 2 - somatic ; 4 - inherited; 8 - paternal; 16 - maternal; 32 - de-novo; 64 - biparental; 128 - uniparental; 256 - not-tested; 512 - tested-inconclusive; 1073741824 - other</td> <td> </td> </tr> <tr> <td>RS</td> <td>dbSNP ID (i.e. rs number)</td> <td> </td> </tr> <tr> <td>SSR</td> <td>Variant Suspect Reason Codes. One or more of the following values may be added: 0 - unspecified, 1 - Paralog, 2 - byEST, 4 - oldAlign, 8 - Para_EST, 16 - 1kg_failed, 1024 - other</td> <td> </td> </tr> </tbody> </table> <p>We now query rs2228145,</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>vep <span class="nt">--id</span> <span class="s2">"1 154426970 154426970 A/C 1"</span> <span class="nt">--species</span> homo_sapiens <span class="nt">-o</span> rs2228145 <span class="nt">--cache</span> <span class="nt">--offline</span> <span class="nt">--force_overwrite</span> <span class="se">\</span>
    <span class="nt">--assembly</span> GRCh37 <span class="nt">--custom</span> clinvar_GRCh37.vcf.gz,ClinVar,vcf,exact,0,CLNSIG,CLNREVSTAT,CLNDN
</code></pre> </div></div> <p>which gives</p> <div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>#Uploaded_variation	Location	Allele	Gene	Feature	Feature_type	Consequence	cDNA_position	CDS_position	Protein_position	Amino_acids	Codons	Existing_variation	Extra
1_154426970_A/C	1:154426970	C	ENSG00000160712	ENST00000344086	Transcript	intron_variant	-	-	-	-	-	-	IMPACT=MODIFIER;STRAND=1;ClinVar=14660;ClinVar_CLNDN=Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus;ClinVar_CLNREVSTAT=no_assertion_criteria_provided;ClinVar_CLNSIG=association;ClinVar_FILTER=.
1_154426970_A/C	1:154426970	C	ENSG00000160712	ENST00000368485	Transcript	missense_variant	1510	1073	358	D/A	gAt/gCt	-	IMPACT=MODERATE;STRAND=1;ClinVar=14660;ClinVar_CLNDN=Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus;ClinVar_CLNREVSTAT=no_assertion_criteria_provided;ClinVar_CLNSIG=association;ClinVar_FILTER=.
1_154426970_A/C	1:154426970	C	ENSG00000160712	ENST00000476006	Transcript	downstream_gene_variant	-	-	-	-	-	-	IMPACT=MODIFIER;DISTANCE=4515;STRAND=1;FLAGS=cds_start_NF,cds_end_NF;ClinVar=14660;ClinVar_CLNDN=Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus;ClinVar_CLNREVSTAT=no_assertion_criteria_provided;ClinVar_CLNSIG=association;ClinVar_FILTER=.
1_154426970_A/C	1:154426970	C	ENSG00000160712	ENST00000502679	Transcript	non_coding_transcript_exon_variant	386	-	-	-	-	-	IMPACT=MODIFIER;STRAND=1;ClinVar=14660;ClinVar_CLNDN=Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus;ClinVar_CLNREVSTAT=no_assertion_criteria_provided;ClinVar_CLNSIG=association;ClinVar_FILTER=.
1_154426970_A/C	1:154426970	C	ENSG00000160712	ENST00000507256	Transcript	non_coding_transcript_exon_variant	271	-	-	-	-	-	IMPACT=MODIFIER;STRAND=1;ClinVar=14660;ClinVar_CLNDN=Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus;ClinVar_CLNREVSTAT=no_assertion_criteria_provided;ClinVar_CLNSIG=association;ClinVar_FILTER=.
1_154426970_A/C	1:154426970	C	ENSG00000160712	ENST00000515190	Transcript	missense_variant	481	482	161	D/A	gAt/gCt	-	IMPACT=MODERATE;STRAND=1;FLAGS=cds_start_NF,cds_end_NF;ClinVar=14660;ClinVar_CLNDN=Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus;ClinVar_CLNREVSTAT=no_assertion_criteria_provided;ClinVar_CLNSIG=association;ClinVar_FILTER=.
</code></pre> </div></div> <p>A <a href="/applications/files/rs2228145_summary.html">HTML summary</a> (somehow the web browser may not display the embedded figures) is also available. The <code class="language-plaintext highlighter-rouge notranslate">Extra</code> column looks clumsy and we could add the <code class="language-plaintext highlighter-rouge notranslate">--tab</code> option to generate a tab-delimited output.</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>vep <span class="nt">--id</span> <span class="s2">"1 154426970 154426970 A/C 1"</span> <span class="nt">--species</span> homo_sapiens <span class="nt">-o</span> rs2228145 <span class="nt">--cache</span> <span class="nt">--offline</span> <span class="nt">--force_overwrite</span> <span class="se">\</span>
    <span class="nt">--custom</span> clinvar_GRCh37.vcf.gz,ClinVar,vcf,exact,0,CLNSIG,CLNREVSTAT,CLNDN <span class="nt">--tab</span> <span class="se">\</span>
    <span class="nt">--fields</span> Uploaded_variation,Gene,Consequence,ClinVar_CLNSIG,ClinVar_CLNREVSTAT,ClinVar_CLNDN
</code></pre> </div></div> <p>to give neatly</p> <div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>#Uploaded_variation	Gene	Consequence	ClinVar_CLNSIG	ClinVar_CLNREVSTAT	ClinVar_CLNDN
1_154426970_A/C	ENSG00000160712	intron_variant	association	no_assertion_criteria_provided	Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus
1_154426970_A/C	ENSG00000160712	missense_variant	association	no_assertion_criteria_provided	Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus
1_154426970_A/C	ENSG00000160712	downstream_gene_variant	association	no_assertion_criteria_provided	Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus
1_154426970_A/C	ENSG00000160712	non_coding_transcript_exon_variant	association	no_assertion_criteria_provided	Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus
1_154426970_A/C	ENSG00000160712	non_coding_transcript_exon_variant	association	no_assertion_criteria_provided	Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus
1_154426970_A/C	ENSG00000160712	missense_variant	association	no_assertion_criteria_provided	Interleukin_6,_serum_level_of,_quantitative_trait_locus|Soluble_interleukin-6_receptor,_serum_level_of,_quantitative_trait_locus
</code></pre> </div></div> <h3 id="dbnsfp"><strong>dbNSFP</strong></h3> <p>Web page: <a href="https://sites.google.com/site/jpopgen/dbNSFP">https://sites.google.com/site/jpopgen/dbNSFP</a>.</p> <p>This is set up as follows,</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>wget ftp://dbnsfp:dbnsfp@dbnsfp.softgenetics.com/dbNSFP4.1a.zip
unzip dbNSFP4.1a.zip <span class="nt">-d</span> dbNSFP4.1a
<span class="nb">cd </span>dbNSFP4.1a
<span class="c"># code for 4.0a</span>
<span class="c"># zcat dbNSFP4.0a_variant.chr1.gz | head -n1 &gt; h</span>
<span class="c"># zgrep -h -v ^#chr dbNSFP4.0a_variant.chr* | sort -k1,1 -k2,2n - | cat h - | bgzip -c &gt; dbNSFP4.0a.gz</span>
<span class="c"># efficient version</span>
<span class="o">(</span>
  zcat dbNSFP4.1a_variant.chr1.gz | <span class="nb">head</span> <span class="nt">-n1</span>
  <span class="nb">export </span><span class="nv">prefix</span><span class="o">=</span>dbNSFP4.1a_variant.chr
  <span class="nb">echo</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span><span class="o">{</span>1..22<span class="o">}</span>.gz <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>M.gz <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>X.gz <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>Y.gz | <span class="se">\</span>
  xargs <span class="nt">-I</span> <span class="o">{}</span> bash <span class="nt">-c</span> <span class="s2">"zcat {} | sed '1d'"</span>
<span class="o">)</span> | bgzip <span class="nt">-c</span> <span class="o">&gt;</span> dbNSFP4.1a.gz
tabix <span class="nt">-s</span> 1 <span class="nt">-b</span> 2 <span class="nt">-e</span> 2 dbNSFP4.1a.gz
<span class="nb">cd</span> -
vep <span class="nt">-i</span> examples/homo_sapiens_GRCh37.vcf <span class="nt">-o</span> <span class="nb">test</span> <span class="nt">--cache</span> <span class="nt">--force_overwrite</span> <span class="nt">--offline</span> <span class="se">\</span>
    <span class="nt">--plugin</span> dbNSFP,dbNSFP4.1a/dbNSFP4.1a.gz,LRT_score,FATHMM_score,MutationTaster_score
</code></pre> </div></div> <p>Since this release is frozen on Ensembl 94's transcript set, one may prefer to use it independently via its Java programs, e.g.,</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>java <span class="nt">-jar</span> search_dbNSFP41a.jar <span class="nt">-i</span> tryhg19.in <span class="nt">-o</span> tryhg19.out <span class="nt">-v</span> hg19
java <span class="nt">-jar</span> search_dbNSFP41a.jar <span class="nt">-i</span> tryhg38.in <span class="nt">-o</span> tryhg38.out
</code></pre> </div></div> <h3 id="-loftee-">— loftee —</h3> <p>GitHub page: <a href="https://github.com/konradjk/loftee">https://github.com/konradjk/loftee</a>.</p> <p>Reference: MacArthur DG et al (2012). A systematic survey of loss-of-function variants in human protein-coding genes. <em>Science</em> 335:823–828</p> <p>It is actually part of the standard VEP plugins.</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="nb">cd </span>loftee
<span class="c"># human_ancestor_fa</span>
<span class="c">## samtools --version gives 0.1.19</span>
wget http://www.broadinstitute.org/~konradk/loftee/human_ancestor.fa.rz
wget http://www.broadinstitute.org/~konradk/loftee/human_ancestor.fa.rz.fai
<span class="c">## samtools --version gives 1.x</span>
wget https://s3.amazonaws.com/bcbio_nextgen/human_ancestor.fa.gz
wget https://s3.amazonaws.com/bcbio_nextgen/human_ancestor.fa.gz.fai
wget https://s3.amazonaws.com/bcbio_nextgen/human_ancestor.fa.gz.gzi
<span class="c"># conservation_file</span>
wget https://personal.broadinstitute.org/konradk/loftee_data/GRCh37/phylocsf_gerp.sql.gz
wget https://www.broadinstitute.org/~konradk/loftee/phylocsf_data.tsv.gz
wget https://personal.broadinstitute.org/konradk/loftee_data/GRCh37/GERP_scores.final.sorted.txt.gz
wget https://personal.broadinstitute.org/konradk/loftee_data/GRCh37/GERP_scores.exons.txt.gz
<span class="c"># annotation</span>
vep <span class="nt">--id</span> <span class="s2">"1 154426970 154426970 A/C 1"</span> <span class="nt">--species</span> homo_sapiens <span class="nt">-o</span> rs2228145 <span class="nt">--cache</span> <span class="nt">--offline</span> <span class="nt">--force_overwrite</span> <span class="se">\</span>
    <span class="nt">--assembly</span> GRCh37 <span class="nt">--plugin</span> LoF,loftee_path:.,human_ancestor_fa:human_ancestor.fa.gz,conservation_file:phylocsf_gerp.sql.gz
<span class="c"># VEP documentation example</span>
vep <span class="nt">--input_file</span> homo_sapiens_GRCh37.vcf <span class="nt">--output_file</span> <span class="nb">test</span> <span class="nt">--cache</span> <span class="nt">--dir_cache</span> <span class="k">${</span><span class="nv">HPC_WORK</span><span class="k">}</span>/ensembl-vep/.vep <span class="nt">--dir_plugins</span> <span class="k">${</span><span class="nv">HPC_WORK</span><span class="k">}</span>/loftee <span class="nt">--offline</span> <span class="se">\</span>
    <span class="nt">--pick</span> <span class="nt">--force_overwrite</span> <span class="nt">--species</span> homo_sapiens <span class="nt">--assembly</span> GRCh37 <span class="se">\</span>
    <span class="nt">--plugin</span> LoF,loftee_path:.,human_ancestor_fa:human_ancestor.fa.gz,conservation_file:phylocsf_gerp.sql.gz
</code></pre> </div></div> <p>If offers implementation for parsing CSQ field but is also possible with R as described below. Note that if loftee_path uses an absolute path, that path should also be within PERL5LIB, e.g.,</p> <div class="language-plaintext highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code>export PERL5LIB=$PERL5LIB:$HPC_WORK/loftee
</code></pre> </div></div> <p>is put in .bashrc. We see now all files are provided, and there are complaints over rz was not generated by bgzip, so we do</p> <div class="language-bash highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="c"># error</span>
samtools faidx human_ancestor.fa.rz
<span class="c"># rework</span>
<span class="nb">gunzip</span> <span class="nt">-c</span> human_ancestor.fa.rz | bgzip <span class="nt">-c</span> <span class="o">&gt;</span> output.fa.gz
<span class="c"># check and put it back</span>
ll human_ancestor.fa.rz output.fa.gz
<span class="nb">mv </span>output.fa.gz human_ancestor.fa.rz
<span class="c"># now possible and check</span>
samtools faidx human_ancestor.fa.rz
ll human_ancestor.fa.<span class="k">*</span>
</code></pre> </div></div> <h3 id="-r-">— R —</h3> <p>This is a wrapper and <code class="language-plaintext highlighter-rouge notranslate">the Ensembl VEP perl script must be installed in your path</code>. Expected to be slower than the <code class="language-plaintext highlighter-rouge notranslate">--offline</code> mode above, it is relatively easy to set up,</p> <div class="language-r highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="n">BiocManager</span><span class="o">::</span><span class="n">install</span><span class="p">(</span><span class="s2">"ensemblVEP"</span><span class="p">)</span><span class="w">
</span><span class="n">vignette</span><span class="p">(</span><span class="s2">"ensemblVEP"</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">ensemblVEP</span><span class="p">)</span><span class="w">
</span><span class="n">file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"ex2.vcf"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"VariantAnnotation"</span><span class="p">)</span><span class="w">
</span><span class="n">param</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">VEPFlags</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">vcf</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">check_existing</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">symbol</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">terms</span><span class="o">=</span><span class="s2">"SO"</span><span class="p">,</span><span class="n">sift</span><span class="o">=</span><span class="s2">"b"</span><span class="p">,</span><span class="n">polyphen</span><span class="o">=</span><span class="s2">"p"</span><span class="p">))</span><span class="w">
</span><span class="n">vep</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensemblVEP</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="w"> </span><span class="n">param</span><span class="p">)</span><span class="w">
</span><span class="n">info</span><span class="p">(</span><span class="n">vep</span><span class="p">)</span><span class="o">$</span><span class="n">CSQ</span><span class="w">
</span><span class="n">csq</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">parseCSQToGRanges</span><span class="p">(</span><span class="n">vep</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">csq</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="c1"># clinvar</span><span class="w">
</span><span class="n">param</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">VEPFlags</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">vcf</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">output_file</span><span class="o">=</span><span class="s2">"clinvar.vcf"</span><span class="p">,</span><span class="n">force_overwrite</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">assembly</span><span class="o">=</span><span class="s2">"GRCh37"</span><span class="p">,</span><span class="n">port</span><span class="o">=</span><span class="m">3337</span><span class="p">,</span><span class="w">
                  </span><span class="n">custom</span><span class="o">=</span><span class="s2">"clinvar_GRCh37.vcf.gz,ClinVar,vcf,exact,0,CLNSIG,CLNREVSTAT,CLNDN"</span><span class="p">))</span><span class="w">
</span><span class="n">ensemblVEP</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="w"> </span><span class="n">param</span><span class="p">)</span><span class="w">
</span></code></pre> </div></div> <p>The second <code class="language-plaintext highlighter-rouge notranslate">ensemblVEP</code> obtains <code class="language-plaintext highlighter-rouge notranslate">clinvar.vcf</code> and <code class="language-plaintext highlighter-rouge notranslate">clinvar.vcf_summary.html</code>. Annotation is made to a VCF file, and returns data with unparsed ‘CSQ'.</p> <p>The facility to parse the CSQ column of a VCF object is done for the documentation example.</p> <div class="language-r highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="c1"># VCF output from the VEP web interface or the call above</span><span class="w">
</span><span class="n">vep</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"INF1.merge.trans.vcf"</span><span class="w">
</span><span class="c1"># Parse into a GRanges and include the 'VCFRowID' column.</span><span class="w">
</span><span class="n">vcf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readVcf</span><span class="p">(</span><span class="n">vep</span><span class="p">,</span><span class="w"> </span><span class="s2">"hg19"</span><span class="p">)</span><span class="w">
</span><span class="n">csq</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">parseCSQToGRanges</span><span class="p">(</span><span class="n">vep</span><span class="p">,</span><span class="w"> </span><span class="n">VCFRowID</span><span class="o">=</span><span class="n">rownames</span><span class="p">(</span><span class="n">vcf</span><span class="p">))</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">mcols</span><span class="p">(</span><span class="n">csq</span><span class="p">),</span><span class="n">file</span><span class="o">=</span><span class="s2">"INF1.merge.trans.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span></code></pre> </div></div> <p>The dbNSFP counterpart is also possible</p> <div class="language-r highlighter-rouge notranslate"><div class="highlight"><pre class="highlight"><code><span class="n">BiocManager</span><span class="o">::</span><span class="n">install</span><span class="p">(</span><span class="s2">"myvariant"</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">VariantAnnotation</span><span class="p">)</span><span class="w">
</span><span class="n">file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"dbsnp_mini.vcf"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"myvariant"</span><span class="p">)</span><span class="w">
</span><span class="n">vcf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readVcf</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="w"> </span><span class="n">genome</span><span class="o">=</span><span class="s2">"hg19"</span><span class="p">)</span><span class="w">
</span><span class="n">rowRanges</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">myvariant</span><span class="p">)</span><span class="w">
</span><span class="n">hgvs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">formatHgvs</span><span class="p">(</span><span class="n">vcf</span><span class="p">,</span><span class="w"> </span><span class="n">variant_type</span><span class="o">=</span><span class="s2">"snp"</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">hgvs</span><span class="p">)</span><span class="w">
</span><span class="n">getVariants</span><span class="p">(</span><span class="n">hgvs</span><span class="p">)</span><span class="w">
</span><span class="n">rsids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="s2">"rs"</span><span class="p">,</span><span class="w"> </span><span class="n">info</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span><span class="o">$</span><span class="n">RS</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">rsids</span><span class="p">)</span><span class="w">
</span><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">queryVariants</span><span class="p">(</span><span class="n">q</span><span class="o">=</span><span class="n">rsids</span><span class="p">,</span><span class="w"> </span><span class="n">scopes</span><span class="o">=</span><span class="s2">"dbsnp.rsid"</span><span class="p">,</span><span class="w"> </span><span class="n">fields</span><span class="o">=</span><span class="s2">"all"</span><span class="p">)</span><span class="w">
</span><span class="n">fields</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">res</span><span class="p">)</span><span class="w">
</span><span class="n">cadd</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s1">'cadd'</span><span class="p">,</span><span class="n">fields</span><span class="p">)</span><span class="w">
</span><span class="n">res</span><span class="p">[</span><span class="n">fields</span><span class="p">[</span><span class="n">cadd</span><span class="p">]]</span><span class="w">
</span></code></pre> </div></div> <p>Note that the CSQ field could also be handled by bcftools split-vep plugin, see <a href="http://samtools.github.io/bcftools/howtos/plugin.split-vep.html">http://samtools.github.io/bcftools/howtos/plugin.split-vep.html</a>.</p> <h3 id="-docker-">— docker —</h3> <p>See <code class="language-plaintext highlighter-rouge notranslate">docker/Dockerfile </code> from the GitHub directory above, or https://github.com/Ensembl/ensembl-vep.</p> <h3 id="-virtual-machine-">— Virtual machine —</h3> <p>See http://www.ensembl.org/info/data/virtual_machine.html which is possibly best for MicroSoft Windows and is not pursued here.</p> <p>ENSEMBL-synonym translation (hg19) file</p> </div> </div> <div class="navigation-bottom d-flex flex-justify-between py-3" role="navigation" aria-label="footer navigation"> <div class="prev"><a href="/applications/tabix.html" class="btn" title="tabix" accesskey="p" rel="prev"> <i class="fa fa-arrow-circle-left"></i> Previous </a></div> <div class="next"><a href="/applications/xlsx.html" class="btn" title="R/xlsx" accesskey="n" rel="next"> Next <i class="fa fa-arrow-circle-right"></i> </a></div> </div><hr> <div class="copyright text-center text-gray" role="contentinfo"> <i class="fa fa-copyright"></i> <span class="time">2021,</span> <a class="text-gray" href="https://github.com/cambridge-ceu" rel="noreferrer" target="_blank">cambridge-ceu</a> Revision <a class="text-gray" href="https://github.com/cambridge-ceu/csd3/commit/b5f8beb971e1e7020fb16231556a1ec40359efd4" title="b5f8beb971e1e7020fb16231556a1ec40359efd4" rel="noreferrer" target="_blank">b5f8beb</a> <br> <div class="generator"> Built with <a href="https://pages.github.com" rel="noreferrer" target="_blank" title="github-pages v209">GitHub Pages</a> using a <a href="https://github.com/rundocs/jekyll-rtd-theme" rel="noreferrer" target="_blank" title="rundocs/jekyll-rtd-theme v2.0.10">theme</a> provided by <a href="https://rundocs.io" rel="noreferrer" target="_blank">RunDocs</a>. </div> </div> </div> </div> <div class="addons-wrap d-flex flex-column overflow-y-auto"> <div class="status d-flex flex-justify-between p-2"> <div class="title p-1"> <i class="fa fa-book"></i> csd3 </div> <div class="branch p-1"> <span class="name"> gh-pages </span> <i class="fa fa-caret-down"></i> </div> </div> <div class="addons d-flex flex-column height-full p-2 d-none"> <dl> <dt>GitHub</dt> <dd> <a href="https://github.com/cambridge-ceu/csd3" title="Stars: "> <i class="fa fa-github"></i> Homepage </a> </dd> <dd> <a href="" title="Open issues: "> <i class="fa fa-question-circle-o"></i> Issues </a> </dd> <dd> <a href="https://github.com/cambridge-ceu/csd3/zipball/gh-pages" title="Size: Kb"> <i class="fa fa-download"></i> Download </a> </dd> </dl> <hr> <div class="license f6 pb-2"> This <a href="/" title="csd3">Software</a> is under the terms of <a href="https://github.com/cambridge-ceu/csd3">The Unlicense</a>. </div> </div> </div> <script src="/assets/js/jquery.min.js"></script><script src="/assets/js/theme.js"></script> </body> </html>